{
 "metadata": {
  "name": "",
  "signature": "sha256:d2465f1badc2809c4e2f69ee381b4ff3350047efce6fa52f14c8b67d45f9ce2e"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "GFF phage file validation program"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###README:\n",
      "\n",
      "This program will check your GFF file. It checks for proper format and use of proper characters  (i.e. no illegal charachters used). \n",
      "\n",
      "**Before you run the program:** copy your GFF file into the \"input\" folder on the Desktop. Make sure you know the EXACT name of the file, it is best to avoid spaces in filenames (rename your files if necessary).  \n",
      "\n",
      "**To run the program:** click anywhere in the code box below near the \"#STEP 1: Click inside this box Here\". You should see the blinking curser appear near where you clicked, once you see the curser hit shift-return (i.e. hold shift while hitting the return key).\n",
      "\n",
      "**What will happen:** the first thing to appear will be a small text entry box at the bottom of the page asking for the name of your gff file. Type the name of the gff file into the box. Once the box has the correct name, hit the return key (no shift). \n",
      "\n",
      "Your results should appear directly below the question box."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#STEP 1: Click inside this box Here\n",
      "\n",
      "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
      "\n",
      "fullGFFFileName = \"/Users/chris/Desktop/input/\" + gffFileName\n",
      "\n",
      "try:\n",
      "    with open(fullGFFFileName) as gffFile:\n",
      "        gffFileContents = gffFile.readlines()\n",
      "except:\n",
      "    gffFileContents = []\n",
      "    print '''Well that didn't work! Could not get GFF file. Are you sure it is in the input folder?\n",
      "    Double check spelling, no spaces in file name, try again. If that fails get help.'''\n",
      "    sys.exit()\n",
      "print \"\\n\\n================================== Results ==================================\\n\"\n",
      "        \n",
      "firstline = gffFileContents[0]\n",
      "\n",
      "print len(firstline)\n",
      "if not validHeader(firstline):\n",
      "    printFailureMessage(\"Check for unix formated file, Header line\")\n",
      "    print firstline\n",
      "    sys.exit()\n",
      "\n",
      "    \n",
      "print \"checking line: \", \n",
      "for lineNum,wholeLine in enumerate(gffFileContents[1:]):\n",
      "    lineNumber = str(lineNum+1)\n",
      "    print \" \" + lineNumber,\n",
      "    \n",
      "    try:\n",
      "        if not validTabStructure(wholeLine):                          #check whole line for proper 8 tab structure\n",
      "            failureMessage = \"Line number: \" + lineNumber + \" tab structure (i.e. 9 entries and 8 tabs)\"\n",
      "            printFailureMessage(failureMessage)\n",
      "            sys.exit()\n",
      "    except:\n",
      "        print \"double check that all functions are defined!\"\n",
      "    \n",
      "    \n",
      "    line = wholeLine.strip().split(\"\\t\")\n",
      "    \n",
      "    if not validSeqname(line[0]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" seq name of \" + line[0]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "    \n",
      "    if not validSource(line[1]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" source definition of \" + line[1] \n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "    if not validType(line[2]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" type definition of \" + line[2]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "    if not validCoordinates(line[3],line[4]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" coordinates left: \" + line[3] + \" right: \" + line[4] + \" have\"\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "    if not validScore(line[5]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" score of \" + line[5]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "    if not validStrand(line[6]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" strand entry of \" + line[6]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "    if not validPhase(line[7]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" phase of \" + line[7]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "    \n",
      "    \n",
      "    if not validAttributes(line[8]):\n",
      "        failMessage = \"Line number: \" + lineNumber + \" attribute entry of: \" + line[8]\n",
      "        printFailureMessage(failMessage)\n",
      "        sys.exit()\n",
      "\n",
      "print \"\\n\\nCongratulations! all tests passed\\nEverything below can be used in DNA Master\"\n",
      "print \"=========================== DNA Master documentation ========================\"\n",
      "text = createDNAMasterFile(gffFileContents)\n",
      "print text"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Enter the name of the gff file (BE EXACT): Mori-3.txt\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "\n",
        "================================== Results ==================================\n",
        "\n",
        "16\n",
        "checking line:   1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99  100  101  102  103  104  105  106  107  108  109  110  111  112  113  114  115  116  117  118  119  120  121 \n",
        "\n",
        "Congratulations! all tests passed\n",
        "Everything below can be used in DNA Master\n",
        "=========================== DNA Master documentation ========================\n",
        "CDS complement (152 - 334)\n",
        "\n",
        "CDS complement (413 - 604)\n",
        "\n",
        "CDS complement (615 - 770)\n",
        "\n",
        "CDS complement (773 - 1150)\n",
        "\n",
        "CDS complement (1147 - 1716)\n",
        "\n",
        "CDS complement (1508 - 2098)\n",
        "\n",
        "CDS complement (2095 - 2187)\n",
        "\n",
        "CDS complement (2184 - 2378)\n",
        "\n",
        "CDS complement (2371 - 2700)\n",
        "\n",
        "CDS complement (2729 - 3349)\n",
        "\n",
        "CDS complement (3346 - 4167)\n",
        "\n",
        "CDS 4656 - 4943\n",
        "\n",
        "CDS 4936 - 5685\n",
        "\n",
        "CDS 5685 - 5840\n",
        "\n",
        "CDS 5866 - 6603\n",
        "\n",
        "CDS 6687 - 7001\n",
        "\n",
        "CDS 6994 - 7305\n",
        "\n",
        "CDS 7293 - 7529\n",
        "\n",
        "CDS 7526 - 7756\n",
        "\n",
        "CDS 7753 - 10116\n",
        "\n",
        "CDS complement (10932 - 11231)\n",
        "\n",
        "CDS 11276 - 11710\n",
        "\n",
        "CDS 11707 - 11943\n",
        "\n",
        "CDS 12010 - 12363\n",
        "\n",
        "CDS 12360 - 12587\n",
        "\n",
        "CDS 12859 - 13158\n",
        "\n",
        "CDS 13118 - 14629\n",
        "\n",
        "CDS 14688 - 15917\n",
        "\n",
        "CDS 15914 - 16498\n",
        "\n",
        "CDS 16491 - 17360\n",
        "\n",
        "CDS 17357 - 18049\n",
        "\n",
        "CDS 18046 - 18666\n",
        "\n",
        "CDS 18735 - 19400\n",
        "\n",
        "CDS 19397 - 19534\n",
        "\n",
        "CDS 19687 - 20919\n",
        "\n",
        "CDS 20919 - 21155\n",
        "\n",
        "CDS 21165 - 21770\n",
        "\n",
        "CDS 21784 - 21921\n",
        "\n",
        "CDS 21924 - 22283\n",
        "\n",
        "CDS 22280 - 22600\n",
        "\n",
        "CDS 22584 - 22970\n",
        "\n",
        "CDS 22963 - 23415\n",
        "\n",
        "CDS 23444 - 24259\n",
        "\n",
        "CDS complement (24356 - 24451)\n",
        "\n",
        "CDS complement (24448 - 24624)\n",
        "\n",
        "CDS 24621 - 24812\n",
        "\n",
        "CDS 24809 - 25264\n",
        "\n",
        "CDS 25335 - 25889\n",
        "\n",
        "CDS 25942 - 26286\n",
        "\n",
        "CDS 26286 - 26885\n",
        "\n",
        "CDS 27046 - 31716\n",
        "\n",
        "CDS 31754 - 33472\n",
        "\n",
        "CDS 33516 - 35252\n",
        "\n",
        "CDS 35298 - 36140\n",
        "\n",
        "CDS 36137 - 38647\n",
        "\n",
        "CDS 38640 - 40130\n",
        "\n",
        "CDS 40142 - 40543\n",
        "\n",
        "CDS 40547 - 40948\n",
        "\n",
        "CDS 40958 - 41182\n",
        "\n",
        "CDS 41195 - 41383\n",
        "\n",
        "CDS 41383 - 42069\n",
        "\n",
        "CDS 42069 - 42188\n",
        "\n",
        "CDS 42238 - 43458\n",
        "\n",
        "CDS 43460 - 44500\n",
        "\n",
        "CDS 44513 - 44809\n",
        "\n",
        "CDS 44806 - 45105\n",
        "\n",
        "CDS 45236 - 45592\n",
        "\n",
        "CDS 45583 - 46080\n",
        "\n",
        "CDS 46086 - 46517\n",
        "\n",
        "CDS complement (46486 - 46893)\n",
        "\n",
        "CDS complement (47021 - 47254)\n",
        "\n",
        "CDS complement (47251 - 47628)\n",
        "\n",
        "CDS complement (47677 - 48108)\n",
        "\n",
        "CDS complement (48151 - 48582)\n",
        "\n",
        "CDS complement (48579 - 49304)\n",
        "\n",
        "CDS complement (49349 - 49981)\n",
        "\n",
        "CDS complement (50097 - 51140)\n",
        "\n",
        "CDS complement (51265 - 52230)\n",
        "\n",
        "CDS complement (52227 - 52493)\n",
        "\n",
        "CDS complement (52490 - 52666)\n",
        "\n",
        "CDS complement (52707 - 53015)\n",
        "\n",
        "CDS complement (52982 - 53296)\n",
        "\n",
        "CDS complement (53293 - 53604)\n",
        "\n",
        "CDS complement (53620 - 54474)\n",
        "\n",
        "CDS complement (54528 - 54839)\n",
        "\n",
        "CDS complement (55215 - 56723)\n",
        "\n",
        "CDS complement (56720 - 57058)\n",
        "\n",
        "CDS complement (57102 - 57551)\n",
        "\n",
        "CDS complement (57657 - 57983)\n",
        "\n",
        "CDS complement (57988 - 58260)\n",
        "\n",
        "CDS complement (58257 - 59057)\n",
        "\n",
        "CDS complement (59109 - 59366)\n",
        "\n",
        "CDS complement (59363 - 59623)\n",
        "\n",
        "CDS complement (59706 - 61406)\n",
        "\n",
        "CDS complement (62029 - 62418)\n",
        "\n",
        "CDS complement (62539 - 62970)\n",
        "\n",
        "CDS complement (63013 - 63303)\n",
        "\n",
        "CDS complement (63300 - 63479)\n",
        "\n",
        "CDS complement (63472 - 63801)\n",
        "\n",
        "CDS complement (63798 - 64205)\n",
        "\n",
        "CDS complement (64205 - 64612)\n",
        "\n",
        "CDS complement (64652 - 64966)\n",
        "\n",
        "CDS complement (64998 - 65270)\n",
        "\n",
        "CDS complement (65359 - 65529)\n",
        "\n",
        "CDS complement (65639 - 65833)\n",
        "\n",
        "CDS complement (65830 - 66162)\n",
        "\n",
        "CDS complement (66182 - 66439)\n",
        "\n",
        "CDS complement (66436 - 66657)\n",
        "\n",
        "CDS complement (66644 - 67159)\n",
        "\n",
        "CDS complement (67149 - 67406)\n",
        "\n",
        "CDS complement (67410 - 67682)\n",
        "\n",
        "CDS complement (67679 - 67834)\n",
        "\n",
        "CDS complement (67831 - 67932)\n",
        "\n",
        "CDS complement (67929 - 68129)\n",
        "\n",
        "CDS complement (68126 - 68305)\n",
        "\n",
        "CDS complement (68351 - 68656)\n",
        "\n",
        "CDS complement (68790 - 69047)\n",
        "\n",
        "CDS complement (69044 - 69187)\n",
        "\n",
        "CDS complement (69184 - 69339)\n",
        "\n",
        "CDS complement (69459 - 69629)\n",
        "\n",
        "CDS complement (70835 - 71557)\n",
        "\n",
        "ORIGIN\n"
       ]
      }
     ],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Definitions of Functions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%writefile gffTester_nose.py\n",
      "import re,sys\n",
      "\n",
      "def validHeader(line):\n",
      "    \"\"\"\n",
      "    Return True if line is a valid gff3 header.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line' - line to check\n",
      "    \"\"\"\n",
      "    if len(line) != 16:\n",
      "        return False\n",
      "    elif line.strip() == \"##gff-version 3\":\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "    \n",
      "def validTabStructure(line):\n",
      "    \"\"\"\n",
      "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line': line to check\n",
      "    \"\"\"\n",
      "    theLine = line.strip().split(\"\\t\")  #strip off extra charachters and split the line on tab\n",
      "\n",
      "    if len(theLine) != 9:               # 9 columns in a gff file, so return False if not exactly 9\n",
      "        return False\n",
      "    else:\n",
      "        return True\n",
      "\n",
      "def validSeqname(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
      "    the sequence name on the Gbrowse database and have only valid characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of entry in column 1 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
      "\n",
      "def validSource(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
      "    which only need validity of the characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)\n",
      "\n",
      "def validType(text, validTypes=None):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
      "    For phage this should be one of types ['gene','mRNA','exon']\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    - 'validTypes': List of valid types for checking if not the default 3\n",
      "    \"\"\"\n",
      "    if validTypes is None:\n",
      "        validTypes = {'gene','mRNA','exon', 'contig'}\n",
      "    \n",
      "    return text in validTypes\n",
      "\n",
      "\n",
      "def validScore(score):\n",
      "    \"\"\"\n",
      "    Checks a score (entries in column 6), should be a number or \".\"\n",
      "    \n",
      "    Parameters:\n",
      "    - 'score': score from column 6\n",
      "    \"\"\"\n",
      "    if score == \".\":\n",
      "        return True\n",
      "    \n",
      "    try:\n",
      "        x = float(score)\n",
      "    except:\n",
      "        isScore = False\n",
      "        pass\n",
      "    else:\n",
      "        isScore = True        \n",
      "\n",
      "    return isScore\n",
      "\n",
      "def validStrand(strand):\n",
      "    \"\"\"\n",
      "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'strand': strand entry from column 7\n",
      "    \"\"\"\n",
      "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
      "    \n",
      "    return strand in validValues\n",
      "\n",
      "def validPhase(phase):\n",
      "    \"\"\"\n",
      "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'phase': strand entry from column 8\n",
      "    \"\"\"\n",
      "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
      "    \n",
      "    return phase in validValues\n",
      "\n",
      "def validAttributes(attributes):\n",
      "    '''\n",
      "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
      "    It is OK to have spaces in the values entry but not the Key value\n",
      "    \n",
      "    \n",
      "    Parameters:\n",
      "    - 'attributes': entire entry from column 9\n",
      "    '''\n",
      "    #ok to have a null string\n",
      "    if attributes == '.':\n",
      "        return True\n",
      "    \n",
      "    #checking for invalid characters goes here if any found\n",
      "    \n",
      "    if bool(charCheck(attributes, search=re.compile(r'[^a-zA-Z0-9.=;_ ]').search)):\n",
      "        return False\n",
      "    \n",
      "    \n",
      "    #ok go ahead and split into the underlyine key=value items \n",
      "\n",
      "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
      "    \n",
      "    validity = True\n",
      "    \n",
      "    for attrPair in attrPairs:\n",
      "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
      "            continue\n",
      "            \n",
      "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
      "            return False\n",
      "        \n",
      "        attrKey, attrValue = attrPair.split('=')\n",
      "        \n",
      "        if len(attrKey) < 1:                #must have an entry for a key\n",
      "            validity = False\n",
      "        \n",
      "        if charCheck(attrKey):              #check key for valid charachters\n",
      "            validity = False\n",
      "        \n",
      "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
      "        if charCheck(attrValue.replace(' ','')):\n",
      "            validity = False\n",
      "            \n",
      "    return True\n",
      "\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters NOT a-zA-Z0-9.=; and returns True if invalid character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    return bool(search(str))\n",
      "\n",
      "def validCoordinate(coord):\n",
      "    \"\"\"\n",
      "    Checks a coordinate (entries in column 4 or 5), should be a positive integer\n",
      "    \n",
      "    Parameters:\n",
      "    - 'coord': coordinate from column 4 or 5\n",
      "    \"\"\"\n",
      "    \n",
      "    return not bool(charCheck(coord, search=re.compile(r'[^0-9]').search))\n",
      "\n",
      "def validCoordinates(leftCoord, rightCoord):\n",
      "    \"\"\"\n",
      "    Checks a coordinates (entries in column 4 and 5), should be a positive integers\n",
      "    and the left Coordinate should be smaller than the right Coordinate\n",
      "    \n",
      "    Parameters:\n",
      "    - 'leftCoord':  coordinate from column 4\n",
      "    - 'rightCoord': coordinate from column 5\n",
      "\n",
      "    \"\"\"\n",
      "    return (validCoordinate(leftCoord) and validCoordinate(rightCoord) and int(leftCoord) <= int(rightCoord))\n",
      "\n",
      "\n",
      "def printFailureMessage(failType):\n",
      "    print \"\\n##### Fatal Error #####\"\n",
      "    print failType + \" failed. No other tests run.\"\n",
      "    print \"\"\"\n",
      "================================ End Results ================================        \n",
      "Computer messages follow below OK to ignore, you should fix file and try again.\n",
      "    \"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Overwriting gffTester_nose.py\n"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Create DNA Master File"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def createDNAMasterFile(gffFileContents):\n",
      "    \"\"\"\n",
      "    Takes in a list of GFF lines, checks for headers and passes lines of type gene to parser.\n",
      "    Returns the complete text of DNA Master text file.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'gffFileContents':  List of text entries, each element a line from GFF# file.\n",
      "    \"\"\"\n",
      "    \n",
      "    text = ''\n",
      "    for wholeLine in gffFileContents:\n",
      "        line = wholeLine.strip().split(\"\\t\")\n",
      "        \n",
      "        if len(line) < 8:\n",
      "            continue\n",
      "        \n",
      "        if line[2] == 'gene':\n",
      "            entry = parseGffLine(line)\n",
      "            text += entry\n",
      "            \n",
      "    epilog = \"ORIGIN\"\n",
      "    text += epilog\n",
      "        \n",
      "    return text\n",
      "            \n",
      "        \n",
      "def parseGffLine(line):\n",
      "    \"\"\"\n",
      "    Takes in a single GFF entry and returns the corresponding DNA Master entry.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line':  List of entries from a single GFF entry.\n",
      "    \"\"\"    \n",
      "    if line[6] == \"+\":\n",
      "        return \"CDS \" + line[3] + \" - \" + line[4] + \"\\n\\n\"\n",
      "    else:\n",
      "        return \"CDS complement (\" + line[3] + \" - \" + line[4] + \")\\n\\n\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Test code"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#%%writefile test_code.py\n",
      "\n",
      "from gffTester_nose import *\n",
      "\n",
      "'''\n",
      "need to check the following functions:\n",
      "\n",
      "def validHeader (line): TEST valid, extra space, extra tab, garbled\n",
      "def validTabStructure(line): test: 7 tabs, 8 tabs, 9 tabs\n",
      "def validSeqname(text): SKIP, just a \"not charCheck\" for now\n",
      "def validSource(text): SKIP, just a \"not charCheck\" for now\n",
      "def validType(text, validTypes=None): TEST: pass with any 3 defaults 'gene','mRNA','exon'; fails on none item, passes with list\n",
      "def validCoordinate(coord): TEST: fails with text, float, negative int; passes with positive int\n",
      "def validScore(score):\n",
      "def validStrand(strand):\n",
      "def validPhase(phase):\n",
      "def validAttributes(attributes):\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "def validGene(seq, line):       \n",
      "def fastaRead(fasta_File):\n",
      "def countStopCodons(seq, strand):\n",
      "'''\n",
      "\n",
      "def test_validHeader_1():\n",
      "    'valid header tests as valid'\n",
      "    result = validHeader('##gff-version 3\\n')\n",
      "    assert result == True\n",
      "\n",
      "def test_validHeader_2():\n",
      "    'header with extra space fails'\n",
      "    result = validHeader('##gff-version 3 \\n')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validHeader_3():\n",
      "    'header with extra tab fails'\n",
      "    result = validHeader('##gff-version 3\\t\\n')\n",
      "    assert result == False\n",
      "\n",
      "    \n",
      "def test_validHeader_4():\n",
      "    'garbled header fails'\n",
      "    result = validHeader('##gff-Version 3\\n')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validTabStructure_1():\n",
      "    'fails 7 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\n\")\n",
      "    assert result == False\n",
      "\n",
      "def test_validTabStructure_2():\n",
      "    'passes 8 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\n\")\n",
      "    assert result == True\n",
      "    \n",
      "def test_validTabStructure_3():\n",
      "    'fails 9 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\t10\\n\")\n",
      "    assert result == False\n",
      "    \n",
      "def test_validType_1():\n",
      "    'Type passes with \"gene\"'\n",
      "    result = validType('gene')\n",
      "    assert result == True\n",
      "\n",
      "def test_validType_2():\n",
      "    'Type passes with \"mRNA\"'\n",
      "    result = validType('mRNA')\n",
      "    assert result == True\n",
      "\n",
      "def test_validType_3():\n",
      "    'Type passes with \"exon\"'\n",
      "    result = validType('exon')\n",
      "    assert result == True\n",
      "    \n",
      "def test_validType_4():\n",
      "    'Type fails with mrna'\n",
      "    result = validType('mrna')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validType_5():\n",
      "    'Type passes with specified list'\n",
      "    result = validType('mrna', ['mRNA', 'mrna'])\n",
      "    assert result == True\n",
      "    \n",
      "\n",
      "def test_validCoordinate_1():\n",
      "    'Coord fails with text as coordinate'\n",
      "    result = validCoordinate('a')\n",
      "    assert result == False\n",
      "\n",
      "def test_validCoordinate_2():\n",
      "    'Coord fails with float'\n",
      "    result = validCoordinate('2.0')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validCoordinate_3():\n",
      "    'Coord fails with negative int'\n",
      "    result = validCoordinate('-3')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validCoordinate_4():\n",
      "    'Coord passes with positive integer'\n",
      "    result = validCoordinate('100')\n",
      "    assert result == True\n",
      "    \n",
      "def test_validCoordinates_1():\n",
      "    'Coordinates fails with left larger than right'\n",
      "    result = validCoordinates('5','4')\n",
      "    assert result == False\n",
      "\n",
      "def test_validCoordinates_2():\n",
      "    'passes with left equal to right'\n",
      "    result = validCoordinates('5','5')\n",
      "    assert result == True\n",
      "    \n",
      "def test_validCoordinates_3():\n",
      "    'passes with left smaller than right'\n",
      "    result = validCoordinates('4','5')\n",
      "    assert result == True\n",
      "    \n",
      "def test_validAttributes_1():\n",
      "    'Attributes passes with null string'\n",
      "    result = validAttributes(\"\")\n",
      "    assert result == True\n",
      "\n",
      "def test_validAttributes_2():\n",
      "    'Attributes passes a=b'\n",
      "    result = validAttributes(\"a=b\")\n",
      "    assert result == True\n",
      "\n",
      "def test_validAttributes_3():\n",
      "    'attrubutes passes with \"a=b;\"'\n",
      "    result = validAttributes(\"a=b;\")\n",
      "    assert result == True\n",
      "\n",
      "def test_validAttributes_4():\n",
      "    'Attributes failes with \"a=bc=d\"'\n",
      "    result = validAttributes(\"a=bc=d\")\n",
      "    assert result == False\n",
      "    \n",
      "def test_validAttributes_5():\n",
      "    'attrubutes passes with \"a=b;c=d\"'\n",
      "    result = validAttributes(\"a=b;c=d\")\n",
      "    assert result == True\n",
      "\n",
      "def test_validAttributes_6():\n",
      "    'attrubutes passes with \"af=b f;c=d\"'\n",
      "    result = validAttributes(\"af=b f;c=d\")\n",
      "    assert result == True\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!nosetests -v test_code.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "valid header tests as valid ... ok\r\n",
        "header with extra space fails ... ok\r\n",
        "header with extra tab fails ... ok\r\n",
        "garbled header fails ... ok\r\n",
        "fails 7 tab structure ... ok\r\n",
        "passes 8 tab structure ... ok\r\n",
        "fails 9 tab structure ... ok\r\n",
        "Type passes with \"gene\" ... ok\r\n",
        "Type passes with \"mRNA\" ... ok\r\n",
        "Type passes with \"exon\" ... ok\r\n",
        "Type fails with mrna ... ok\r\n",
        "Type passes with specified list ... ok\r\n",
        "Coord fails with text as coordinate ... ok\r\n",
        "Coord fails with float ... ok\r\n",
        "Coord fails with negative int ... ok\r\n",
        "Coord passes with positive integer ... ok\r\n",
        "Coordinates fails with left larger than right ... ok\r\n",
        "passes with left equal to right ... ok\r\n",
        "passes with left smaller than right ... ok\r\n",
        "Attributes passes with null string ... ok\r\n",
        "Attributes passes a=b ... ok\r\n",
        "attrubutes passes with \"a=b;\" ... ok\r\n",
        "Attributes failes with \"a=bc=d\" ... ok\r\n",
        "attrubutes passes with \"a=b;c=d\" ... ok\r\n",
        "attrubutes passes with \"af=b f;c=d\" ... ok\r\n",
        "\r\n",
        "----------------------------------------------------------------------\r\n",
        "Ran 25 tests in 0.004s\r\n",
        "\r\n",
        "OK\r\n"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print len(\"##gff-version 3\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "15\n"
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}