{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GFF phage file validation program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###README:\n",
    "\n",
    "This program will check your GFF file. It checks for proper format and use of proper characters  (i.e. no illegal charachters used). \n",
    "\n",
    "**Before you run the program:** copy your GFF file into the \"input\" folder on the Desktop. Make sure you know the EXACT name of the file, it is best to avoid spaces in filenames (rename your files if necessary).  \n",
    "\n",
    "**To run the program:** click anywhere in the code box below near the \"#STEP 1: Click inside this box Here\". You should see the blinking curser appear near where you clicked, once you see the curser hit shift-return (i.e. hold shift while hitting the return key).\n",
    "\n",
    "**What will happen:** the first thing to appear will be a small text entry box at the bottom of the page asking for the name of your gff file. Type the name of the gff file into the box. Once the box has the correct name, hit the return key (no shift). \n",
    "\n",
    "Your results should appear directly below the question box."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the name of the gff file (BE EXACT): Joy99_v2.txt\n",
      "\n",
      "\n",
      "================================== Results ==================================\n",
      "\n",
      "16\n",
      "checking line:   1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91  92  93  94 \n",
      "\n",
      "Congratulations! all tests passed\n",
      "Everything below can be used in DNA Master\n",
      "=========================== DNA Master documentation ========================\n",
      "CDS 79 - 273\n",
      "  /gene=Joy99.1\n",
      "CDS 270 - 455\n",
      "  /gene=Joy99.2\n",
      "CDS 452 - 706\n",
      "  /gene=Joy99.3\n",
      "CDS 703 - 930\n",
      "  /gene=Joy99.4\n",
      "CDS 1156 - 1359\n",
      "  /gene=Joy99.5\n",
      "CDS 1352 - 1579\n",
      "  /gene=Joy99.6\n",
      "CDS 1563 - 2987\n",
      "  /gene=Joy99.7\n",
      "CDS 2999 - 4555\n",
      "  /gene=Joy99.8\n",
      "CDS 4560 - 7070\n",
      "  /gene=Joy99.9\n",
      "CDS 7067 - 7252\n",
      "  /gene=Joy99.10\n",
      "CDS 7289 - 7825\n",
      "  /gene=Joy99.11\n",
      "CDS 7900 - 8835\n",
      "  /gene=Joy99.12\n",
      "CDS 8989 - 9333\n",
      "  /gene=Joy99.13\n",
      "CDS 9330 - 9686\n",
      "  /gene=Joy99.14\n",
      "CDS 9667 - 9948\n",
      "  /gene=Joy99.15\n",
      "CDS 9945 - 10370\n",
      "  /gene=Joy99.16\n",
      "CDS 10470 - 11084\n",
      "  /gene=Joy99.17\n",
      "CDS 11198 - 11635\n",
      "  /gene=Joy99.18\n",
      "CDS 11692 - 12045\n",
      "  /gene=Joy99.19\n",
      "CDS 12045 - 15878\n",
      "  /gene=Joy99.20\n",
      "CDS 15980 - 17116\n",
      "  /gene=Joy99.21\n",
      "CDS 17117 - 18883\n",
      "  /gene=Joy99.22\n",
      "CDS 18883 - 19359\n",
      "  /gene=Joy99.23\n",
      "CDS 19443 - 20525\n",
      "  /gene=Joy99.24\n",
      "CDS 20532 - 20840\n",
      "  /gene=Joy99.25\n",
      "CDS 20841 - 23273\n",
      "  /gene=Joy99.26\n",
      "CDS 23285 - 24298\n",
      "  /gene=Joy99.27\n",
      "CDS 24382 - 24753\n",
      "  /gene=Joy99.28\n",
      "CDS 24771 - 26426\n",
      "  /gene=Joy99.29\n",
      "CDS 26423 - 27295\n",
      "  /gene=Joy99.30\n",
      "CDS 27306 - 27746\n",
      "  /gene=Joy99.31\n",
      "CDS 27749 - 28087\n",
      "  /gene=Joy99.32\n",
      "CDS 28084 - 28335\n",
      "  /gene=Joy99.33\n",
      "CDS 28322 - 29530\n",
      "  /gene=Joy99.34\n",
      "CDS 29723 - 30154\n",
      "  /gene=Joy99.35\n",
      "CDS 30151 - 30513\n",
      "  /gene=Joy99.36\n",
      "CDS 30590 - 31276\n",
      "  /gene=Joy99.37\n",
      "CDS 31392 - 31790\n",
      "  /gene=Joy99.38\n",
      "CDS complement (31864 - 32166)\n",
      "  /gene=Joy99.39\n",
      "CDS 32524 - 33612\n",
      "  /gene=Joy99.40\n",
      "CDS complement (33716 - 34510)\n",
      "  /gene=Joy99.41\n",
      "CDS complement (34645 - 35025)\n",
      "  /gene=Joy99.42\n",
      "CDS 35188 - 35445\n",
      "  /gene=Joy99.43\n",
      "CDS 35442 - 35708\n",
      "  /gene=Joy99.44\n",
      "CDS 35710 - 36060\n",
      "  /gene=Joy99.45\n",
      "CDS 36239 - 36406\n",
      "  /gene=Joy99.46\n",
      "CDS 36403 - 36615\n",
      "  /gene=Joy99.47\n",
      "CDS 36612 - 37403\n",
      "  /gene=Joy99.48\n",
      "CDS 37400 - 37663\n",
      "  /gene=Joy99.49\n",
      "CDS 37660 - 38601\n",
      "  /gene=Joy99.50\n",
      "CDS 38815 - 39165\n",
      "  /gene=Joy99.51\n",
      "CDS 39162 - 39545\n",
      "  /gene=Joy99.52\n",
      "CDS 39546 - 39824\n",
      "  /gene=Joy99.53\n",
      "CDS 39821 - 40021\n",
      "  /gene=Joy99.54\n",
      "CDS 40033 - 40587\n",
      "  /gene=Joy99.55\n",
      "CDS 40587 - 40853\n",
      "  /gene=Joy99.56\n",
      "CDS 40850 - 41182\n",
      "  /gene=Joy99.57\n",
      "CDS 41185 - 41343\n",
      "  /gene=Joy99.58\n",
      "CDS 41340 - 42218\n",
      "  /gene=Joy99.59\n",
      "CDS 42215 - 42448\n",
      "  /gene=Joy99.60\n",
      "CDS 42524 - 42739\n",
      "  /gene=Joy99.61\n",
      "CDS 42739 - 42858\n",
      "  /gene=Joy99.62\n",
      "CDS 42940 - 43449\n",
      "  /gene=Joy99.63\n",
      "CDS 43551 - 43793\n",
      "  /gene=Joy99.64\n",
      "CDS 43790 - 44029\n",
      "  /gene=Joy99.65\n",
      "CDS 44026 - 44397\n",
      "  /gene=Joy99.66\n",
      "CDS 44439 - 47057\n",
      "  /gene=Joy99.67\n",
      "CDS 47466 - 48164\n",
      "  /gene=Joy99.68\n",
      "CDS 48157 - 48618\n",
      "  /gene=Joy99.69\n",
      "CDS 48615 - 48818\n",
      "  /gene=Joy99.70\n",
      "CDS 48815 - 50002\n",
      "  /gene=Joy99.71\n",
      "CDS 50015 - 50248\n",
      "  /gene=Joy99.72\n",
      "CDS 50245 - 50517\n",
      "  /gene=Joy99.73\n",
      "CDS 50590 - 51051\n",
      "  /gene=Joy99.74\n",
      "CDS 51176 - 51451\n",
      "  /gene=Joy99.75\n",
      "CDS 51538 - 51666\n",
      "  /gene=Joy99.76\n",
      "CDS 51663 - 52565\n",
      "  /gene=Joy99.77\n",
      "CDS 52562 - 53038\n",
      "  /gene=Joy99.78\n",
      "CDS 53032 - 53313\n",
      "  /gene=Joy99.79\n",
      "CDS 53313 - 53615\n",
      "  /gene=Joy99.80\n",
      "CDS 53615 - 53812\n",
      "  /gene=Joy99.81\n",
      "CDS 53809 - 53961\n",
      "  /gene=Joy99.82\n",
      "CDS 54073 - 55266\n",
      "  /gene=Joy99.83\n",
      "CDS 55263 - 55649\n",
      "  /gene=Joy99.84\n",
      "CDS 55649 - 56188\n",
      "  /gene=Joy99.85\n",
      "CDS 56181 - 56753\n",
      "  /gene=Joy99.86\n",
      "CDS 56750 - 56911\n",
      "  /gene=Joy99.87\n",
      "CDS 57083 - 57709\n",
      "  /gene=Joy99.88\n",
      "CDS 57741 - 57941\n",
      "  /gene=Joy99.89\n",
      "CDS 58064 - 58279\n",
      "  /gene=Joy99.90\n",
      "CDS 58427 - 58696\n",
      "  /gene=Joy99.91\n",
      "CDS 59070 - 59300\n",
      "  /gene=Joy99.92\n",
      "CDS 59345 - 59479\n",
      "  /gene=Joy99.93\n",
      "CDS 59476 - 59778\n",
      "  /gene=Joy99.94\n",
      "ORIGIN\n"
     ]
    }
   ],
   "source": [
    "#STEP 1: Click inside this box Here\n",
    "\n",
    "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
    "\n",
    "fullGFFFileName = \"/Users/chris/Desktop/input/\" + gffFileName\n",
    "\n",
    "try:\n",
    "    with open(fullGFFFileName) as gffFile:\n",
    "        gffFileContents = gffFile.readlines()\n",
    "except:\n",
    "    gffFileContents = []\n",
    "    print '''Well that didn't work! Could not get GFF file. Are you sure it is in the input folder?\n",
    "    Double check spelling, no spaces in file name, try again. If that fails get help.'''\n",
    "    sys.exit()\n",
    "print \"\\n\\n================================== Results ==================================\\n\"\n",
    "        \n",
    "firstline = gffFileContents[0]\n",
    "\n",
    "print len(firstline)\n",
    "if not validHeader(firstline):\n",
    "    printFailureMessage(\"Check for unix formated file, Header line\")\n",
    "    print firstline\n",
    "    sys.exit()\n",
    "\n",
    "    \n",
    "print \"checking line: \", \n",
    "for lineNum,wholeLine in enumerate(gffFileContents[1:]):\n",
    "    lineNumber = str(lineNum+1)\n",
    "    print \" \" + lineNumber,\n",
    "    \n",
    "    try:\n",
    "        if not validTabStructure(wholeLine):                          #check whole line for proper 8 tab structure\n",
    "            failureMessage = \"Line number: \" + lineNumber + \" tab structure (i.e. 9 entries and 8 tabs)\"\n",
    "            printFailureMessage(failureMessage)\n",
    "            sys.exit()\n",
    "    except:\n",
    "        print \"double check that all functions are defined!\"\n",
    "    \n",
    "    \n",
    "    line = wholeLine.strip().split(\"\\t\")\n",
    "    \n",
    "    if not validSeqname(line[0]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" seq name of \" + line[0]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "    \n",
    "    if not validSource(line[1]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" source definition of \" + line[1] \n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validType(line[2]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" type definition of \" + line[2]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validCoordinates(line[3],line[4]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" coordinates left: \" + line[3] + \" right: \" + line[4] + \" have\"\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validScore(line[5]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" score of \" + line[5]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validStrand(line[6]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" strand entry of \" + line[6]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validPhase(line[7]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" phase of \" + line[7]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "    \n",
    "    \n",
    "    if not validAttributes(line[8]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" attribute entry of: \" + line[8]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "print \"\\n\\nCongratulations! all tests passed\\nEverything below can be used in DNA Master\"\n",
    "print \"=========================== DNA Master documentation ========================\"\n",
    "text = createDNAMasterFile(gffFileContents)\n",
    "print text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Definitions of Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#%%writefile gffTester_nose.py\n",
    "import re,sys\n",
    "\n",
    "def validHeader(line):\n",
    "    \"\"\"\n",
    "    Return True if line is a valid gff3 header.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line' - line to check\n",
    "    \"\"\"\n",
    "    if len(line) != 16:\n",
    "        return False\n",
    "    elif line.strip() == \"##gff-version 3\":\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "def validTabStructure(line):\n",
    "    \"\"\"\n",
    "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line': line to check\n",
    "    \"\"\"\n",
    "    theLine = line.strip().split(\"\\t\")  #strip off extra charachters and split the line on tab\n",
    "\n",
    "    if len(theLine) != 9:               # 9 columns in a gff file, so return False if not exactly 9\n",
    "        print 'Length = ' + str(len(theLine))\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "def validSeqname(text):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
    "    the sequence name on the Gbrowse database and have only valid characters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of entry in column 1 to check\n",
    "    \"\"\"\n",
    "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
    "\n",
    "def validSource(text):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
    "    which only need validity of the characters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of the entry in column 2 to check\n",
    "    \"\"\"\n",
    "    return not charCheck(text)\n",
    "\n",
    "def validType(text, validTypes=None):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
    "    For phage this should be one of types ['gene','mRNA','exon']\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of the entry in column 2 to check\n",
    "    - 'validTypes': List of valid types for checking if not the default 3\n",
    "    \"\"\"\n",
    "    if validTypes is None:\n",
    "        validTypes = {'gene','mRNA','exon', 'contig'}\n",
    "    \n",
    "    return text in validTypes\n",
    "\n",
    "\n",
    "def validScore(score):\n",
    "    \"\"\"\n",
    "    Checks a score (entries in column 6), should be a number or \".\"\n",
    "    \n",
    "    Parameters:\n",
    "    - 'score': score from column 6\n",
    "    \"\"\"\n",
    "    if score == \".\":\n",
    "        return True\n",
    "    \n",
    "    try:\n",
    "        x = float(score)\n",
    "    except:\n",
    "        isScore = False\n",
    "        pass\n",
    "    else:\n",
    "        isScore = True        \n",
    "\n",
    "    return isScore\n",
    "\n",
    "def validStrand(strand):\n",
    "    \"\"\"\n",
    "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
    "    \n",
    "    Parameters:\n",
    "    - 'strand': strand entry from column 7\n",
    "    \"\"\"\n",
    "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
    "    \n",
    "    return strand in validValues\n",
    "\n",
    "def validPhase(phase):\n",
    "    \"\"\"\n",
    "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
    "    \n",
    "    Parameters:\n",
    "    - 'phase': strand entry from column 8\n",
    "    \"\"\"\n",
    "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
    "    \n",
    "    return phase in validValues\n",
    "\n",
    "def validAttributes(attributes):\n",
    "    '''\n",
    "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
    "    It is OK to have spaces in the values entry but not the Key value\n",
    "    \n",
    "    \n",
    "    Parameters:\n",
    "    - 'attributes': entire entry from column 9\n",
    "    '''\n",
    "    #ok to have a null string\n",
    "    if attributes == '.':\n",
    "        return True\n",
    "    \n",
    "    #checking for invalid characters     \n",
    "    if bool(charCheck(attributes, search=re.compile(r'[^a-zA-Z0-9.=;_ \"]').search)):\n",
    "        return False\n",
    "    \n",
    "    \n",
    "    #ok go ahead and split into the underlyine key=value items \n",
    "\n",
    "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
    "    \n",
    "    for attrPair in attrPairs:\n",
    "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
    "            continue\n",
    "            \n",
    "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
    "            return False\n",
    "        \n",
    "        attrKey, attrValue = attrPair.split('=')\n",
    "        \n",
    "        if len(attrKey) < 1:                #must have an entry for a key\n",
    "            print len(attrKey)\n",
    "            return False\n",
    "        \n",
    "        if charCheck(attrKey):              #check key for valid charachters\n",
    "            return False\n",
    "        \n",
    "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
    "        if charCheck(attrValue.replace(' ','')):\n",
    "            return False\n",
    "            \n",
    "        return True\n",
    "\n",
    "\n",
    "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
    "    \"\"\"\n",
    "    Checks a string for characters NOT a-zA-Z0-9.=; and returns True if invalid character is\n",
    "    found.\n",
    "\n",
    "    Parameters:\n",
    "    -'str': string being checked.\n",
    "    -'search': default search all characters in valid set\n",
    "    \"\"\"\n",
    "    return bool(search(str))\n",
    "\n",
    "def validCoordinate(coord):\n",
    "    \"\"\"\n",
    "    Checks a coordinate (entries in column 4 or 5), should be a positive integer\n",
    "    \n",
    "    Parameters:\n",
    "    - 'coord': coordinate from column 4 or 5\n",
    "    \"\"\"\n",
    "    \n",
    "    return not bool(charCheck(coord, search=re.compile(r'[^0-9]').search))\n",
    "\n",
    "def validCoordinates(leftCoord, rightCoord):\n",
    "    \"\"\"\n",
    "    Checks a coordinates (entries in column 4 and 5), should be a positive integers\n",
    "    and the left Coordinate should be smaller than the right Coordinate\n",
    "    \n",
    "    Parameters:\n",
    "    - 'leftCoord':  coordinate from column 4\n",
    "    - 'rightCoord': coordinate from column 5\n",
    "\n",
    "    \"\"\"\n",
    "    return (validCoordinate(leftCoord) and validCoordinate(rightCoord) and int(leftCoord) <= int(rightCoord))\n",
    "\n",
    "\n",
    "def printFailureMessage(failType):\n",
    "    print \"\\n##### Fatal Error #####\"\n",
    "    print failType + \" failed. No other tests run.\"\n",
    "    print \"\"\"\n",
    "================================ End Results ================================        \n",
    "Computer messages follow below OK to ignore, you should fix file and try again.\n",
    "    \"\"\"\n",
    "    \n",
    "def createDNAMasterFile(gffFileContents):\n",
    "    \"\"\"\n",
    "    Takes in a list of GFF lines, checks for headers and passes lines of type gene to parser.\n",
    "    Returns the complete text of DNA Master text file.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'gffFileContents':  List of text entries, each element a line from GFF# file.\n",
    "    \"\"\"\n",
    "    \n",
    "    text = ''\n",
    "    for wholeLine in gffFileContents:\n",
    "        line = wholeLine.strip().split(\"\\t\")\n",
    "        \n",
    "        if len(line) < 8:\n",
    "            continue\n",
    "        \n",
    "        if line[2] == 'gene':\n",
    "            entry = gene2CDS(line)\n",
    "            text += entry\n",
    "            \n",
    "        text += parseAttributes(line[8])\n",
    "            \n",
    "    epilog = \"ORIGIN\"\n",
    "    text += epilog\n",
    "        \n",
    "    return text\n",
    "            \n",
    "        \n",
    "def gene2CDS(line):\n",
    "    \"\"\"\n",
    "    Takes in a single GFF entry and returns the corresponding DNA Master entry for the CDS line.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line':  List of entries from a single GFF entry.\n",
    "    \"\"\"    \n",
    "    if line[6] == \"+\":\n",
    "        return \"CDS \" + line[3] + \" - \" + line[4] + \"\\n\"\n",
    "    else:\n",
    "        return \"CDS complement (\" + line[3] + \" - \" + line[4] + \")\\n\"\n",
    "    \n",
    "def parseAttributes(attributes):\n",
    "    \"\"\"\n",
    "    takes in the Attributes entry of the GFF. If there is an ID= or Name=\n",
    "    entry then create the /gene line. If there is a notes= entry then create\n",
    "    the /note line\n",
    "    \"\"\"\n",
    "    #ok go ahead and split into the underlyine key=value items \n",
    "\n",
    "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
    "    \n",
    "    returnString = \"\"\n",
    "    \n",
    "    for attrPair in attrPairs:\n",
    "        attrKey = \"\"\n",
    "        attrValue = \"\"\n",
    "        \n",
    "        if len(attrPair.split('=')) == 2:\n",
    "            attrKey = attrPair.split(\"=\")[0].lower()\n",
    "            attrValue =  attrPair.split(\"=\")[1]\n",
    "        \n",
    "        if attrKey in [\"id\", \"name\"]:\n",
    "            returnString += '  /gene=' + attrValue + '\\n'\n",
    "        if attrKey == \"note\":\n",
    "            if attrValue[0:1] == attrValue[-1:] == '\"':\n",
    "                attrValue = attrValue[1:-1]\n",
    "            returnString += '    /note=\"' + attrValue + '\"\\n'\n",
    "    \n",
    "    return returnString"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%writefile test_code.py\n",
    "\n",
    "from gffTester_nose import *\n",
    "\n",
    "'''\n",
    "need to check the following functions:\n",
    "\n",
    "def validHeader (line): TEST valid, extra space, extra tab, garbled\n",
    "def validTabStructure(line): test: 7 tabs, 8 tabs, 9 tabs\n",
    "def validSeqname(text): SKIP, just a \"not charCheck\" for now\n",
    "def validSource(text): SKIP, just a \"not charCheck\" for now\n",
    "def validType(text, validTypes=None): TEST: pass with any 3 defaults 'gene','mRNA','exon'; fails on none item, passes with list\n",
    "def validCoordinate(coord): TEST: fails with text, float, negative int; passes with positive int\n",
    "def validScore(score):\n",
    "def validStrand(strand):\n",
    "def validPhase(phase):\n",
    "def validAttributes(attributes):\n",
    "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
    "def validGene(seq, line):       \n",
    "def fastaRead(fasta_File):\n",
    "def countStopCodons(seq, strand):\n",
    "'''\n",
    "\n",
    "def test_validHeader_1():\n",
    "    'valid header tests as valid'\n",
    "    result = validHeader('##gff-version 3\\n')\n",
    "    assert result == True\n",
    "\n",
    "def test_validHeader_2():\n",
    "    'header with extra space fails'\n",
    "    result = validHeader('##gff-version 3 \\n')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validHeader_3():\n",
    "    'header with extra tab fails'\n",
    "    result = validHeader('##gff-version 3\\t\\n')\n",
    "    assert result == False\n",
    "\n",
    "def test_validHeader_4():\n",
    "    'garbled header fails'\n",
    "    result = validHeader('##gff-Version 3\\n')\n",
    "    assert result == False\n",
    "    \n",
    "\n",
    "def test_validTabStructure_1():\n",
    "    'fails 7 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\n\")\n",
    "    assert result == False\n",
    "\n",
    "def test_validTabStructure_2():\n",
    "    'passes 8 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\n\")\n",
    "    assert result == True\n",
    "    \n",
    "def test_validTabStructure_3():\n",
    "    'fails 9 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\t10\\n\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validType_1():\n",
    "    'Type passes with \"gene\"'\n",
    "    result = validType('gene')\n",
    "    assert result == True\n",
    "\n",
    "def test_validType_2():\n",
    "    'Type passes with \"mRNA\"'\n",
    "    result = validType('mRNA')\n",
    "    assert result == True\n",
    "\n",
    "def test_validType_3():\n",
    "    'Type passes with \"exon\"'\n",
    "    result = validType('exon')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validType_4():\n",
    "    'Type fails with mrna'\n",
    "    result = validType('mrna')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validType_5():\n",
    "    'Type passes with specified list'\n",
    "    result = validType('mrna', ['mRNA', 'mrna'])\n",
    "    assert result == True\n",
    "    \n",
    "\n",
    "def test_validCoordinate_1():\n",
    "    'Coord fails with text as coordinate'\n",
    "    result = validCoordinate('a')\n",
    "    assert result == False\n",
    "\n",
    "def test_validCoordinate_2():\n",
    "    'Coord fails with float'\n",
    "    result = validCoordinate('2.0')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validCoordinate_3():\n",
    "    'Coord fails with negative int'\n",
    "    result = validCoordinate('-3')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validCoordinate_4():\n",
    "    'Coord passes with positive integer'\n",
    "    result = validCoordinate('100')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validCoordinates_1():\n",
    "    'Coordinates fails with left larger than right'\n",
    "    result = validCoordinates('5','4')\n",
    "    assert result == False\n",
    "\n",
    "def test_validCoordinates_2():\n",
    "    'passes with left equal to right'\n",
    "    result = validCoordinates('5','5')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validCoordinates_3():\n",
    "    'passes with left smaller than right'\n",
    "    result = validCoordinates('4','5')\n",
    "    assert result == True\n",
    "\n",
    "    \n",
    "def test_validAttributes_1():\n",
    "    'Attributes passes with null string'\n",
    "    result = validAttributes(\"\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_2():\n",
    "    'Attributes passes a=b'\n",
    "    result = validAttributes(\"a=b\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_3():\n",
    "    'attrubutes passes with \"a=b;\"'\n",
    "    result = validAttributes(\"a=b;\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_4():\n",
    "    'Attributes failes with \"a=bc=d\"'\n",
    "    result = validAttributes(\"a=bc=d\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validAttributes_5():\n",
    "    'attrubutes passes with \"a=b;c=d\"'\n",
    "    result = validAttributes(\"a=b;c=d\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_6():\n",
    "    'attrubutes passes with \"af=b f;c=d\"'\n",
    "    result = validAttributes(\"af=b f;c=d\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_7():\n",
    "    'attrubutes fails with blank key \"a=b;=b\"'\n",
    "    result = validAttributes(\"a=b;=b\")\n",
    "    assert result == False\n",
    "\n",
    "def test_validAttributes_8():\n",
    "    'attrubutes fails with illegal character in key \"key&=ab\"'\n",
    "    result = validAttributes(\"key&=ab\")\n",
    "    assert result == False\n",
    "\n",
    "\n",
    "def test_parseAttributes_1():\n",
    "    'Attribute test 1 single ID entry'\n",
    "    returned = parseAttributes(\"ID=gene1\")\n",
    "    assert returned == '  /gene=gene1\\n'\n",
    "\n",
    "def test_parseAttributes_2():\n",
    "    'Attribute test 2 single ID entry with semicolon'\n",
    "    returned = parseAttributes(\"ID=gene1;\")\n",
    "    assert returned == '  /gene=gene1\\n'\n",
    "    \n",
    "def test_parseAttributes_3():\n",
    "    'Attribute test 3 single Note entry with spaces'\n",
    "    returned = parseAttributes(\"Note=note note\")\n",
    "    assert returned == '    /note=\"note note\"\\n'\n",
    "    \n",
    "def test_parseAttributes_4():\n",
    "    'Attribute test 4 single Note entry with semicolon'\n",
    "    returned = parseAttributes(\"Note=gene;\")\n",
    "    assert returned == '    /note=\"gene\"\\n'\n",
    "    \n",
    "def test_parseAttributes_5():\n",
    "    'Attribute test 5 single Note entry preceded by ;'\n",
    "    returned = parseAttributes(\";Note=gene2\")\n",
    "    assert returned == '    /note=\"gene2\"\\n'\n",
    "    \n",
    "def test_parseAttributes_6():\n",
    "    'Attribute test 6 ID + Note entry'\n",
    "    returned = parseAttributes(\"ID=gene1;Note=note4\")\n",
    "    assert returned == '  /gene=gene1\\n    /note=\"note4\"\\n'\n",
    "    \n",
    "def test_parseAttributes_7():\n",
    "    'Attribute test 7 single Note entry with quotes'\n",
    "    returned = parseAttributes('Note=\"gene7\"')\n",
    "    assert returned == '    /note=\"gene7\"\\n'\n",
    "    \n",
    "def test_parseAttributes_8():\n",
    "    'Attribute test 8 single id entry with quotes'\n",
    "    returned = parseAttributes('id=\"gene8\"')\n",
    "    assert returned == '  /gene=\"gene8\"\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!nosetests -v test_code.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print \"grene\" in [\"gene\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "a = '\"jkLmn\"'\n",
    "b = '3'\n",
    "c = '\"'\n",
    "print a[1:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%who"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "info = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "info['pham']= None\n",
    "print type(info['pham'])\n",
    "if info['pham']:\n",
    "    print \"true\"\n",
    "else:\n",
    "    print \"false\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "whos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
