{
 "metadata": {
  "name": "",
  "signature": "sha256:f940a07fe08b5df54e0607781f0bb17bb9c327c8d368a584b9d5a404f684a527"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "GFF phage file validation program"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###README:\n",
      "\n",
      "This program will check your GFF file. It checks for proper format and use of proper characters (format checks). If you give the program the fasta file with the DNA sequence of your phage, the program will also check that the \"gene\" entries start with a proper start codon and have only a single stop codon as the last codon (biology checks).\n",
      "\n",
      "**Before you run the program:** copy both your phage fasta file and your GFF file into the \"input\" folder on the Desktop. Make sure you know the EXACT name of the files, it is best to avoid spaces in filenames (rename your files if necessary).  \n",
      "\n",
      "**To run the program:** click anywhere in the code box below near the \"#Click in Here\" so that you see the curser, hit shift-return (i.e. hold shift while hitting the return key).\n",
      "\n",
      "**What will happen:** the first thing to appear will be a small box asking for the name of your gff file. Type the name of the gff file into the box. Once the box has the correct name, hit the return key (no shift). You will then see another box asking for the name of your fasta file. If you only want format checks, make sure the cursor is in the box, leave the box blank and hit enter. If you want the program to also run biology checks, type the name of the fasta file, then hit enter.\n",
      "\n",
      "Your results should appear directly below the two question boxes."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Click in Here\n",
      "\n",
      "fastaFileName = raw_input('Enter the name of the fasta file (BE EXACT): ')\n",
      "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
      "\n",
      "fullFastaFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + fastaFileName\n",
      "fullGFFFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + gffFileName\n",
      "\n",
      "try:\n",
      "    with open(fullFastaFileName) as fastaFile:\n",
      "        fastaFileContents = fastaFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get fasta file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "try:\n",
      "    with open(fullGFFFileName) as gffFile:\n",
      "        gffFileContents = gffFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get GFF file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "print \"\\n\\n================================ Results ================================\\n\"\n",
      "        \n",
      "firstline = gffFileContents[0]\n",
      "\n",
      "if not validHeader(firstline):\n",
      "    print \"##### Fatal Error #####\"\n",
      "    print \"\"\"\n",
      "    GFF header line failed. No other tests run. \n",
      "    Check 1st line, it should be EXACTLY\n",
      "    ##gff-version 3\n",
      "    Most Common error: spaces and/or tabs at end of line after 3.\n",
      "    \n",
      "Computer messages follow, ignore and try again:\n",
      "    \"\"\"\n",
      "\n",
      "    sys.exit()\n",
      "\n",
      "for lineNumber,line in enumerate(gffFileContents[1:]):\n",
      "    print lineNumber, line"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Definitions of Functions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%writefile gffTester_nose.py\n",
      "import re,sys\n",
      "\n",
      "def validHeader (line):\n",
      "    \"\"\"\n",
      "    Return True if line is a valid gff3 header.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line' - line to check\n",
      "    \"\"\"\n",
      "    if line == \"##gff-version 3\\n\":\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "    \n",
      "def validTabStructure(line):\n",
      "    \"\"\"\n",
      "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line': line to check\n",
      "    \"\"\"\n",
      "    theLine = line.strip().split(\"\\t\")  #strip off extra charachters and split the line on tab\n",
      "    if len(theLine) != 9:               # 9 columns in a gff file, so return False if not exactly 9\n",
      "        return False\n",
      "    else:\n",
      "        return True\n",
      "\n",
      "def validSeqname(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
      "    the sequence name on the Gbrowse database and have only valid characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of entry in column 1 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
      "\n",
      "def validSource(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
      "    which only need validity of the characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)\n",
      "\n",
      "def validType(text, validTypes=None):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
      "    For phage this should be one of types ['gene','mRNA','exon']\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    - 'validTypes': List of valid types for checking if not the default 3\n",
      "    \"\"\"\n",
      "    if validTypes is None:\n",
      "        validTypes = ['gene','mRNA','exon']\n",
      "    \n",
      "    return text in validTypes\n",
      "\n",
      "\n",
      "def validScore(score):\n",
      "    \"\"\"\n",
      "    Checks a score (entries in column 6), should be a number or \".\"\n",
      "    \n",
      "    Parameters:\n",
      "    - 'score': score from column 6\n",
      "    \"\"\"\n",
      "    if score == \".\":\n",
      "        return True\n",
      "    \n",
      "    try:\n",
      "        x = float(score)\n",
      "    except:\n",
      "        isScore = False\n",
      "        pass\n",
      "    else:\n",
      "        isScore = True        \n",
      "\n",
      "    return isScore\n",
      "\n",
      "def validStrand(strand):\n",
      "    \"\"\"\n",
      "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'strand': strand entry from column 7\n",
      "    \"\"\"\n",
      "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
      "    \n",
      "    return strand in validValues\n",
      "\n",
      "def validPhase(phase):\n",
      "    \"\"\"\n",
      "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'phase': strand entry from column 8\n",
      "    \"\"\"\n",
      "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
      "    \n",
      "    return phase in validValues\n",
      "\n",
      "def validAttributes(attributes):\n",
      "    \"\"\"\n",
      "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
      "    It is OK to have spaces in the values entry but not the Key value\n",
      "    \n",
      "    \n",
      "    Parameters:\n",
      "    - 'attributes': entire entry from column 9\n",
      "    \"\"\"\n",
      "    \n",
      "    #ok to have a null string\n",
      "    if attributes == '':\n",
      "        return True\n",
      "    \n",
      "     #checking for invalid characters, spaces OK in column 9 values so remove any before checking with charCheck\n",
      "\n",
      "        \n",
      "    #ok go ahead and split into the underlyine key=value items    \n",
      "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
      "    \n",
      "    validity = True\n",
      "    \n",
      "    for attrPair in attrPairs:\n",
      "        print attrPair\n",
      "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
      "            continue\n",
      "            \n",
      "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
      "            validity = False\n",
      "        \n",
      "        attrKey, attrValue = attrPair.split('=')\n",
      "        \n",
      "        if len(attrKey) < 1:                #must have an entry for a key\n",
      "            validity = False\n",
      "        \n",
      "        if charCheck(attrKey):              #check key for valid charachters\n",
      "            validity = False\n",
      "        \n",
      "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
      "        if charCheck(attrValue.replace(' ','')):\n",
      "            validity = False\n",
      "            \n",
      "    return validity\n",
      "\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters NOT a-zA-Z0-9.=; and returns True if invalid character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    return bool(search(str))\n",
      "\n",
      "def validCoordinate(coord):\n",
      "    \"\"\"\n",
      "    Checks a coordinate (entries in column 4 and 5), should be a positive integer\n",
      "    \n",
      "    Parameters:\n",
      "    - 'coord': coordinate from column 4 or 5\n",
      "    \"\"\"\n",
      "    \n",
      "    return not bool(charCheck(coord, search=re.compile(r'[^0-9]').search))\n",
      "\n",
      "def validGene(seq, line):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is define a proper gene.\n",
      "    \n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has one stop codon and it is the last codon.\n",
      "    - begins with a valid start codon [ATG, GTG, CTG].\n",
      "    \n",
      "    Parameters:\n",
      "    -'Seq': string nucleotide sequence of entire phage.\n",
      "    -'line': single line in gff3 format for checking.\n",
      "    \"\"\"\n",
      "    \n",
      "    leftCoordinate= int(line.split(\"\\t\")[3])\n",
      "    rightCoordinate= int(line.split(\"\\t\")[4])\n",
      "    strand = line.split(\"\\t\")[6]\n",
      "    geneSeq = seq[leftCoordinate-1:rightCoordinate]   #not really the gene for negative strand genes but good enough\n",
      "    \n",
      "    if  strand == \"+\":\n",
      "        firstCodon = geneSeq[:3].upper()\n",
      "        lastCodon = geneSeq[-3:].upper()\n",
      "        startCodons = ['ATG', 'TTG', 'GTG']\n",
      "        stopCodons = ['TAA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        firstCodon = geneSeq[-3:].upper()\n",
      "        lastCodon = geneSeq[:3].upper()\n",
      "        startCodons = ['CAT', 'CAA', 'CAC']   #reverse complement of the start codons\n",
      "        stopCodons = ['TTA', 'CTA', 'TCA']   #reverse complement of the stop codons\n",
      "\n",
      "    #first make sure the values of leftCoordinate and rightCoordinate are OK\n",
      "    \n",
      "    if (leftCoordinate >= len(seq)) or (rightCoordinate > len(seq)):  #GFF is 1 based counting \n",
      "        return False\n",
      "    \n",
      "    if (rightCoordinate - leftCoordinate) < 30:                       #minimum gene size for this check is 10 a.a.\n",
      "        return False\n",
      "    \n",
      "    if len(geneSeq)%3 != 0:\n",
      "        return False\n",
      "    \n",
      "    if firstCodon not in startCodons:\n",
      "        return False\n",
      "    \n",
      "    if lastCodon not in stopCodons:\n",
      "        return False\n",
      "    \n",
      "    stopCodonCount = countStopCodons(geneSeq,strand)\n",
      "    if  stopCodonCount > 1:\n",
      "        return False\n",
      "    \n",
      "    return True\n",
      "\n",
      "def fastaRead(fasta_File):\n",
      "    \"\"\"\n",
      "    Reads in a nucleotide sequence from a text file in .fasta format.\n",
      "\n",
      "    Parameters:\n",
      "    -'fasta_File': uploaded text file in .fasta format\n",
      "\n",
      "    Output:\n",
      "    -'seq': string of a nucleotide sequence.\n",
      "    \"\"\"\n",
      "\n",
      "    seq = []\n",
      "    try:\n",
      "        f1 = open(fasta_File, \"r\")\n",
      "    except:\n",
      "        return \"Error: unable to read fasta file.\"\n",
      "    \n",
      "    for line in f1:\n",
      "        line = line.rstrip()\n",
      "        if line.startswith(\">\"):\n",
      "            continue\n",
      "        else:\n",
      "            seq.append(line)\n",
      "\n",
      "    f1.close()\n",
      "    return ''.join(seq)\n",
      "\n",
      "def countStopCodons(seq, strand):\n",
      "    \"\"\"\n",
      "    Used to count the number of stop codons in a given sequence\n",
      "\n",
      "    Parameters:\n",
      "    -'seq': nucleotide sequence of plus strand.\n",
      "    -'strand': indication if gene is on the plus or minus strand\n",
      "\n",
      "    Output:\n",
      "    -'count': total number of stop codons found in the sequence.\n",
      "    \"\"\"\n",
      "    \n",
      "    count = 0\n",
      "    \n",
      "    if strand == \"+\":\n",
      "        stopCodons = ['TTA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        stopCodons = ['TAA', 'CTA', 'TCA']   ##reverse complement of stop codons\n",
      "        \n",
      "    #now create a list of all codons using list comprehension, this only works well if len%3 == 0\n",
      "    if strand == \"+\":\n",
      "        codons=[seq[x:x+3] for x in range(0,len(seq),3)]\n",
      "    else:\n",
      "        codons=[seq[x:x+3] for x in range(-3,-len(seq),-3)] #works for all but first codon becuase string[-3:0]=''\n",
      "        codons[0] = seq[-3:]                                #so fudge in the \"first\" codon on the negative side\n",
      "    \n",
      "    for stop in stopCodons:\n",
      "        count += codons.count(stop)\n",
      "        \n",
      "    return count"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Overwriting gffTester_nose.py\n"
       ]
      }
     ],
     "prompt_number": 49
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Test Code"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%writefile test_code.py\n",
      "\n",
      "from gffTester_nose import *\n",
      "\n",
      "'''\n",
      "need to check the following functions:\n",
      "\n",
      "def validHeader (line): TEST valid, extra space, extra tab, garbled\n",
      "def validTabStructure(line): test: 7 tabs, 8 tabs, 9 tabs\n",
      "def validSeqname(text): SKIP, just a \"not charCheck\" for now\n",
      "def validSource(text): SKIP, just a \"not charCheck\" for now\n",
      "def validType(text, validTypes=None): TEST: pass with any 3 defaults 'gene','mRNA','exon'; fails on none item, passes with list\n",
      "def validCoordinate(coord): TEST: fails with text, float, negative int; passes with positive int\n",
      "def validScore(score):\n",
      "def validStrand(strand):\n",
      "def validPhase(phase):\n",
      "def validAttributes(attributes):\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "def validGene(seq, line):       \n",
      "def fastaRead(fasta_File):\n",
      "def countStopCodons(seq, strand):\n",
      "'''\n",
      "\n",
      "def test_validHeader_1():\n",
      "    'valid header tests as valid'\n",
      "    result = validHeader('##gff-version 3\\n')\n",
      "    assert result == True\n",
      "\n",
      "def test_validHeader_2():\n",
      "    'header with extra space fails'\n",
      "    result = validHeader('##gff-version 3 \\n')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validHeader_3():\n",
      "    'header with extra tab fails'\n",
      "    result = validHeader('##gff-version 3\\t\\n')\n",
      "    assert result == False\n",
      "\n",
      "    \n",
      "def test_validHeader_4():\n",
      "    'garbled header fails'\n",
      "    result = validHeader('##gff-Version 3\\n')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validTabStructure_1():\n",
      "    'fails 7 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\n\")\n",
      "    assert result == False\n",
      "\n",
      "def test_validTabStructure_2():\n",
      "    'passes 8 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\n\")\n",
      "    assert result == True\n",
      "    \n",
      "def test_validTabStructure_3():\n",
      "    'fails 9 tab structure'\n",
      "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\t10\\n\")\n",
      "    assert result == False\n",
      "    \n",
      "def test_validType_1():\n",
      "    'passes with \"gene\"'\n",
      "    result = validType('gene')\n",
      "    assert result == True\n",
      "\n",
      "def test_validType_2():\n",
      "    'passes with \"mRNA\"'\n",
      "    result = validType('mRNA')\n",
      "    assert result == True\n",
      "\n",
      "def test_validType_3():\n",
      "    'passes with \"exon\"'\n",
      "    result = validType('exon')\n",
      "    assert result == True\n",
      "    \n",
      "def test_validType_4():\n",
      "    'fails with mrna'\n",
      "    result = validType('mrna')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validType_5():\n",
      "    'passes with specified list'\n",
      "    result = validType('mrna', ['mRNA', 'mrna'])\n",
      "    assert result == True\n",
      "    \n",
      "\n",
      "def test_validCoordinate_1():\n",
      "    'fails with text as coordinate'\n",
      "    result = validCoordinate('a')\n",
      "    assert result == False\n",
      "\n",
      "def test_validCoordinate_2():\n",
      "    'fails with float'\n",
      "    result = validCoordinate('2.0')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validCoordinate_3():\n",
      "    'fails with negative int'\n",
      "    result = validCoordinate('-3')\n",
      "    assert result == False\n",
      "    \n",
      "def test_validCoordinate_4():\n",
      "    'passes with positive integer'\n",
      "    result = validCoordinate('100')\n",
      "    assert result == True"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Overwriting test_code.py\n"
       ]
      }
     ],
     "prompt_number": 53
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!nosetests -v test_code.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "valid header tests as valid ... ok\r\n",
        "header with extra space fails ... ok\r\n",
        "header with extra tab fails ... ok\r\n",
        "garbled header fails ... ok\r\n",
        "fails 7 tab structure ... ok\r\n",
        "passes 8 tab structure ... ok\r\n",
        "fails 9 tab structure ... ok\r\n",
        "passes with \"gene\" ... ok\r\n",
        "passes with \"mRNA\" ... ok\r\n",
        "passes with \"exon\" ... ok\r\n",
        "fails with mrna ... ok\r\n",
        "passes with specified list ... ok\r\n",
        "fails with text as coordinate ... ok\r\n",
        "fails with float ... ok\r\n",
        "fails with negative int ... ok\r\n",
        "passes with positive integer ... ok\r\n",
        "\r\n",
        "----------------------------------------------------------------------\r\n",
        "Ran 16 tests in 0.003s\r\n",
        "\r\n",
        "OK\r\n"
       ]
      }
     ],
     "prompt_number": 54
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'gene','mRNA','exon'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}