{
 "metadata": {
  "name": "",
  "signature": "sha256:c406647c520c887e016b16cf5a0b69bba7260636ed2b78f97417b9c5576c9bf3"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pwd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1,
       "text": [
        "u'/Users/chris/Documents/code/python/gff_validator'"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys, re, Bio\n",
      "from Bio.Seq import Seq\n",
      "from Bio.Alphabet import IUPAC\n",
      "from Bio import SeqIO"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "set up functions"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def sortGff3"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def sortGff3(gff3_File, types = ['gene','mRNA','exon']): ## don't forget to add name checking\n",
      "    \"\"\"   \n",
      "    This checks for proper \"structure\" for correct format:\n",
      "    - file starts with proper header\n",
      "    - each line is tab delimited and has 9 components.\n",
      "    - each line has a proper type at the 3rd component.\n",
      "    - example line:\n",
      "    1stComp\\t2ndComp\\t3rdComp\\t4thComp\\t...\n",
      "    \n",
      "    It also sorts each line in the gff file according to the location of the entry (i.e. base position) and the\n",
      "    type hierarchy. Valid types are 'gene', 'mRNA', 'exon' and must match exactly. \n",
      "    Gene != gene, mrna != mRNA, etc.\n",
      "    \n",
      "    Parameters:\n",
      "    -'gff3_File': the uploaded gff file.\n",
      "    -'types': a list indicating the types of each gff line and the order the types should\n",
      "              be sorted in. The first type in the list will be ordered before the second, \n",
      "              the second type before the third, etc.\n",
      "\n",
      "    Output:\n",
      "    -'keyList': a list of keys sorted in the order that the lines will be sorted.\n",
      "    -'holder': a dictionary of lines paired with keys in 'keyList'.\n",
      "    \"\"\"\n",
      "    \n",
      "    holder = dict()                    #keys are <type>_leftcoord; (e.g. gene_410) value is the entire line from the GFF3 file\n",
      "    counter = dict()                   #keys are leftcoord; values are sum of indexs of types found with types list\n",
      "    structure = []                     #this will hold the errors from this function, which are considered \"structural\"\n",
      "    contigCount = False\n",
      "\n",
      "    formatCount = 0\n",
      "    \n",
      "    with open(gff3_File, \"r\") as file:\n",
      "        fi = file.readlines()              #OK for phages, may fail for large gff files\n",
      "\n",
      "    try:   #check if first line is proper header\n",
      "        if notValidHeader(fi[0]):\n",
      "            raise LineError(\"0026\")\n",
      "    except LineError as er:\n",
      "        structure.append(\"[ 0 ] \" + er.returnError() + \"\\n\" + fi[0])\n",
      "\n",
      "        \n",
      "    for lineCount,line in enumerate(fi[1:]):\n",
      "\n",
      "        try:         # this could be a method \"splitline\"\n",
      "            theLine = line.strip().split(\"\\t\")  #split on tab\n",
      "            if len(theLine) > 9: # 9 columns in a proper gff file\n",
      "                raise FormatError(\"0200\")\n",
      "            elif len(theLine) < 9:\n",
      "                raise FormatError(\"0100\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"[\" + str(lineCount+1) + \"] \" + er.returnError()  + \":\\n\" + line)\n",
      "            continue\n",
      "        except:\n",
      "            structure.append(\"Validation Error: unkown error. Check line# \" + str(lineCount+1) + \":\\n\" + line)\n",
      "            continue\n",
      "\n",
      "        if theLine[2] == \"contig\":\n",
      "            holder[theLine[2] + \"_\" + theLine[3]] = line\n",
      "            continue\n",
      "\n",
      "        try:\n",
      "            if theLine[2] not in types:\n",
      "                raise FormatError(\"0700\")\n",
      "        except FormatError:\n",
      "            continue                  #skip here as unparsable below, but don't report as these are \"entry\" errors\n",
      "\n",
      "        try:\n",
      "            try:\n",
      "                leftPosition = int(theLine[3])\n",
      "                rightPosition = int(theLine[4])\n",
      "            except ValueError:\n",
      "                raise FormatError(\"0800\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"[\" + str(lineCount+1) + \"] \" + er.returnError()  + \":\\n\" + line)\n",
      "            continue\n",
      "\n",
      "        #to get here line will is partially validated, it will:\n",
      "        # 1. not be type contig\n",
      "        # 2. have 9 components \n",
      "        # 3. have a valid type in 3rd column\n",
      "        # 4. have integers in 4th and 5th column\n",
      "\n",
      "        holder[theLine[2] + \"_\" + theLine[3]] = line #FIXME: problem if 2 genes with same start number (i.e. chaperones)\n",
      "\n",
      "        if theLine[3] in counter:\n",
      "            counter[theLine[3]] = counter[theLine[3]] + types.index(theLine[2])\n",
      "        else:\n",
      "            counter[theLine[3]] = types.index(theLine[2])\n",
      "\n",
      "\n",
      "    #this part checks that you have the correct number of entries for each start coordinate:\n",
      "    counterKeys = counter.keys()\n",
      "    \n",
      "    expectedNum = ((len(types))*(len(types)-1))/2\n",
      "    print \"coutnerKeys are: \" + str(counterKeys)\n",
      "\n",
      "    for key in counterKeys:  \n",
      "        thisNum = counter[key]\n",
      "        try:\n",
      "            if thisNum == expectedNum:\n",
      "                continue\n",
      "            else:\n",
      "                raise FormatError(\"0500\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"Gene at Coordinate \" + key + \" \" + er.returnError())\n",
      "    \n",
      "    \n",
      "    Errors[\"structure\"] = structure                    #add the list of structural errors to the Errors dictionary\n",
      "    \n",
      "    #need to sort all the keys \n",
      "    keyList = holder.keys()\n",
      "    \n",
      "\n",
      "    #first sort on the secondary charachteriestic which is type:\n",
      "    typeHierarchy = ['contig','gene','mRNA','exon']\n",
      "    typeSortList = sorted(keyList, key=lambda item: typeHierarchy.index(item.split(\"_\")[0] ))\n",
      "                      \n",
      "    #now sort on the primary charachteristic which is the left coordinate \n",
      "    sortedKeyList = sorted(typeSortList, key=lambda item: int(item.split(\"_\")[1]))\n",
      "            \n",
      "    tempS = str(gff3_File)\n",
      "    newName = tempS[0:-5] + \"_sorted.gff3\"\n",
      "    with open(newName,\"w\") as outfile:\n",
      "        outfile.write(\"##gff-version 3\\n\")\n",
      "        for key in sortedKeyList:\n",
      "            outfile.write(holder[key])\n",
      "            \n",
      "    return sortedKeyList, holder"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def notValidHeader"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validHeader (line):\n",
      "    \"\"\"\n",
      "    Check if line is not a valid gff3 header.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line' - line to check\n",
      "    \"\"\"\n",
      "    if line == \"##gff-version 3\\n\":\n",
      "        return True\n",
      "    else:\n",
      "        return False"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validTabStructure"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validTabStructure(line):\n",
      "    \"\"\"\n",
      "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line': line to check\n",
      "    \"\"\"\n",
      "    theLine = line.strip().split(\"\\t\")  #split on tab\n",
      "    if len(theLine) != 9: # 9 columns in a proper gff file\n",
      "        return False\n",
      "    else:\n",
      "        return True"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validColumn1"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validColumn1(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
      "    the sequence name on the Gbrowse database and have only valid characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of entry in column 1 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)  # need name check, for now just check for valid charachters"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validColumn2"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validColumn2(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
      "    which only need validity of the characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validColumn3"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validColumn3(text, validTypes=None):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
      "    For phage this should be one of types typically ['gene','mRNA','exon']\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    - 'validTypes': List of valid types for checking\n",
      "    \"\"\"\n",
      "    if validTypes is None:\n",
      "        validTypes = {'gene','mRNA','exon'}\n",
      "    \n",
      "    return text in validTypes"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validCoordinate"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validCoordinate(coord):\n",
      "    \"\"\"\n",
      "    Checks a coordinate (entries in column 4 and 5), should be a positive integer\n",
      "    \n",
      "    Parameters:\n",
      "    - 'coord': coordinate from column 4 or 5\n",
      "    \"\"\"\n",
      "    try:\n",
      "        x = int(coord)\n",
      "    except:\n",
      "        isCoordinate = False\n",
      "        pass\n",
      "    else:\n",
      "        if x > 0:\n",
      "            print \"tested positive\"\n",
      "            isCoordinate = True\n",
      "        else:\n",
      "            print \"tested negative\"\n",
      "            isCoordinate = False\n",
      "    \n",
      "    return isCoordinate"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validScore"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validScore(score):\n",
      "    \"\"\"\n",
      "    Checks a score (entries in column 6), should be a number or \".\"\n",
      "    \n",
      "    Parameters:\n",
      "    - 'score': score from column 6\n",
      "    \"\"\"\n",
      "    if score == \".\":\n",
      "        return True\n",
      "    \n",
      "    try:\n",
      "        x = float(score)\n",
      "    except:\n",
      "        isScore = False\n",
      "        pass\n",
      "    else:\n",
      "        isScore = True        \n",
      "\n",
      "    return isScore"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validStrand"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validStrand(strand):\n",
      "    \"\"\"\n",
      "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'strand': strand entry from column 7\n",
      "    \"\"\"\n",
      "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
      "    \n",
      "    return strand in validValues"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validPhase"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validPhase(phase):\n",
      "    \"\"\"\n",
      "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'phase': strand entry from column 8\n",
      "    \"\"\"\n",
      "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
      "    \n",
      "    return phase in validValues"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validAttributes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validAttributes(attributes):\n",
      "    \"\"\"\n",
      "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
      "    It is OK to have spaces in the values entry but not the Key value\n",
      "    \n",
      "    \n",
      "    Parameters:\n",
      "    - 'attributes': entire entry from column 9\n",
      "    \"\"\"\n",
      "    \n",
      "    #ok to have a null string\n",
      "    if attributes == '':\n",
      "        return True\n",
      "    \n",
      "     #checking for invalid characters, spaces OK in column 9 values so remove any before checking with charCheck\n",
      "\n",
      "        \n",
      "    #ok go ahead and split into the underlyine key=value items    \n",
      "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
      "    \n",
      "    validity = True\n",
      "    \n",
      "    for attrPair in attrPairs:\n",
      "        print attrPair\n",
      "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
      "            continue\n",
      "            \n",
      "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
      "            validity = False\n",
      "        \n",
      "        attrKey, attrValue = attrPair.split('=')\n",
      "        \n",
      "        if len(attrKey) < 1:                #must have an entry for a key\n",
      "            validity = False\n",
      "        \n",
      "        if charCheck(attrKey):              #check key for valid charachters\n",
      "            validity = False\n",
      "        \n",
      "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
      "        if charCheck(attrValue.replace(' ','')):\n",
      "            validity = False\n",
      "            \n",
      "    return validity"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "string = \"name!=big dye;x=y\"\n",
      "validAttributes(string)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def fileCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fileCheck(keyList, holder, Seq, types = ['gene','mRNA','exon']):\n",
      "    \"\"\"\n",
      "    Checks each component of each line of the gff file for proper format. These are considered \"entry\" errors.\n",
      "    All entry errors will be added to Errors dictionary.\n",
      "\n",
      "    Parameters:\n",
      "    -'keyList': list of sorted keys.\n",
      "    -'holder': dictionary of lines paired with keys in 'keyList'.\n",
      "    -'Seq': string of the nucleotide sequence extracted from uploaded fasta file.\n",
      "    -'types': a list indicating the types of each gff line and the order the types should\n",
      "              be sorted in. The first type in the list will be ordered before the second, \n",
      "              the second type before the third, etc.\n",
      "\n",
      "\"\"\"\n",
      "        \n",
      "    entry = []                  \n",
      "    idList = set()                 #list of ID's seen to check for uniquenss\n",
      "    \n",
      "    \n",
      "    \n",
      "    for line,key in enumerate(keyList[1:]):   #skipping first line as it is the contig\n",
      "        attributes = dict()                   #will be used to hold each Key/value pair of column 9 for each line\n",
      "        \n",
      "        lineCount = line + 1    #adding 1 for unacccounted for header line\n",
      "        \n",
      "        try:\n",
      "            theLine = holder[key].strip().split(\"\\t\") ## try here to catch if  not tab deliminated or adding extra lines\n",
      "        except:\n",
      "            entry.append(\"Format Error: each line needs to be tab deliminated.\")\n",
      "            continue\n",
      "        \n",
      "        if theLine[2] == \"contig\":\n",
      "            continue\n",
      "            \n",
      "        if len(theLine) != 9:   #this is just to skip improper lines, error already caught by sortGFF3\n",
      "            continue\n",
      "        \n",
      "            \n",
      "        ### 1ST and 2ND ITEM ### \n",
      "        try:\n",
      "            if charCheck(theLine[0]):\n",
      "                raise LineError(\"0002\")\n",
      "            elif charCheck(theLine[1]):\n",
      "                raise LineError(\"0003\")\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "        \n",
      "        ### 3RD ITEM ###\n",
      "        try:\n",
      "            if theLine[2] not in types:\n",
      "                raise LineError(\"0004\")  ## types can be changed if more types of line needed\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "            continue                     ##no other checks done if not a proper type\n",
      "        \n",
      "        ### 4TH ITEM ###\n",
      "        try:\n",
      "            if not int(theLine[3]) > 0:\n",
      "                raise LineError(\"0005\") ## needs to be greater than 0\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "        except ValueError:\n",
      "            entry.append(\"\\n[\" + holder[key].strip() +\"]:\\n\" + \"] coordinates in column 4 not readable as number\") \n",
      "        \n",
      "        \n",
      "        ### 5TH ITEM ###\n",
      "        try:\n",
      "            if not int(theLine[4]) > int(theLine[3]):\n",
      "                raise LineError(\"0006\") ## needs to be greater than 1st coordinate\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "        except ValueError:                  #lines of this type should have been removed but just in case....\n",
      "            entry.append(\"\\n[\" + holder[key].strip() +\"]:\\n\" + \" coordinates in column 5 not readable as number\") \n",
      "        \n",
      "        ### 6TH ITEM ###\n",
      "        try:\n",
      "            if theLine[5] != \".\": ## can also be a number\n",
      "                try:\n",
      "                    int(theLine[5])\n",
      "                except ValueError:\n",
      "                    raise LineError(\"0007\")\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "            \n",
      "        ### 7TH ITEM ###\n",
      "        try:\n",
      "            if theLine[6] != \"+\":\n",
      "                if theLine[6] != \"-\":\n",
      "                    raise LineError(\"0008\")  ## + or - strand\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "                \n",
      "        ### 8TH ITEM ### \n",
      "        try:\n",
      "            if theLine[7] != \".\":\n",
      "                raise LineError(\"0009\")\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "                \n",
      "        \n",
      "        ### 9TH ITEM ###      \n",
      "        ### 9th item is an attributes list \";\" separated list of key/value pairs in the form key=value\n",
      "\n",
      "        #start by checking for invalid characters, spaces OK in here so remove before checking\n",
      "        try:\n",
      "            if charCheck(theLine[8].replace(\" \", \"\")):\n",
      "                raise LineError(\"0011\")\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "            continue              #don't even try to parse if invalid charachters, just move to next line \n",
      "\n",
      "\n",
      "        #checking for proper separation of key/value pairs with the ';' \n",
      "        try:\n",
      "            if theLine[8].count(\"=\") > (theLine[8].count(\";\") + 1):\n",
      "                raise LineError(\"0028\")\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "            continue               #again don't even try with this much failure\n",
      "                \n",
      "        #ok go ahead and split into the underlyine key=value items    \n",
      "        attrItems = theLine[8].split(\";\")   #list of each attribute pair\n",
      "\n",
      "\n",
      "        #run through the list of attrItems; make you they are in key=value and put everything in attributes dict\n",
      "        for attr in attrItems:\n",
      "            if len(attr) == 0:           #catch attr from extra ; but just skip it\n",
      "                continue\n",
      "            try:        ## start with \"=\" search\n",
      "                if attr.find(\"=\") < 0:\n",
      "                    raise LineError(\"0027\")\n",
      "            except LineError as er:\n",
      "                entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "                continue                 #if it doesn't have an \"=\" don't bother with the checks below\n",
      "                    \n",
      "            #to get to here attr has to have at least one \"=\"; so split to check each of key and value\n",
      "            attributeName, attributeValue = attr.split(\"=\")\n",
      "            try:                        #make sure of both a attributeName (key) and attributeValue (value)\n",
      "                if len(attributeName) < 1 or len(attributeValue) < 1:\n",
      "                    raise LineError(\"0027\")\n",
      "                else:                   #ok to add to attributes dict\n",
      "                    attributes[attributeName] = attributeValue\n",
      "            except LineError as er:\n",
      "                entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "                    \n",
      "        #start parsing the attributes dictionary\n",
      "        try:            #do checks for ID presence and uniqueness\n",
      "            if \"ID\" in attributes:\n",
      "                try:\n",
      "                    if attributes[\"ID\"] in idList:\n",
      "                        raise LineError(\"0012\") ## each id can only be used once \n",
      "                    else:\n",
      "                        idList.add(attributes[\"ID\"])\n",
      "                except LineError as er:\n",
      "                    entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "                                \n",
      "            else:\n",
      "                raise LineError(\"0013\")  ##ID is missing\n",
      "        except LineError as er:\n",
      "            entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "\n",
      "                     \n",
      "\n",
      "        ######################## checks that are specific to certain types #############\n",
      "        if theLine[2] == types[0]:       #is a gene line; check the biology and check for a Name\n",
      "            geneCheck(Seq,holder[key])\n",
      "            try:  #do checks for Name presence\n",
      "                if \"Name\" in attributes:\n",
      "                    _Name = True\n",
      "                else:\n",
      "                    raise LineError(\"0029\")\n",
      "            except LineError as er:\n",
      "                entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "\n",
      "        elif types.index(theLine[2]) > 0:    #is an mRNA or exon, check for parent key FIXME: and valid value\n",
      "            try:  #do checks for parent presence\n",
      "                if \"Parent\" in attributes:\n",
      "                    _Parent = True\n",
      "                else:\n",
      "                    raise LineError(\"0025\")\n",
      "            except LineError as er:\n",
      "                entry.append(er.returnError() + \"\\n[\" + holder[key].strip() +\"]\\n\")\n",
      "        \n",
      "        else:\n",
      "            continue                   #this should not happen, but just in case, skip to the next line\n",
      "        \n",
      "                \n",
      "    print \"entry\"\n",
      "    print len(entry)\n",
      "    Errors[\"entry\"] = entry                    #add the list of entry errors to the Errors dictionary\n",
      "\n",
      "    return\n",
      "    \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def charCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters not a-zA-Z0-9.=; and returns True if such a character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    return bool(search(str))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "string = \"&\"\n",
      "charCheck(string) & False\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def geneCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def geneCheck(seq, line):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is a proper gene. Writes errors to global list\n",
      "    'Errors'.\n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has one stop codon as last codon.\n",
      "    - no internal stop codon.\n",
      "    - begins with a valid start codon [ATG, GTG, CTG].\n",
      "    \n",
      "    Parameters:\n",
      "    -'Seq': string nucleotide sequence of entire phage.\n",
      "    -'line': single line in gff3 format for checking.\n",
      "    \"\"\"\n",
      "    #FIXME: need to add check that all coord are within the range of the contig\n",
      "    \n",
      "\n",
      "    if \"biology\" not in Errors:\n",
      "        Errors[\"biology\"]=[]        #dictionary of errors based on biology like start codon, stop codon\n",
      "    \n",
      "    errorCode = 1000\n",
      "    \n",
      "    leftCoordinate= int(line.split(\"\\t\")[3])\n",
      "    rightCoordinate= int(line.split(\"\\t\")[4])\n",
      "    strand = line.split(\"\\t\")[6]\n",
      "    geneSeq = seq[leftCoordinate-1:rightCoordinate]   #not really the gene for negative strand genes but good enough\n",
      "    \n",
      "    print \"checking gene\"\n",
      "    print line\n",
      "    print str(geneSeq[:10]) + \" ... \" + str(geneSeq[-10:])\n",
      "    \n",
      "    ##set up values based on strand of gene\n",
      "    if  strand == \"+\":\n",
      "        firstCodon = geneSeq[:3].upper()\n",
      "        lastCodon = geneSeq[-3:].upper()\n",
      "        startCodons = ['ATG', 'TTG', 'GTG']\n",
      "        stopCodons = ['TAA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        firstCodon = geneSeq[-3:].upper()\n",
      "        lastCodon = geneSeq[:3].upper()\n",
      "        startCodons = ['CAT', 'CAA', 'CAC']   #reverse complement of the start codons\n",
      "        stopCodons = ['TTA', 'CTA', 'TCA']   #reverse complement of the stop codons\n",
      "    \n",
      "    \n",
      "    #test length is multiple of 3\n",
      "    try:\n",
      "        print \"going into 3div check\"\n",
      "        if len(geneSeq)%3 != 0:\n",
      "            print \"found div 3 error, calling error\"\n",
      "            print \"geneSeq \" + geneSeq\n",
      "            raise BiologyError(\"1010\")\n",
      "    except BiologyError as er:\n",
      "        Errors[\"biology\"].append(er.returnError() + \"\\n\" + \"[\" + line + \"]\\n\")\n",
      "        return\n",
      "    \n",
      "    if firstCodon not in startCodons:\n",
      "        print \"no start, adding 100\"\n",
      "        errorCode += 100\n",
      "    \n",
      "    stopCodonCount = countStopCodons(geneSeq,strand)\n",
      "    if  stopCodonCount == 0:\n",
      "        print \"no stop, adding 1000\"\n",
      "        errorCode += 1000\n",
      "            \n",
      "    if stopCodonCount > 1:\n",
      "        print \"Too many stop codons adding 2000\"\n",
      "        errorCode += 2000\n",
      "\n",
      "    try:\n",
      "        print \"end of error encoding\"\n",
      "        print errorCode\n",
      "        if errorCode > 1000:\n",
      "            print \"in the main section, errorCode = \" + str(errorCode)\n",
      "            raise BiologyError(str(errorCode)) ## use sum to pick the             \n",
      "    except BiologyError as er:\n",
      "        Errors[\"biology\"].append(er.returnError() + \"\\n\" + \"[\" + line + \"]\\n\")\n",
      "    \n",
      "                        #add the list of biology errors to the Errors dictionary\n",
      "    \n",
      "\n",
      "    return\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "countStopCodons"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def countStopCodons(seq, strand):\n",
      "    \"\"\"\n",
      "    Used to count the number of stop codons in a given sequence\n",
      "\n",
      "    Parameters:\n",
      "    -'seq': nucleotide sequence of plus strand.\n",
      "    -'strand': indication if gene is on the plus or minus strand\n",
      "\n",
      "    Output:\n",
      "    -'count': total number of stop codons found in the sequence.\n",
      "    \"\"\"\n",
      "    \n",
      "    count = 0\n",
      "    \n",
      "    if strand == \"+\":\n",
      "        stopCodons = ['TTA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        stopCodons = ['TAA', 'CTA', 'TCA']   ##reverse complement of stop codons\n",
      "        \n",
      "    #now create a list of all codons using list comprehension, this only works well if len%3 == 0\n",
      "    if strand == \"+\":\n",
      "        codons=[seq[x:x+3] for x in range(0,len(seq),3)]\n",
      "    else:\n",
      "        codons=[seq[x:x+3] for x in range(-3,-len(seq),-3)] #works for all but first codon becuase string[-3:0]=''\n",
      "        codons[0] = seq[-3:]                                #so fudge in the \"first\" codon on the negative side\n",
      "    \n",
      "    for stop in stopCodons:\n",
      "        count += codons.count(stop)\n",
      "        \n",
      "    return count\n",
      "\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def validGene"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validGene(seq, line):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is define a proper gene.\n",
      "    \n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has one stop codon and it is the last codon.\n",
      "    - begins with a valid start codon [ATG, GTG, CTG].\n",
      "    \n",
      "    Parameters:\n",
      "    -'Seq': string nucleotide sequence of entire phage.\n",
      "    -'line': single line in gff3 format for checking.\n",
      "    \"\"\"\n",
      "    \n",
      "    leftCoordinate= int(line.split(\"\\t\")[3])\n",
      "    rightCoordinate= int(line.split(\"\\t\")[4])\n",
      "    strand = line.split(\"\\t\")[6]\n",
      "    geneSeq = seq[leftCoordinate-1:rightCoordinate]   #not really the gene for negative strand genes but good enough\n",
      "    \n",
      "    if  strand == \"+\":\n",
      "        firstCodon = geneSeq[:3].upper()\n",
      "        lastCodon = geneSeq[-3:].upper()\n",
      "        startCodons = ['ATG', 'TTG', 'GTG']\n",
      "        stopCodons = ['TAA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        firstCodon = geneSeq[-3:].upper()\n",
      "        lastCodon = geneSeq[:3].upper()\n",
      "        startCodons = ['CAT', 'CAA', 'CAC']   #reverse complement of the start codons\n",
      "        stopCodons = ['TTA', 'CTA', 'TCA']   #reverse complement of the stop codons\n",
      "\n",
      "    #first make sure the values of leftCoordinate and rightCoordinate are OK\n",
      "    \n",
      "    if (leftCoordinate >= len(seq)) or (rightCoordinate > len(seq)):  #GFF3 is 1 based counting \n",
      "        return False\n",
      "    \n",
      "    if (rightCoordinate - leftCoordinate) < 30:                       #minimum gene size for this check is 10 a.a.\n",
      "        return False\n",
      "    \n",
      "    if len(geneSeq)%3 != 0:\n",
      "        return False\n",
      "    \n",
      "    if firstCodon not in startCodons:\n",
      "        return False\n",
      "    \n",
      "    if lastCodon not in stopCodons:\n",
      "        return False\n",
      "    \n",
      "    stopCodonCount = countStopCodons(geneSeq,strand)\n",
      "    if  stopCodonCount > 1:\n",
      "        return False\n",
      "    \n",
      "    return True"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "report results"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "global Errors\n",
      "Errors =dict()\n",
      "Keylist,holder = sortGff3(\"docs/Phabio_biology.gff3\")\n",
      "\n",
      "seq = fastaRead(\"docs/Phabio.fasta\")\n",
      "\n",
      "fileCheck(Keylist, holder, seq, types = ['gene','mRNA','exon'])\n",
      "\n",
      "print \"===========Format or Structural Issues================\"\n",
      "print \"====These errors are severe enough that these lines===\" \n",
      "print \"========are removed before subsequent checking========\\n\"\n",
      "\n",
      "for l in Errors['structure']:\n",
      "    print l\n",
      "    \n",
      "    \n",
      "print \"\\n==============Data issues=================\"\n",
      "print   \"=====These lines fail data validation=====\"\n",
      "\n",
      "for l in Errors['entry']:\n",
      "    print l\n",
      "\n",
      "print \"\\n===================Biological issues===================\"\n",
      "print   \"=====These lines fail proper biological structure======\"\n",
      "\n",
      "\n",
      "for l in Errors['biology']:\n",
      "    print l"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def fastaRead"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fastaRead(fasta_File):\n",
      "    \"\"\"\n",
      "    Reads in a nucleotide sequence from a text file in .fasta format.\n",
      "\n",
      "    Parameters:\n",
      "    -'fasta_File': uploaded text file in .fasta format\n",
      "\n",
      "    Output:\n",
      "    -'seq': string of a nucleotide sequence.\n",
      "    \"\"\"\n",
      "\n",
      "    seq = []\n",
      "    try:\n",
      "        f1 = open(fasta_File, \"r\")\n",
      "    except:\n",
      "        return \"Error: unable to read fasta file.\"\n",
      "    \n",
      "    for line in f1:\n",
      "        line = line.rstrip()\n",
      "        if line.startswith(\">\"):\n",
      "            continue\n",
      "        else:\n",
      "            seq.append(line)\n",
      "\n",
      "    f1.close()\n",
      "    return ''.join(seq)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def translate"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def translate(nucSeq):\n",
      "    \"\"\"\n",
      "    Translates a nucleotide sequence into a protein sequence. Uses a dictionary of nucleotides\n",
      "    paired with the protein they translate for.\n",
      "\n",
      "    Parameters:\n",
      "    -'nucSeq': string of nucleotides.\n",
      "\n",
      "    Output:\n",
      "    -'protSeq': string of proteins.\n",
      "    \"\"\"\n",
      "    \n",
      "    codonLib = {'TTT':'F','TTC':'F','TTA':'L','TTG':'L','CTT':'L','CTC':'L','CTA':'L',\n",
      "                'CTG':'L','ATT':'I','ATC':'I','ATA':'I','ATG':'M','GTT':'V','GTC':'V',\n",
      "                'GTA':'V','GTG':'V','TCT':'S','TCC':'S','TCA':'S','TCG':'S','CCT':'P',\n",
      "                'CCC':'P','CCA':'P','CCG':'P','ACT':'T','ACC':'T','ACA':'T','ACG':'T',\n",
      "                'GCT':'A','GCC':'A','GCA':'A','GCG':'A','TAT':'Y','TAC':'Y','TAA':'*',\n",
      "                'TAG':'*','CAT':'H','CAC':'H','CAA':'Q','CAG':'Q','AAT':'N','AAC':'N',\n",
      "                'AAA':'K','AAG':'K','GAT':'D','GAC':'D','GAA':'E','GAG':'E','TGT':'C',\n",
      "                'TGC':'C','TGA':'*','TGG':'W','CGT':'R','CGC':'R','CGA':'R','CGG':'R',\n",
      "                'AGT':'S','AGC':'S','AGA':'R','AGG':'R','GGT':'G','GGC':'G','GGA':'G','GGG':'G'}\n",
      "    \n",
      "    cntLoc = 0\n",
      "    cntCodon = 0\n",
      "    codon = ''\n",
      "    protSeq = ''\n",
      "    while cntLoc < len(nucSeq):\n",
      "        codon = codon + nucSeq[cntLoc]\n",
      "        cntCodon += 1\n",
      "        if cntCodon == 3:\n",
      "            temp = codonLib[codon]\n",
      "            protSeq = protSeq + temp\n",
      "            codon = ''\n",
      "            cntCodon = 0\n",
      "        cntLoc += 1\n",
      "        \n",
      "    return protSeq"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def outFile"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def outFile(keyS, holderS, nameS, errors, nameE, incLine):\n",
      "    \"\"\"\n",
      "    Writes a sorted gff file and an errors text file.\n",
      "    \n",
      "    Parameters:\n",
      "    -'keyS': list of sorted keys.\n",
      "    -'holderS': dictionary of lines from the gff file paired with keys in 'keyS'.\n",
      "    -'nameS': name of new sorted gff file. Also location where the file is written.\n",
      "    -'errors': list of all errors found in uploaded files to be written to text file.\n",
      "    -'nameE': name of new errors text file. Also location where the file is written.\n",
      "    -'incLine': boolean indicating whether lines from the sorted file should be included in \n",
      "                errors file.\n",
      "    \"\"\"\n",
      "    outE = nameE\n",
      "    outS = nameS\n",
      "    \n",
      "    if incLine:                                        #1\n",
      "        for item in errors:                             #2  \n",
      "            if \"Coordinate\" in item:                     #3\n",
      "                tempE = item.split(\" \")[1]                #4\n",
      "                for key in keyS:                          #4\n",
      "                    if tempE == key.split(\"_\")[1]:         #5\n",
      "                        outE.write(holderS[key])            #6\n",
      "                        break                               #6    \n",
      "\n",
      "            else:                                        #3   \n",
      "                tempE = int(item[1])                      #4\n",
      "                outE.write(holderS[keyS[tempE]])          #4\n",
      "\n",
      "            outE.write(item)                             #3                          \n",
      "            outE.write(\"\\n\")                             #3\n",
      "            outE.write(\"\\n\")                             #3\n",
      "\n",
      "    else:                                              #1\n",
      "        for item in errors:                             #2\n",
      "            outE.write(item)                             #3\n",
      "            outE.write(\"\\n\")                             #3\n",
      "\n",
      "    for key in keyS:                                   #1\n",
      "        outS.write(holderS[key])                        #2\n",
      "        outS.write(\"\\n\")                                #2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "seq = \"TTAGTCATCATCATCATCATCTAGCATCATCATCATCATCATCATCATCATCATCATCATA=AATGATCTGTATAGACTGTTGACGATG\"\n",
      "strand = '-'\n",
      "x,y = seq.split('=')\n",
      "print seq\n",
      "print x\n",
      "print y"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "seq = \"TTAGTCATCATCATCATCATCTAGCATCATCATCATCATCATCATCATCATCATCATCATAAATGATCTGTATAGACTGTTGACGATG\"\n",
      "strand = '-'\n",
      "x = countStopCodons(seq,strand)\n",
      "seq[-3:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Define Error Classes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class ValidationError(Exception):\n",
      "    \"\"\"\n",
      "    Exception classes:\n",
      "    - 'Validation Error': basic exception class. Used to catch unknown errors. \n",
      "    - 'Format Error': catches errors dealing with line format.\n",
      "    - 'Line Error': catches errors dealing with individuel components of a line.\n",
      "    - 'Biology Error': catches errors dealing with the described genes.\n",
      "    \"\"\"\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        _dict = dict()\n",
      "        _dict['0000'] = \"Usage Error: Unknown error code given used. Use printDict() to see dictionary of errors\"\n",
      "        _dict['1000'] = \"Validation Error: unknown\"\n",
      "        \n",
      "        self.code = str(code)\n",
      "        self._dict = _dict\n",
      "        self.message = message\n",
      "        \n",
      "        super(ValidationError, self).__init__(code, message)\n",
      "        \n",
      "    def printDict(self):\n",
      "        for k, v in self._dict.iteritems():\n",
      "            print k + \":  \" + v\n",
      "            \n",
      "    def returnError(self):\n",
      "        try:\n",
      "            return self._dict[self.code]\n",
      "        except:\n",
      "            return self._dict['0000']\n",
      "\n",
      "class FormatError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(FormatError, self).__init__(code, message)\n",
      "        \n",
      "        self._dict['0100'] = \"Format Error: less than 9 columns separated by <tab>\"\n",
      "        self._dict['0200'] = \"Format Error: more than 9 columns separated by <tab>\"\n",
      "        self._dict['0300'] = \"Format Error: tab deliminated\"\n",
      "        self._dict['0400'] = \"Format Error: multiple contigs. There should only be one contig line per file\"\n",
      "        self._dict['0500'] = \"Format Error: Errors in defined types. Must have one each of [gene, mRNA, exon]\"\n",
      "        self._dict['0600'] = \"Format Error: ensure only one sequence in fasta file\"\n",
      "        self._dict['0700'] = \"Format Error: invalid type in column 3; must be EXACTLY (case sensitive) 'gene' or 'mRNA' or 'exon'\"\n",
      "        self._dict['0800'] = \"Format Error: coordinates in column 4 and/or 5 cannot be read as numbers\"\n",
      "\n",
      "\n",
      "class LineError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(LineError, self).__init__(code, message)\n",
      "\n",
      "        self._dict['0001'] = \"Error: unknown\"\n",
      "        self._dict['0002'] = \"1st column: restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0003'] = \"2nd column: restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0004'] = \"3rd column: type invalid, must exactly match one of [gene, mRNA, exon]\"\n",
      "        self._dict['0005'] = \"4th column: 1st coordinate must be a positive integer\"\n",
      "        self._dict['0006'] = \"5th column: 2nd coordinate must be an integer (and be) greater than the 1st coordinate\"\n",
      "        self._dict['0007'] = \"6th column: must be a number or '.'\"\n",
      "        self._dict['0008'] = \"7th column: must indicate strand as either '+' or '-'; no other characters allowed\"\n",
      "        self._dict['0009'] = \"8th column: For phage (with no introns) this column must be '.'\"\n",
      "        self._dict['0011'] = \"9th column: restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0012'] = \"9th column: ID already used. Each gene must have a unique ID\"\n",
      "        self._dict['0013'] = \"9th column: Lines must have a valid ID\"\n",
      "        self._dict['0021'] = \"9th column: bad sort or a gene does not have a line for each type. Default types are [gene, mRNA, exon]\"\n",
      "        self._dict['0022'] = \"9th column: restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0023'] = \"9th column: ID already used. Each ID must be unique\"\n",
      "        self._dict['0024'] = \"9th column: last type, must at least have a Parent\"\n",
      "        self._dict['0025'] = \"9th column: Lines of this type must have a Parent\"\n",
      "        self._dict['0026'] = \"First line must be EXACTLY: '##gff-version 3' (HINT: check for unseen spaces or tabs)\"\n",
      "        self._dict['0027'] = \"9th column: items in this column must be in the form <name>=<value>, separated by ';'\"\n",
      "        self._dict['0028'] = \"9th column: items must be separated by ';', too many '=' without enough ';' \"\n",
      "        self._dict['0029'] = \"9th column: Genes must have a valid Name\"\n",
      "\n",
      "\n",
      "        \n",
      "class BiologyError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(BiologyError, self).__init__(code, message)\n",
      "        \n",
      "        self._dict['1010'] = \"Gene is not a multiple of 3, start/stop codons not checked\"\n",
      "        self._dict['1100'] = \"Start codon missing, recheck coordinates\"\n",
      "        self._dict['2000'] = \"Stop codon missing, recheck coordinates\"\n",
      "        self._dict['2100'] = \"Start & stop codons missing, recheck coordinates\"\n",
      "        self._dict['4000'] = \"Start codon present but more than 1 stop codon, recheck coordinates\"        \n",
      "        self._dict['4100'] = \"Start codon missing & more than 1 stop codon, recheck coordinates\"\n",
      "        \n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Main"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def main(gff3_File, seq_File):\n",
      "    global Errors\n",
      "    Errors = dict()\n",
      "    seq = fastaRead(seq_File)\n",
      "    \n",
      "    sorted_File = sortGff3(gff3_File)       \n",
      "    \n",
      "    report = fileCheck(sorted_File, seq)\n",
      "    \n",
      "    newName = gff3_File[0:-5] + \"_sorted_Errors.txt\"\n",
      "    out = open(newName,\"w\")\n",
      "    for item in Errors:\n",
      "        out.write(item)\n",
      "        out.write(\"\\n\")\n",
      "    out.close()\n",
      "    return [sorted_File, newName]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "test code"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "read fasta"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "get name of fasta file\n",
      "read in fasta file\n",
      "check fasta file\n",
      "    check line 1 is proper header\n",
      "    check every other line\n",
      "        check structure # must have 9 components\n",
      "        check every item"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "phabioSeq[40:371]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "test sortGFF3"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "global Errors\n",
      "Errors = []\n",
      "\n",
      "#f1 = open(\"Jerm2-final-genes.txt\", \"r\")\n",
      "#for line in f1:\n",
      "#    theLine = line.strip().split(\"\\t\")\n",
      "#    for comp in theLine:\n",
      "#        print comp\n",
      "#    print \"\\n\\n\"\n",
      "\n",
      "derp = sortGff3(\"Jerm2-final-genes.txt\")\n",
      "print derp[1]['gene_599']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fileCheck(\"Jerm2-final-gene_sorted.gff3\", \"Jerm2.fasta\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Errors=[]\n",
      "derp = fastaRead(\"Jerm2.fasta\")\n",
      "geneCheck(31199,31429,derp,21)\n",
      "print Errors"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene = \"N/A\"\n",
      "smlCount = 0\n",
      "\n",
      "for theSeq in SeqIO.parse(\"Jerm2.fasta\", \"fasta\"):\n",
      "    gene = theSeq\n",
      "gene.seq.translate() \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Errors = []\n",
      "prot = geneCheck(1, 60, \"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAAATGATGATGATGATGATGATG\", 35)\n",
      "print Errors"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(\"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAA\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def geneCheckBio(coord1, coord2, seq, count):\n",
      "    gene = \"N/A\"\n",
      "    smlCount = 0\n",
      "    \n",
      "    try:\n",
      "        for theSeq in SeqIO.parse(seq, \"fasta\"):\n",
      "            gene = theSeq[coord1:coord2]\n",
      "            smlCount += 1\n",
      "        if smlCount != 1:\n",
      "            raise FormatError(\"0600\")\n",
      "    except FormatError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "        return\n",
      "    except:\n",
      "        Errors.append(\"Error in parsing the fasta file. Ensure correct fasta format\")\n",
      "        return\n",
      "    \n",
      "    try:\n",
      "        if len(gene)%3 != 0:\n",
      "            raise BiologyError(\"0050\") ## divisible by three\n",
      "    except BiologyError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "        \n",
      "    try:\n",
      "        protein = gene.seq.translate() ## errors in translating the gene\n",
      "    except:\n",
      "        Errors.append(\"[\" + str(count) + \"]  Biology Error: unable to translate sequence. Ensure sequence provided is a nucleotide sequence.\")\n",
      "        return\n",
      "    \n",
      "    try:\n",
      "        if protein.count(\"*\") == 0:\n",
      "            raise BiologyError(\"0030\") ## no stop codon\n",
      "        if protein.count(\"*\") > 1:\n",
      "            raise BiologyError(\"0040\") ## internal stop codons\n",
      "        if protein[0] != \"M\":\n",
      "            raise BiologyError(\"0020\") ## has to start with M\n",
      "    except BiologyError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "    \n",
      "    return "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Errors['structure']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(\"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAAATGATGATGATGATGATGATG\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def errorsToFile(errorsList, gff3_File_name):\n",
      "    \n",
      "    \n",
      "    \n",
      "    newName = gff3_File_name[0:-5] + \"_sorted_Errors.txt\"\n",
      "    out = open(newName,\"w\")\n",
      "    for item in errorsList:\n",
      "        out.write(item)\n",
      "    out.close()\n",
      "    return newName\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def main2():\n",
      "    global bigG\n",
      "    bigG = ['derp','merp']\n",
      "    sec()\n",
      "    printGl()\n",
      "    \n",
      "def sec():\n",
      "    bigG.append('poke')\n",
      "    \n",
      "def printGl():\n",
      "    print bigG"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "once = ['one','two','three']\n",
      "once.append('fourth')\n",
      "print once[3]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cay = False\n",
      "\n",
      "try:\n",
      "    if cay:\n",
      "        print \"nom\"\n",
      "    else:\n",
      "        raise ValidationError(\"1000\")\n",
      "except ValidationError as e:\n",
      "    print \"HELLO \" + e.returnError()\n",
      "except:\n",
      "    print \"Missed\"\n",
      "print \"END\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "kay = False\n",
      "\n",
      "try:\n",
      "    if kay:\n",
      "        print \"om\"\n",
      "    else:\n",
      "        raise FormatError(\"100\")\n",
      "except FormatError as e:\n",
      "    print e.printDict()\n",
      "except:\n",
      "    print \"missed me\"\n",
      "print \"end\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "try:\n",
      "    raise FormatError(\"010\")\n",
      "except FormatError as e:\n",
      "    print \"meep\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "prac = dict()\n",
      "prac[\"0300\"] = \"5\"\n",
      "prac[\"0001\"] = \"6\"\n",
      "prac[\"0002\"] = \"7\"\n",
      "prac[\"0003\"] = \"8\"\n",
      "prac[\"0000\"] = \"1\"\n",
      "prac[\"1000\"] = \"2\"\n",
      "prac[\"0100\"] = \"3\"\n",
      "prac[\"0200\"] = \"4\"\n",
      "prac[\"0010\"] = \"11\"\n",
      "prac[\"0020\"] = \"12\"\n",
      "prac[\"0030\"] = \"13\"\n",
      "prac[\"0040\"] = \"14\"\n",
      "\n",
      "prac.keys()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "listKeys = prac.keys()\n",
      "codeNum = len(listKeys)\n",
      "codeLen = len(listKeys[0])\n",
      "prevKey = \"\"\n",
      "index = 0\n",
      "while index < codeNum:\n",
      "    currKey = listKeys[index]\n",
      "    if index == 0:\n",
      "        prevKey = currKey\n",
      "        index += 1\n",
      "        continue\n",
      "    else:\n",
      "        for y in range(0, codeLen):\n",
      "            if currKey[y] == prevKey[y]:\n",
      "                continue\n",
      "            elif currKey[y] > prevKey[y]:\n",
      "                listKeys[index-1] = currKey\n",
      "                listKeys[index] = prevKey\n",
      "                index -= 2\n",
      "                break\n",
      "            else:\n",
      "                continue\n",
      "        index += 1\n",
      "        prevKey = currKey\n",
      "for key in listKeys:\n",
      "    print key\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "validPhase = {\".\", 1, 2, 0 }\n",
      "\"1\" in validPhase"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "startHere()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filename = raw_input('Enter a file name: ')\n",
      "print \"file = \" + filename"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def startHere():\n",
      "    fastaFile = raw_input('Enter the name of the fasta file (BE EXACT): ')\n",
      "    gffFile = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
      "    return fastaFile,gffFile"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Click in Here\n",
      "\n",
      "fastaFileName = raw_input('Enter the name of the fasta file (BE EXACT): ')\n",
      "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
      "\n",
      "fullFastaFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + fastaFileName\n",
      "fullGFFFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + gffFileName\n",
      "\n",
      "try:\n",
      "    with open(fullFastaFileName) as fastaFile:\n",
      "        fastaFileContents = fastaFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get fasta file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "try:\n",
      "    with open(fullGFFFileName) as gffFile:\n",
      "        gffFileContents = gffFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get GFF file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "print \"\\n\\n================================ Results ================================\\n\"\n",
      "        \n",
      "firstline = gffFileContents[0]\n",
      "print firstline\n",
      "\n",
      "if not validHeader(firstline):\n",
      "    print \"##### Fatal Error #####\"\n",
      "    print \"\"\"\n",
      "    GFF header line failed. No other tests run. \n",
      "    Check 1st line, it should be EXACTLY\n",
      "    ##gff-version 3\n",
      "    Most Common error: spaces and/or tabs at end of line after 3.\n",
      "    \n",
      "Computer messages follow, ignore and try again:\n",
      "    \"\"\"\n",
      "\n",
      "    sys.exit()\n",
      "\n",
      "print \"ready for data lines\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gffFileContents[1,2]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters not a-zA-Z0-9.=; and returns True if such a character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    print bool(search(str))\n",
      "    return not bool(search(str))\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bool(charCheck('1.4', search=re.compile(r'[^0-9]').search))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "True\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 32,
       "text": [
        "False"
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "x =  {1,2}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "type(x)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 34,
       "text": [
        "set"
       ]
      }
     ],
     "prompt_number": 34
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}