{
 "metadata": {
  "name": "",
  "signature": "sha256:526ef547372aecf7ecc4523680f34e04e0997ac2865bf7753fea76503bb8f6ba"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def validHeader (line):\n",
      "    \"\"\"\n",
      "    Return True if line is a valid gff3 header.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line' - line to check\n",
      "    \"\"\"\n",
      "    if line != \"##gff-version 3\\n\":\n",
      "        return False\n",
      "    else:\n",
      "        return True\n",
      "    \n",
      "def validTabStructure(line):\n",
      "    \"\"\"\n",
      "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line': line to check\n",
      "    \"\"\"\n",
      "    theLine = line.strip().split(\"\\t\")  #split on tab\n",
      "    if len(theLine) != 9: # 9 columns in a proper gff file\n",
      "        return False\n",
      "    else:\n",
      "        return True\n",
      "\n",
      "def validSeqname(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
      "    the sequence name on the Gbrowse database and have only valid characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of entry in column 1 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
      "\n",
      "def validSource(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
      "    which only need validity of the characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)\n",
      "\n",
      "def validType(text, validTypes=None):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
      "    For phage this should be one of types typically ['gene','mRNA','exon']\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    - 'validTypes': List of valid types for checking\n",
      "    \"\"\"\n",
      "    if validTypes is None:\n",
      "        validTypes = {'gene','mRNA','exon'}\n",
      "    \n",
      "    return text in validTypes\n",
      "\n",
      "def validCoordinate(coord):\n",
      "    \"\"\"\n",
      "    Checks a coordinate (entries in column 4 and 5), should be a positive integer\n",
      "    \n",
      "    Parameters:\n",
      "    - 'coord': coordinate from column 4 or 5\n",
      "    \"\"\"\n",
      "    try:\n",
      "        x = int(coord)\n",
      "    except:\n",
      "        isCoordinate = False\n",
      "        pass\n",
      "    else:\n",
      "        if x > 0:\n",
      "            print \"tested positive\"\n",
      "            isCoordinate = True\n",
      "        else:\n",
      "            print \"tested negative\"\n",
      "            isCoordinate = False\n",
      "    \n",
      "    return isCoordinate\n",
      "\n",
      "def validScore(score):\n",
      "    \"\"\"\n",
      "    Checks a score (entries in column 6), should be a number or \".\"\n",
      "    \n",
      "    Parameters:\n",
      "    - 'score': score from column 6\n",
      "    \"\"\"\n",
      "    if score == \".\":\n",
      "        return True\n",
      "    \n",
      "    try:\n",
      "        x = float(score)\n",
      "    except:\n",
      "        isScore = False\n",
      "        pass\n",
      "    else:\n",
      "        isScore = True        \n",
      "\n",
      "    return isScore\n",
      "\n",
      "def validStrand(strand):\n",
      "    \"\"\"\n",
      "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'strand': strand entry from column 7\n",
      "    \"\"\"\n",
      "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
      "    \n",
      "    return strand in validValues\n",
      "\n",
      "def validPhase(phase):\n",
      "    \"\"\"\n",
      "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'phase': strand entry from column 8\n",
      "    \"\"\"\n",
      "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
      "    \n",
      "    return phase in validValues\n",
      "\n",
      "def validAttributes(attributes):\n",
      "    \"\"\"\n",
      "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
      "    It is OK to have spaces in the values entry but not the Key value\n",
      "    \n",
      "    \n",
      "    Parameters:\n",
      "    - 'attributes': entire entry from column 9\n",
      "    \"\"\"\n",
      "    \n",
      "    #ok to have a null string\n",
      "    if attributes == '':\n",
      "        return True\n",
      "    \n",
      "     #checking for invalid characters, spaces OK in column 9 values so remove any before checking with charCheck\n",
      "\n",
      "        \n",
      "    #ok go ahead and split into the underlyine key=value items    \n",
      "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
      "    \n",
      "    validity = True\n",
      "    \n",
      "    for attrPair in attrPairs:\n",
      "        print attrPair\n",
      "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
      "            continue\n",
      "            \n",
      "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
      "            validity = False\n",
      "        \n",
      "        attrKey, attrValue = attrPair.split('=')\n",
      "        \n",
      "        if len(attrKey) < 1:                #must have an entry for a key\n",
      "            validity = False\n",
      "        \n",
      "        if charCheck(attrKey):              #check key for valid charachters\n",
      "            validity = False\n",
      "        \n",
      "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
      "        if charCheck(attrValue.replace(' ','')):\n",
      "            validity = False\n",
      "            \n",
      "    return validity\n",
      "\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters not a-zA-Z0-9.=; and returns True if such a character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    return bool(search(str))\n",
      "\n",
      "def validGene(seq, line):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is define a proper gene.\n",
      "    \n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has one stop codon and it is the last codon.\n",
      "    - begins with a valid start codon [ATG, GTG, CTG].\n",
      "    \n",
      "    Parameters:\n",
      "    -'Seq': string nucleotide sequence of entire phage.\n",
      "    -'line': single line in gff3 format for checking.\n",
      "    \"\"\"\n",
      "    \n",
      "    leftCoordinate= int(line.split(\"\\t\")[3])\n",
      "    rightCoordinate= int(line.split(\"\\t\")[4])\n",
      "    strand = line.split(\"\\t\")[6]\n",
      "    geneSeq = seq[leftCoordinate-1:rightCoordinate]   #not really the gene for negative strand genes but good enough\n",
      "    \n",
      "    if  strand == \"+\":\n",
      "        firstCodon = geneSeq[:3].upper()\n",
      "        lastCodon = geneSeq[-3:].upper()\n",
      "        startCodons = ['ATG', 'TTG', 'GTG']\n",
      "        stopCodons = ['TAA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        firstCodon = geneSeq[-3:].upper()\n",
      "        lastCodon = geneSeq[:3].upper()\n",
      "        startCodons = ['CAT', 'CAA', 'CAC']   #reverse complement of the start codons\n",
      "        stopCodons = ['TTA', 'CTA', 'TCA']   #reverse complement of the stop codons\n",
      "\n",
      "    #first make sure the values of leftCoordinate and rightCoordinate are OK\n",
      "    \n",
      "    if (leftCoordinate > len(seq)) or (rightCoordinate >= len(seq)):  #zero based counting \n",
      "        return False\n",
      "    \n",
      "    if (rightCoordinate - leftCoordinate) < 30:                       #minimum gene size for this check is 10 a.a.\n",
      "        return False\n",
      "    \n",
      "    if len(geneSeq)%3 != 0:\n",
      "        return False\n",
      "    \n",
      "    if firstCodon not in startCodons:\n",
      "        return False\n",
      "    \n",
      "    if lastCodon not in stopCodons:\n",
      "        return False\n",
      "    \n",
      "    stopCodonCount = countStopCodons(geneSeq,strand)\n",
      "    if  stopCodonCount > 1:\n",
      "        return False\n",
      "    \n",
      "    return True\n",
      "\n",
      "def fastaRead(fasta_File):\n",
      "    \"\"\"\n",
      "    Reads in a nucleotide sequence from a text file in .fasta format.\n",
      "\n",
      "    Parameters:\n",
      "    -'fasta_File': uploaded text file in .fasta format\n",
      "\n",
      "    Output:\n",
      "    -'seq': string of a nucleotide sequence.\n",
      "    \"\"\"\n",
      "\n",
      "    seq = []\n",
      "    try:\n",
      "        f1 = open(fasta_File, \"r\")\n",
      "    except:\n",
      "        return \"Error: unable to read fasta file.\"\n",
      "    \n",
      "    for line in f1:\n",
      "        line = line.rstrip()\n",
      "        if line.startswith(\">\"):\n",
      "            continue\n",
      "        else:\n",
      "            seq.append(line)\n",
      "\n",
      "    f1.close()\n",
      "    return ''.join(seq)\n",
      "\n",
      "def countStopCodons(seq, strand):\n",
      "    \"\"\"\n",
      "    Used to count the number of stop codons in a given sequence\n",
      "\n",
      "    Parameters:\n",
      "    -'seq': nucleotide sequence of plus strand.\n",
      "    -'strand': indication if gene is on the plus or minus strand\n",
      "\n",
      "    Output:\n",
      "    -'count': total number of stop codons found in the sequence.\n",
      "    \"\"\"\n",
      "    \n",
      "    count = 0\n",
      "    \n",
      "    if strand == \"+\":\n",
      "        stopCodons = ['TTA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        stopCodons = ['TAA', 'CTA', 'TCA']   ##reverse complement of stop codons\n",
      "        \n",
      "    #now create a list of all codons using list comprehension, this only works well if len%3 == 0\n",
      "    if strand == \"+\":\n",
      "        codons=[seq[x:x+3] for x in range(0,len(seq),3)]\n",
      "    else:\n",
      "        codons=[seq[x:x+3] for x in range(-3,-len(seq),-3)] #works for all but first codon becuase string[-3:0]=''\n",
      "        codons[0] = seq[-3:]                                #so fudge in the \"first\" codon on the negative side\n",
      "    \n",
      "    for stop in stopCodons:\n",
      "        count += codons.count(stop)\n",
      "        \n",
      "    return count"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}