{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GFF phage file validation program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###README:\n",
    "\n",
    "This program will check your GFF file. It checks for proper format and use of proper characters  (i.e. no illegal charachters used). \n",
    "\n",
    "**Before you run the program:** copy your GFF file into the \"input\" folder on the Desktop. Make sure you know the EXACT name of the file, it is best to avoid spaces in filenames (rename your files if necessary).  \n",
    "\n",
    "**To run the program:** click anywhere in the code box below near the \"#STEP 1: Click inside this box Here\". You should see the blinking curser appear near where you clicked, once you see the curser hit shift-return (i.e. hold shift while hitting the return key).\n",
    "\n",
    "**What will happen:** the first thing to appear will be a small text entry box at the bottom of the page asking for the name of your gff file. Type the name of the gff file into the box. Once the box has the correct name, hit the return key (no shift). \n",
    "\n",
    "Your results should appear directly below the question box."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the name of the gff file (BE EXACT): enkatz_v2.txt\n",
      "\n",
      "\n",
      "================================== Results ==================================\n",
      "\n",
      "16\n",
      "checking line:   1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 \n",
      "\n",
      "Congratulations! all tests passed\n",
      "Everything below can be used in DNA Master\n",
      "=========================== DNA Master documentation ========================\n",
      "CDS 901 - 1092\n",
      "  /gene=Enkatz.1\n",
      "CDS 1128 - 1568\n",
      "  /gene=Enkatz.2\n",
      "CDS 1933 - 3003\n",
      "  /gene=Enkatz.3\n",
      "CDS 3013 - 3717\n",
      "  /gene=Enkatz.4\n",
      "CDS 3731 - 3991\n",
      "  /gene=Enkatz.5\n",
      "CDS 3999 - 4256\n",
      "  /gene=Enkatz.6\n",
      "CDS 4256 - 5689\n",
      "  /gene=Enkatz.7\n",
      "CDS 5679 - 6647\n",
      "  /gene=Enkatz.8\n",
      "CDS 6667 - 7185\n",
      "  /gene=Enkatz.9\n",
      "CDS 7169 - 8857\n",
      "  /gene=Enkatz.10\n",
      "CDS 8854 - 10296\n",
      "  /gene=Enkatz.11\n",
      "CDS 10293 - 11174\n",
      "  /gene=Enkatz.12\n",
      "CDS 11232 - 11744\n",
      "  /gene=Enkatz.13\n",
      "CDS 11754 - 12755\n",
      "  /gene=Enkatz.14\n",
      "CDS 12834 - 13070\n",
      "  /gene=Enkatz.15\n",
      "CDS 13070 - 13519\n",
      "  /gene=Enkatz.16\n",
      "CDS 13519 - 13881\n",
      "  /gene=Enkatz.17\n",
      "CDS 13881 - 14276\n",
      "  /gene=Enkatz.18\n",
      "CDS 14280 - 14726\n",
      "  /gene=Enkatz.19\n",
      "CDS 14735 - 15586\n",
      "  /gene=Enkatz.20\n",
      "CDS 15608 - 16087\n",
      "  /gene=Enkatz.21\n",
      "CDS 16108 - 16509\n",
      "  /gene=Enkatz.22\n",
      "CDS 16521 - 18992\n",
      "  /gene=Enkatz.23\n",
      "CDS 19016 - 21073\n",
      "  /gene=Enkatz.24\n",
      "CDS 21092 - 21274\n",
      "  /gene=Enkatz.25\n",
      "CDS 21274 - 23076\n",
      "  /gene=Enkatz.26\n",
      "CDS 23114 - 23560\n",
      "  /gene=Enkatz.27\n",
      "CDS 23557 - 23904\n",
      "  /gene=Enkatz.28\n",
      "CDS 23901 - 24248\n",
      "  /gene=Enkatz.29\n",
      "CDS 24273 - 26126\n",
      "  /gene=Enkatz.30\n",
      "CDS 26119 - 27597\n",
      "  /gene=Enkatz.31\n",
      "CDS 27690 - 28010\n",
      "  /gene=Enkatz.32\n",
      "CDS 28020 - 28211\n",
      "  /gene=Enkatz.33\n",
      "CDS 28208 - 29167\n",
      "  /gene=Enkatz.34\n",
      "CDS 29288 - 29572\n",
      "  /gene=Enkatz.35\n",
      "CDS 29723 - 31225\n",
      "  /gene=Enkatz.36\n",
      "CDS complement (31348 - 31623)\n",
      "  /gene=Enkatz.37\n",
      "CDS complement (31620 - 31799)\n",
      "  /gene=Enkatz.38\n",
      "CDS complement (31787 - 31882)\n",
      "  /gene=Enkatz.39\n",
      "CDS complement (31879 - 32166)\n",
      "  /gene=Enkatz.40\n",
      "CDS complement (32163 - 32525)\n",
      "  /gene=Enkatz.41\n",
      "CDS complement (32547 - 34373)\n",
      "  /gene=Enkatz.42\n",
      "CDS complement (34360 - 34833)\n",
      "  /gene=Enkatz.43\n",
      "CDS complement (34901 - 35131)\n",
      "  /gene=Enkatz.44\n",
      "CDS complement (35174 - 35668)\n",
      "  /gene=Enkatz.45\n",
      "CDS complement (35665 - 35850)\n",
      "  /gene=Enkatz.46\n",
      "CDS complement (35847 - 36638)\n",
      "  /gene=Enkatz.47\n",
      "CDS complement (36631 - 36771)\n",
      "  /gene=Enkatz.48\n",
      "CDS complement (36776 - 37543)\n",
      "  /gene=Enkatz.49\n",
      "CDS complement (37540 - 37719)\n",
      "  /gene=Enkatz.50\n",
      "CDS complement (37716 - 38114)\n",
      "  /gene=Enkatz.51\n",
      "CDS complement (38111 - 38755)\n",
      "  /gene=Enkatz.52\n",
      "CDS complement (38972 - 39136)\n",
      "  /gene=Enkatz.53\n",
      "CDS complement (39198 - 39332)\n",
      "  /gene=Enkatz.54\n",
      "CDS complement (39329 - 39820)\n",
      "  /gene=Enkatz.55\n",
      "CDS complement (39890 - 40249)\n",
      "  /gene=Enkatz.56\n",
      "CDS complement (40246 - 40512)\n",
      "  /gene=Enkatz.57\n",
      "CDS complement (40512 - 41069)\n",
      "  /gene=Enkatz.58\n",
      "CDS complement (41066 - 41878)\n",
      "  /gene=Enkatz.59\n",
      "CDS complement (41865 - 42011)\n",
      "  /gene=Enkatz.60\n",
      "CDS complement (42008 - 42301)\n",
      "  /gene=Enkatz.61\n",
      "CDS complement (42298 - 42618)\n",
      "  /gene=Enkatz.62\n",
      "CDS complement (42622 - 42804)\n",
      "  /gene=Enkatz.63\n",
      "CDS complement (42801 - 42959)\n",
      "  /gene=Enkatz.64\n",
      "CDS complement (42987 - 43805)\n",
      "  /gene=Enkatz.65\n",
      "CDS complement (43972 - 44289)\n",
      "  /gene=Enkatz.66\n",
      "CDS complement (44286 - 44681)\n",
      "  /gene=Enkatz.67\n",
      "CDS complement (44720 - 45235)\n",
      "  /gene=Enkatz.68\n",
      "CDS complement (45497 - 45796)\n",
      "  /gene=Enkatz.69\n",
      "CDS complement (45793 - 46062)\n",
      "  /gene=Enkatz.70\n",
      "CDS complement (46059 - 46262)\n",
      "  /gene=Enkatz.71\n",
      "CDS complement (46259 - 46447)\n",
      "  /gene=Enkatz.72\n",
      "CDS complement (46447 - 46614)\n",
      "  /gene=Enkatz.73\n",
      "CDS complement (46611 - 46739)\n",
      "  /gene=Enkatz.74\n",
      "CDS complement (46760 - 46909)\n",
      "  /gene=Enkatz.75\n",
      "CDS complement (46906 - 47223)\n",
      "  /gene=Enkatz.76\n",
      "CDS complement (47495 - 47707)\n",
      "  /gene=Enkatz.77\n",
      "CDS complement (47774 - 48058)\n",
      "  /gene=Enkatz.78\n",
      "CDS complement (48327 - 48482)\n",
      "  /gene=Enkatz.79\n",
      "CDS complement (48692 - 48931)\n",
      "  /gene=Enkatz.80\n",
      "CDS complement (48960 - 49058)\n",
      "  /gene=Enkatz.81\n",
      "CDS complement (49153 - 49356)\n",
      "  /gene=Enkatz.82\n",
      "ORIGIN\n"
     ]
    }
   ],
   "source": [
    "#STEP 1: Click inside this box Here\n",
    "\n",
    "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
    "\n",
    "fullGFFFileName = \"/Users/chris/Desktop/input/\" + gffFileName\n",
    "\n",
    "try:\n",
    "    with open(fullGFFFileName) as gffFile:\n",
    "        gffFileContents = gffFile.readlines()\n",
    "except:\n",
    "    gffFileContents = []\n",
    "    print '''Well that didn't work! Could not get GFF file. Are you sure it is in the input folder?\n",
    "    Double check spelling, no spaces in file name, try again. If that fails get help.'''\n",
    "    sys.exit()\n",
    "print \"\\n\\n================================== Results ==================================\\n\"\n",
    "        \n",
    "firstline = gffFileContents[0]\n",
    "\n",
    "print len(firstline)\n",
    "if not validHeader(firstline):\n",
    "    printFailureMessage(\"Check for unix formated file, Header line\")\n",
    "    print firstline\n",
    "    sys.exit()\n",
    "\n",
    "    \n",
    "print \"checking line: \", \n",
    "for lineNum,wholeLine in enumerate(gffFileContents[1:]):\n",
    "    lineNumber = str(lineNum+1)\n",
    "    print \" \" + lineNumber,\n",
    "    \n",
    "    try:\n",
    "        if not validTabStructure(wholeLine):                          #check whole line for proper 8 tab structure\n",
    "            failureMessage = \"Line number: \" + lineNumber + \" tab structure (i.e. 9 entries and 8 tabs)\"\n",
    "            printFailureMessage(failureMessage)\n",
    "            sys.exit()\n",
    "    except:\n",
    "        print \"double check that all functions are defined!\"\n",
    "    \n",
    "    \n",
    "    line = wholeLine.strip().split(\"\\t\")\n",
    "    \n",
    "    if not validSeqname(line[0]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" seq name of \" + line[0]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "    \n",
    "    if not validSource(line[1]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" source definition of \" + line[1] \n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validType(line[2]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" type definition of \" + line[2]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validCoordinates(line[3],line[4]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" coordinates left: \" + line[3] + \" right: \" + line[4] + \" have\"\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validScore(line[5]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" score of \" + line[5]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validStrand(line[6]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" strand entry of \" + line[6]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "    if not validPhase(line[7]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" phase of \" + line[7]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "    \n",
    "    \n",
    "    if not validAttributes(line[8]):\n",
    "        failMessage = \"Line number: \" + lineNumber + \" attribute entry of: \" + line[8]\n",
    "        printFailureMessage(failMessage)\n",
    "        sys.exit()\n",
    "\n",
    "print \"\\n\\nCongratulations! all tests passed\\nEverything below can be used in DNA Master\"\n",
    "print \"=========================== DNA Master documentation ========================\"\n",
    "text = createDNAMasterFile(gffFileContents)\n",
    "print text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Definitions of Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#%%writefile gffTester_nose.py\n",
    "import re,sys\n",
    "\n",
    "def validHeader(line):\n",
    "    \"\"\"\n",
    "    Return True if line is a valid gff3 header.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line' - line to check\n",
    "    \"\"\"\n",
    "    if len(line) != 16:\n",
    "        return False\n",
    "    elif line.strip() == \"##gff-version 3\":\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "def validTabStructure(line):\n",
    "    \"\"\"\n",
    "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line': line to check\n",
    "    \"\"\"\n",
    "    theLine = line.strip().split(\"\\t\")  #strip off extra charachters and split the line on tab\n",
    "\n",
    "    if len(theLine) != 9:               # 9 columns in a gff file, so return False if not exactly 9\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "def validSeqname(text):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
    "    the sequence name on the Gbrowse database and have only valid characters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of entry in column 1 to check\n",
    "    \"\"\"\n",
    "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
    "\n",
    "def validSource(text):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
    "    which only need validity of the characters\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of the entry in column 2 to check\n",
    "    \"\"\"\n",
    "    return not charCheck(text)\n",
    "\n",
    "def validType(text, validTypes=None):\n",
    "    \"\"\"\n",
    "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
    "    For phage this should be one of types ['gene','mRNA','exon']\n",
    "    \n",
    "    Parameters:\n",
    "    - 'text': string value of the entry in column 2 to check\n",
    "    - 'validTypes': List of valid types for checking if not the default 3\n",
    "    \"\"\"\n",
    "    if validTypes is None:\n",
    "        validTypes = {'gene','mRNA','exon', 'contig'}\n",
    "    \n",
    "    return text in validTypes\n",
    "\n",
    "\n",
    "def validScore(score):\n",
    "    \"\"\"\n",
    "    Checks a score (entries in column 6), should be a number or \".\"\n",
    "    \n",
    "    Parameters:\n",
    "    - 'score': score from column 6\n",
    "    \"\"\"\n",
    "    if score == \".\":\n",
    "        return True\n",
    "    \n",
    "    try:\n",
    "        x = float(score)\n",
    "    except:\n",
    "        isScore = False\n",
    "        pass\n",
    "    else:\n",
    "        isScore = True        \n",
    "\n",
    "    return isScore\n",
    "\n",
    "def validStrand(strand):\n",
    "    \"\"\"\n",
    "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
    "    \n",
    "    Parameters:\n",
    "    - 'strand': strand entry from column 7\n",
    "    \"\"\"\n",
    "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
    "    \n",
    "    return strand in validValues\n",
    "\n",
    "def validPhase(phase):\n",
    "    \"\"\"\n",
    "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
    "    \n",
    "    Parameters:\n",
    "    - 'phase': strand entry from column 8\n",
    "    \"\"\"\n",
    "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
    "    \n",
    "    return phase in validValues\n",
    "\n",
    "def validAttributes(attributes):\n",
    "    '''\n",
    "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
    "    It is OK to have spaces in the values entry but not the Key value\n",
    "    \n",
    "    \n",
    "    Parameters:\n",
    "    - 'attributes': entire entry from column 9\n",
    "    '''\n",
    "    print \"here\"\n",
    "    #ok to have a null string\n",
    "    if attributes == '.':\n",
    "        return True\n",
    "    \n",
    "    \n",
    "    #ok go ahead and split into the underlyine key=value items \n",
    "\n",
    "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
    "    \n",
    "    for attrPair in attrPairs:\n",
    "        if attrPair == '':                  #ignore two ;; in a row or end with an extra ;\n",
    "            continue\n",
    "            \n",
    "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
    "            return False\n",
    "        \n",
    "        attrKey, attrValue = attrPair.split(\"=\")\n",
    "        \n",
    "        if attrKey == '':                #must have an entry for a key\n",
    "            return False\n",
    "        \n",
    "        if charCheck(attrKey):          #check key for valid characters\n",
    "            return False\n",
    "        \n",
    "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
    "        if charCheck(attrValue.replace(' ','')):\n",
    "            return False\n",
    "            \n",
    "    return True\n",
    "\n",
    "\n",
    "def charCheck(str, searcher=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
    "    \"\"\"\n",
    "    Checks a string for characters NOT a-zA-Z0-9.=; and returns True if invalid character is\n",
    "    found.\n",
    "\n",
    "    Parameters:\n",
    "    -'str': string being checked.\n",
    "    -'search': default search all characters in valid set\n",
    "    \"\"\"\n",
    "    return bool(searcher(str))\n",
    "\n",
    "def validCoordinate(coord):\n",
    "    \"\"\"\n",
    "    Checks a coordinate (entries in column 4 or 5), should be a positive integer\n",
    "    \n",
    "    Parameters:\n",
    "    - 'coord': coordinate from column 4 or 5\n",
    "    \"\"\"\n",
    "    \n",
    "    return not bool(charCheck(coord, search=re.compile(r'[^0-9]').search))\n",
    "\n",
    "def validCoordinates(leftCoord, rightCoord):\n",
    "    \"\"\"\n",
    "    Checks a coordinates (entries in column 4 and 5), should be a positive integers\n",
    "    and the left Coordinate should be smaller than the right Coordinate\n",
    "    \n",
    "    Parameters:\n",
    "    - 'leftCoord':  coordinate from column 4\n",
    "    - 'rightCoord': coordinate from column 5\n",
    "\n",
    "    \"\"\"\n",
    "    return (validCoordinate(leftCoord) and validCoordinate(rightCoord) and int(leftCoord) <= int(rightCoord))\n",
    "\n",
    "\n",
    "def printFailureMessage(failType):\n",
    "    print \"\\n##### Fatal Error #####\"\n",
    "    print failType + \" failed. No other tests run.\"\n",
    "    print \"\"\"\n",
    "================================ End Results ================================        \n",
    "Computer messages follow below OK to ignore, you should fix file and try again.\n",
    "    \"\"\"\n",
    "    \n",
    "def createDNAMasterFile(gffFileContents):\n",
    "    \"\"\"\n",
    "    Takes in a list of GFF lines, checks for headers and passes lines of type gene to parser.\n",
    "    Returns the complete text of DNA Master text file.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'gffFileContents':  List of text entries, each element a line from GFF3 file.\n",
    "    \"\"\"\n",
    "    \n",
    "    text = ''\n",
    "    for wholeLine in gffFileContents:\n",
    "        line = wholeLine.strip().split(\"\\t\")\n",
    "        \n",
    "        if len(line) < 8:\n",
    "            continue\n",
    "        \n",
    "        if line[2] == 'gene':\n",
    "            entry = gene2CDS(line)\n",
    "            text += entry\n",
    "            \n",
    "        text += parseAttributes(line[8])\n",
    "            \n",
    "    epilog = \"ORIGIN\"\n",
    "    text += epilog\n",
    "        \n",
    "    return text\n",
    "            \n",
    "        \n",
    "def gene2CDS(line):\n",
    "    \"\"\"\n",
    "    Takes in a single GFF entry and returns the corresponding DNA Master entry for the CDS line.\n",
    "    \n",
    "    Parameters:\n",
    "    - 'line':  List of entries from a single GFF entry.\n",
    "    \"\"\"    \n",
    "    if line[6] == \"+\":\n",
    "        return \"CDS \" + line[3] + \" - \" + line[4] + \"\\n\"\n",
    "    else:\n",
    "        return \"CDS complement (\" + line[3] + \" - \" + line[4] + \")\\n\"\n",
    "    \n",
    "def parseAttributes(attributes):\n",
    "    \"\"\"\n",
    "    takes in the Attributes entry of the GFF. If there is an ID= or Name=\n",
    "    entry then create the /gene line. If there is a notes= entry then create\n",
    "    the /note line\n",
    "    \"\"\"\n",
    "    #ok go ahead and split into the underlyine key=value items \n",
    "\n",
    "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
    "    \n",
    "    returnString = \"\"\n",
    "    \n",
    "    for attrPair in attrPairs:\n",
    "        attrKey = \"\"\n",
    "        attrValue = \"\"\n",
    "        \n",
    "        if len(attrPair.split('=')) == 2:\n",
    "            attrKey = attrPair.split(\"=\")[0].lower()\n",
    "            attrValue =  attrPair.split(\"=\")[1]\n",
    "        \n",
    "        if attrKey in [\"id\", \"name\"]:\n",
    "            returnString += '  /gene=' + attrValue + '\\n'\n",
    "        if attrKey == \"note\":\n",
    "            if attrValue[0:1] == attrValue[-1:] == '\"':\n",
    "                attrValue = attrValue[1:-1]\n",
    "            returnString += '    /note=\"' + attrValue + '\"\\n'\n",
    "    \n",
    "    return returnString"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting test_code.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile test_code.py\n",
    "\n",
    "from gffTester_nose import *\n",
    "\n",
    "'''\n",
    "need to check the following functions:\n",
    "\n",
    "def validHeader (line): TEST valid, extra space, extra tab, garbled\n",
    "def validTabStructure(line): test: 7 tabs, 8 tabs, 9 tabs\n",
    "def validSeqname(text): SKIP, just a \"not charCheck\" for now\n",
    "def validSource(text): SKIP, just a \"not charCheck\" for now\n",
    "def validType(text, validTypes=None): TEST: pass with any 3 defaults 'gene','mRNA','exon'; fails on none item, passes with list\n",
    "def validCoordinate(coord): TEST: fails with text, float, negative int; passes with positive int\n",
    "def validScore(score):\n",
    "def validStrand(strand):\n",
    "def validPhase(phase):\n",
    "def validAttributes(attributes):\n",
    "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
    "\n",
    "'''\n",
    "\n",
    "def test_validHeader_1():\n",
    "    'valid header tests as valid'\n",
    "    result = validHeader('##gff-version 3\\n')\n",
    "    assert result == True\n",
    "\n",
    "def test_validHeader_2():\n",
    "    'header with extra space fails'\n",
    "    result = validHeader('##gff-version 3 \\n')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validHeader_3():\n",
    "    'header with extra tab fails'\n",
    "    result = validHeader('##gff-version 3\\t\\n')\n",
    "    assert result == False\n",
    "\n",
    "def test_validHeader_4():\n",
    "    'garbled header fails'\n",
    "    result = validHeader('##gff-Version 3\\n')\n",
    "    assert result == False\n",
    "    \n",
    "\n",
    "def test_validTabStructure_1():\n",
    "    'fails 7 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\n\")\n",
    "    assert result == False\n",
    "\n",
    "def test_validTabStructure_2():\n",
    "    'passes 8 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\n\")\n",
    "    assert result == True\n",
    "    \n",
    "def test_validTabStructure_3():\n",
    "    'fails 9 tab structure'\n",
    "    result = validTabStructure(\"1\\t2\\t3\\t4\\t5\\t6\\t7\\t8\\t9\\t10\\n\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validType_1():\n",
    "    'Type passes with \"gene\"'\n",
    "    result = validType('gene')\n",
    "    assert result == True\n",
    "\n",
    "def test_validType_2():\n",
    "    'Type passes with \"mRNA\"'\n",
    "    result = validType('mRNA')\n",
    "    assert result == True\n",
    "\n",
    "def test_validType_3():\n",
    "    'Type passes with \"exon\"'\n",
    "    result = validType('exon')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validType_4():\n",
    "    'Type fails with mrna'\n",
    "    result = validType('mrna')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validType_5():\n",
    "    'Type passes with specified list'\n",
    "    result = validType('mrna', ['mRNA', 'mrna'])\n",
    "    assert result == True\n",
    "    \n",
    "\n",
    "def test_validCoordinate_1():\n",
    "    'Coord fails with text as coordinate'\n",
    "    result = validCoordinate('a')\n",
    "    assert result == False\n",
    "\n",
    "def test_validCoordinate_2():\n",
    "    'Coord fails with float'\n",
    "    result = validCoordinate('2.0')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validCoordinate_3():\n",
    "    'Coord fails with negative int'\n",
    "    result = validCoordinate('-3')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validCoordinate_4():\n",
    "    'Coord passes with positive integer'\n",
    "    result = validCoordinate('100')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validCoordinate_5():\n",
    "    'Coord fails with digits with space'\n",
    "    result = validCoordinate('1 2')\n",
    "    assert result == False\n",
    "    \n",
    "def test_validCoordinates_1():\n",
    "    'Coordinates fails with left larger than right'\n",
    "    result = validCoordinates('5','4')\n",
    "    assert result == False\n",
    "\n",
    "def test_validCoordinates_2():\n",
    "    'passes with left equal to right'\n",
    "    result = validCoordinates('5','5')\n",
    "    assert result == True\n",
    "    \n",
    "def test_validCoordinates_3():\n",
    "    'passes with left smaller than right'\n",
    "    result = validCoordinates('4','5')\n",
    "    assert result == True\n",
    "\n",
    "    \n",
    "def test_validAttributes_1():\n",
    "    'Attributes passes with null string'\n",
    "    result = validAttributes(\"\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_2():\n",
    "    'Attributes passes a=b'\n",
    "    result = validAttributes(\"a=b\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_3():\n",
    "    'attrubutes passes with \"a=b;\"'\n",
    "    result = validAttributes(\"a=b;\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_4():\n",
    "    'Attributes fails with \"a=bc=d\"'\n",
    "    result = validAttributes(\"a=bc=d\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validAttributes_5():\n",
    "    'attrubutes passes with \"a=b;c=d\"'\n",
    "    result = validAttributes(\"a=b;c=d\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_6():\n",
    "    'attrubutes passes with \"af=b f;c=d\"'\n",
    "    result = validAttributes(\"af=b f;c=d\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_7():\n",
    "    'attrubutes fails with blank key \"a*b=c\"'\n",
    "    result = validAttributes(\"a*b=c\")\n",
    "    assert result == False\n",
    "\n",
    "def test_validAttributes_8():\n",
    "    'attrubutes fails with illegal character in key \"key&=ab\"'\n",
    "    result = validAttributes(\"key&=ab\")\n",
    "    assert result == False\n",
    "\n",
    "def test_validAttributes_9():\n",
    "    'attrubutes fails if missing = \"ab\"'\n",
    "    result = validAttributes(\"ab\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validAttributes_10():\n",
    "    'attrubutes fails if missing = in 2nd \"a=b;ab\"'\n",
    "    result = validAttributes(\"a=b;ab\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_validAttributes_11():\n",
    "    'attrubutes passes with 3 pairs \"a=b;c=d;e=f\"'\n",
    "    result = validAttributes(\"a=b;c=d;e=f\")\n",
    "    assert result == True\n",
    "\n",
    "def test_validAttributes_12():\n",
    "    'attrubutes fails error in 3rd pair \"a=b;c=d;e==f\"'\n",
    "    result = validAttributes(\"a=b;c=d;e==f\")\n",
    "    assert result == False\n",
    "\n",
    "def test_charCheck_1():\n",
    "    'charCheck give true for illegal & \"a&f\"'\n",
    "    result = charCheck(\"a&f\")\n",
    "    assert result == True\n",
    "\n",
    "def test_charCheck_2():\n",
    "    'charCheck true for illegal space \"a f\"'\n",
    "    result = charCheck(\"a f\")\n",
    "    assert result == True\n",
    "    \n",
    "def test_charCheck_3():\n",
    "    'charCheck gives false for simple text \"abcdefghijklmnopqrstuvwxyz\"'\n",
    "    result = charCheck(\"abcdefghijklmnopqrstuvwxyz\")\n",
    "    assert result == False\n",
    "\n",
    "def test_charCheck_4():\n",
    "    'charCheck gives false for simple text \"ABCDEFGHIJKLMNOPQRSTUVWXYZ\"'\n",
    "    result = charCheck(\"ABCDEFGHIJKLMNOPQRSTUVWXYZ\")\n",
    "    assert result == False   \n",
    "    \n",
    "def test_charCheck_5():\n",
    "    'charCheck gives false for simple text \"1234567890\"'\n",
    "    result = charCheck(\"1234567890\")\n",
    "    assert result == False\n",
    "    \n",
    "def test_parseAttributes_1():\n",
    "    'Attribute test 1 single ID entry'\n",
    "    returned = parseAttributes(\"ID=gene1\")\n",
    "    assert returned == '  /gene=gene1\\n'\n",
    "\n",
    "def test_parseAttributes_2():\n",
    "    'Attribute test 2 single ID entry with semicolon'\n",
    "    returned = parseAttributes(\"ID=gene1;\")\n",
    "    assert returned == '  /gene=gene1\\n'\n",
    "    \n",
    "def test_parseAttributes_3():\n",
    "    'Attribute test 3 single Note entry with spaces'\n",
    "    returned = parseAttributes(\"Note=note note\")\n",
    "    assert returned == '    /note=\"note note\"\\n'\n",
    "    \n",
    "def test_parseAttributes_4():\n",
    "    'Attribute test 4 single Note entry with semicolon'\n",
    "    returned = parseAttributes(\"Note=gene;\")\n",
    "    assert returned == '    /note=\"gene\"\\n'\n",
    "    \n",
    "def test_parseAttributes_5():\n",
    "    'Attribute test 5 single Note entry preceded by ;'\n",
    "    returned = parseAttributes(\";Note=gene2\")\n",
    "    assert returned == '    /note=\"gene2\"\\n'\n",
    "    \n",
    "def test_parseAttributes_6():\n",
    "    'Attribute test 6 ID + Note entry'\n",
    "    returned = parseAttributes(\"ID=gene1;Note=note4\")\n",
    "    assert returned == '  /gene=gene1\\n    /note=\"note4\"\\n'\n",
    "    \n",
    "def test_parseAttributes_7():\n",
    "    'Attribute test 7 single Note entry with quotes'\n",
    "    returned = parseAttributes('Note=\"gene7\"')\n",
    "    assert returned == '    /note=\"gene7\"\\n'\n",
    "    \n",
    "def test_parseAttributes_8():\n",
    "    'Attribute test 8 single id entry with quotes'\n",
    "    returned = parseAttributes('id=\"gene8\"')\n",
    "    assert returned == '  /gene=\"gene8\"\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "valid header tests as valid ... ok\r\n",
      "header with extra space fails ... ok\r\n",
      "header with extra tab fails ... ok\r\n",
      "garbled header fails ... ok\r\n",
      "fails 7 tab structure ... ok\r\n",
      "passes 8 tab structure ... ok\r\n",
      "fails 9 tab structure ... ok\r\n",
      "Type passes with \"gene\" ... ok\r\n",
      "Type passes with \"mRNA\" ... ok\r\n",
      "Type passes with \"exon\" ... ok\r\n",
      "Type fails with mrna ... ok\r\n",
      "Type passes with specified list ... ok\r\n",
      "Coord fails with text as coordinate ... ok\r\n",
      "Coord fails with float ... ok\r\n",
      "Coord fails with negative int ... ok\r\n",
      "Coord passes with positive integer ... ok\r\n",
      "Coord fails with digits with space ... ok\r\n",
      "Coordinates fails with left larger than right ... ok\r\n",
      "passes with left equal to right ... ok\r\n",
      "passes with left smaller than right ... ok\r\n",
      "Attributes passes with null string ... ok\r\n",
      "Attributes passes a=b ... ok\r\n",
      "attrubutes passes with \"a=b;\" ... ok\r\n",
      "Attributes fails with \"a=bc=d\" ... ok\r\n",
      "attrubutes passes with \"a=b;c=d\" ... ok\r\n",
      "attrubutes passes with \"af=b f;c=d\" ... ok\r\n",
      "attrubutes fails with blank key \"a*b=c\" ... ok\r\n",
      "attrubutes fails with illegal character in key \"key&=ab\" ... ok\r\n",
      "attrubutes fails if missing = \"ab\" ... ok\r\n",
      "attrubutes fails if missing = in 2nd \"a=b;ab\" ... ok\r\n",
      "attrubutes passes with 3 pairs \"a=b;c=d;e=f\" ... ok\r\n",
      "attrubutes fails error in 3rd pair \"a=b;c=d;e==f\" ... ok\r\n",
      "charCheck give true for illegal & \"a&f\" ... ok\r\n",
      "charCheck true for illegal space \"a f\" ... ok\r\n",
      "charCheck gives false for simple text \"abcdefghijklmnopqrstuvwxyz\" ... ok\r\n",
      "charCheck gives false for simple text \"ABCDEFGHIJKLMNOPQRSTUVWXYZ\" ... ok\r\n",
      "charCheck gives false for simple text \"1234567890\" ... ok\r\n",
      "Attribute test 1 single ID entry ... ok\r\n",
      "Attribute test 2 single ID entry with semicolon ... ok\r\n",
      "Attribute test 3 single Note entry with spaces ... ok\r\n",
      "Attribute test 4 single Note entry with semicolon ... ok\r\n",
      "Attribute test 5 single Note entry preceded by ; ... ok\r\n",
      "Attribute test 6 ID + Note entry ... ok\r\n",
      "Attribute test 7 single Note entry with quotes ... ok\r\n",
      "Attribute test 8 single id entry with quotes ... ok\r\n",
      "\r\n",
      "----------------------------------------------------------------------\r\n",
      "Ran 45 tests in 0.007s\r\n",
      "\r\n",
      "OK\r\n"
     ]
    }
   ],
   "source": [
    "!nosetests -v test_code.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print \"grene\" in [\"gene\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "a = '\"jkLmn\"'\n",
    "b = '3'\n",
    "c = '\"'\n",
    "print a[1:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "charCheck\t createDNAMasterFile\t gene2CDS\t parseAttributes\t printFailureMessage\t re\t sys\t validAttributes\t validCoordinate\t \n",
      "validCoordinates\t validHeader\t validPhase\t validScore\t validSeqname\t validSource\t validStrand\t validTabStructure\t validType\t \n",
      "\n"
     ]
    }
   ],
   "source": [
    "%who"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "b\n"
     ]
    }
   ],
   "source": [
    "yin, yang = \"=b\".split('=')\n",
    "print len(yin)\n",
    "print yang"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<type 'NoneType'>\n",
      "false\n"
     ]
    }
   ],
   "source": [
    "info['pham']= None\n",
    "print type(info['pham'])\n",
    "if info['pham']:\n",
    "    print \"true\"\n",
    "else:\n",
    "    print \"false\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable              Type        Data/Info\n",
      "-------------------------------------------\n",
      "charCheck             function    <function charCheck at 0x1035257d0>\n",
      "createDNAMasterFile   function    <function createDNAMasterFile at 0x1035259b0>\n",
      "gene2CDS              function    <function gene2CDS at 0x103525a28>\n",
      "parseAttributes       function    <function parseAttributes at 0x103525aa0>\n",
      "printFailureMessage   function    <function printFailureMessage at 0x103525938>\n",
      "re                    module      <module 're' from '/Appli<...>ts/lib/python2.7/re.pyc'>\n",
      "sys                   module      <module 'sys' (built-in)>\n",
      "validAttributes       function    <function validAttributes at 0x103525758>\n",
      "validCoordinate       function    <function validCoordinate at 0x103525848>\n",
      "validCoordinates      function    <function validCoordinates at 0x1035258c0>\n",
      "validHeader           function    <function validHeader at 0x102de96e0>\n",
      "validPhase            function    <function validPhase at 0x103525668>\n",
      "validScore            function    <function validScore at 0x103525488>\n",
      "validSeqname          function    <function validSeqname at 0x103512ed8>\n",
      "validSource           function    <function validSource at 0x103512f50>\n",
      "validStrand           function    <function validStrand at 0x103525578>\n",
      "validTabStructure     function    <function validTabStructure at 0x103512e60>\n",
      "validType             function    <function validType at 0x103525500>\n"
     ]
    }
   ],
   "source": [
    "whos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "charCheck(\"&\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
