{
 "metadata": {
  "name": "",
  "signature": "sha256:75a3998384421deefee8ae01de0141a115a5dd371a14d975a668a2b0d5962262"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "GFF phage file validation program"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###README:\n",
      "\n",
      "This program will check your GFF file. It checks for proper format and use of proper characters (format checks). If you give the program the fasta file with the DNA sequence of your phage, the program will also check that the \"gene\" entries start with a proper start codon and have only a single stop codon as the last codon (biology checks).\n",
      "\n",
      "**Before you run the program:** copy both your phage fasta file and your GFF file into the \"input\" folder on the Desktop. Make sure you know the EXACT name of the files, it is best to avoid spaces in filenames (rename your files if necessary).  \n",
      "\n",
      "**To run the program:** click anywhere in the code box below near the \"#Click in Here\" so that you see the curser, hit shift-return (i.e. hold shift while hitting the return key).\n",
      "\n",
      "**What will happen:** the first thing to appear will be a small box asking for the name of your gff file. Type the name of the gff file into the box. Once the box has the correct name, hit the return key (no shift). You will then see another box asking for the name of your fasta file. If you only want format checks, make sure the cursor is in the box, leave the box blank and hit enter. If you want the program to also run biology checks, type the name of the fasta file, then hit enter.\n",
      "\n",
      "Your results should appear directly below the two question boxes."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Click in Here\n",
      "\n",
      "fastaFileName = raw_input('Enter the name of the fasta file (BE EXACT): ')\n",
      "gffFileName = raw_input('Enter the name of the gff file (BE EXACT): ')\n",
      "\n",
      "fullFastaFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + fastaFileName\n",
      "fullGFFFileName = \"/Users/chris/Documents/code/python/gff_validator/docs/\" + gffFileName\n",
      "\n",
      "try:\n",
      "    with open(fullFastaFileName) as fastaFile:\n",
      "        fastaFileContents = fastaFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get fasta file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "try:\n",
      "    with open(fullGFFFileName) as gffFile:\n",
      "        gffFileContents = gffFile.readlines()\n",
      "except:\n",
      "    print \"Well that didn't work! Could not get GFF file. Double check spelling, no spaces in file name, try again.\"\n",
      "\n",
      "print \"\\n\\n================================ Results ================================\\n\"\n",
      "        \n",
      "firstline = gffFileContents[0]\n",
      "\n",
      "if not validHeader(firstline):\n",
      "    print \"##### Fatal Error #####\"\n",
      "    print \"\"\"\n",
      "    GFF header line failed. No other tests run. \n",
      "    Check 1st line, it should be EXACTLY\n",
      "    ##gff-version 3\n",
      "    Most Common error: spaces and/or tabs at end of line after 3.\n",
      "    \n",
      "Computer messages follow, ignore and try again:\n",
      "    \"\"\"\n",
      "\n",
      "    sys.exit()\n",
      "\n",
      "for lineNumber,line in enumerate(gffFileContents[1:]):\n",
      "    print lineNumber, line"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Enter the name of the fasta file (BE EXACT): a.fasta\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Enter the name of the gff file (BE EXACT): a.gff3\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "\n",
        "================================ Results ================================\n",
        "\n",
        "0 Phabio_draft\tGroup\tgene\t43\t371\t200\t+\t.\tID=Phabio.1;Name=Phabio.1\n",
        "\n",
        "1 Phabio_draft\tGroup\tmRNA\t43\t371\t200\t+\t.\tID=Phabio.1.mRNA;Parent=Phabio.1\n",
        "\n",
        "2 Phabio_draft\tGroup\texon\t43\t371\t200\t+\t.\tID=Phabio.1.exon;Parent=Phabio.1.mRNA\n",
        "\n",
        "3 Phabio_draft\tGroup\tgene\t445\t699\t200\t+\t.\tID=Phabio.2;Name=Phabio.2\n",
        "\n",
        "4 Phabio_draft\tGroup\tmRNA\t445\t699\t200\t+\t.\tID=Phabio.2.mRNA;Parent=Phabio.2\n",
        "\n",
        "5 Phabio_draft\tGroup\texon\t445\t699\t200\t+\t.\tID=Phabio.2.exon;Parent=Phabio.2.mRNA\n",
        "\n",
        "6 Phabio_draft\tGroup\tgene\t671\t2158\t200\t+\t.\tID=Phabio.3;Name=Phabio.3\n",
        "\n",
        "7 Phabio_draft\tGroup\tmRNA\t671\t2158\t200\t+\t.\tID=Phabio.3.mRNA;Parent=Phabio.3\n",
        "\n",
        "8 Phabio_draft\tGroup\texon\t671\t2158\t200\t+\t.\tID=Phabio.3.exon;Parent=Phabio.3.mRNA\n",
        "\n",
        "9 Phabio_draft\tGroup\tgene\t2199\t3620\t200\t+\t.\tID=Phabio.4;Name=Phabio.4\n",
        "\n",
        "10 Phabio_draft\tGroup\tmRNA\t2199\t3620\t200\t+\t.\tID=Phabio.4.mRNA;Parent=Phabio.4\n",
        "\n",
        "11 Phabio_draft\tGroup\texon\t2199\t3620\t200\t+\t.\tID=Phabio.4.exon;Parent=Phabio.4.mRNA\n",
        "\n",
        "12 Phabio_draft\tGroup\tgene\t3604\t4329\t200\t+\t.\tID=Phabio.5;Name=Phabio.5\n",
        "\n",
        "13 Phabio_draft\tGroup\tmRNA\t3604\t4329\t200\t+\t.\tID=Phabio.5.mRNA;Parent=Phabio.5\n",
        "\n",
        "14 Phabio_draft\tGroup\texon\t3604\t4329\t200\t+\t.\tID=Phabio.5.exon;Parent=Phabio.5.mRNA\n",
        "\n",
        "15 Phabio_draft\tGroup\tgene\t4408\t4944\t200\t+\t.\tID=Phabio.6;Name=Phabio.6\n",
        "\n",
        "16 Phabio_draft\tGroup\tmRNA\t4408\t4944\t200\t+\t.\tID=Phabio.6.mRNA;Parent=Phabio.6\n",
        "\n",
        "17 Phabio_draft\tGroup\texon\t4408\t4944\t200\t+\t.\tID=Phabio.6.exon;Parent=Phabio.6.mRNA\n",
        "\n",
        "18 Phabio_draft\tGroup\tgene\t5007\t5915\t200\t+\t.\tID=Phabio.7;Name=Phabio.7\n",
        "\n",
        "19 Phabio_draft\tGroup\tmRNA\t5007\t5915\t200\t+\t.\tID=Phabio.7.mRNA;Parent=Phabio.7\n",
        "\n",
        "20 Phabio_draft\tGroup\texon\t5007\t5915\t200\t+\t.\tID=Phabio.7.exon;Parent=Phabio.7.mRNA\n",
        "\n",
        "21 Phabio_draft\tGroup\tgene\t5928\t6299\t200\t+\t.\tID=Phabio.8;Name=Phabio.8\n",
        "\n",
        "22 Phabio_draft\tGroup\tmRNA\t5928\t6299\t200\t+\t.\tID=Phabio.8.mRNA;Parent=Phabio.8\n",
        "\n",
        "23 Phabio_draft\tGroup\texon\t5928\t6299\t200\t+\t.\tID=Phabio.8.exon;Parent=Phabio.8.mRNA\n",
        "\n",
        "24 Phabio_draft\tGroup\tgene\t6296\t6625\t200\t+\t.\tID=Phabio.9;Name=Phabio.9\n",
        "\n",
        "25 Phabio_draft\tGroup\tmRNA\t6296\t6625\t200\t+\t.\tID=Phabio.9.mRNA;Parent=Phabio.9\n",
        "\n",
        "26 Phabio_draft\tGroup\texon\t6296\t6625\t200\t+\t.\tID=Phabio.9.exon;Parent=Phabio.9.mRNA\n",
        "\n",
        "27 Phabio_draft\tGroup\tgene\t6632\t6958\t200\t+\t.\tID=Phabio.10;Name=Phabio.10\n",
        "\n",
        "28 Phabio_draft\tGroup\tmRNA\t6632\t6958\t200\t+\t.\tID=Phabio.10.mRNA;Parent=Phabio.10\n",
        "\n",
        "29 Phabio_draft\tGroup\texon\t6632\t6958\t200\t+\t.\tID=Phabio.10.exon;Parent=Phabio.10.mRNA\n",
        "\n",
        "30 Phabio_draft\tGroup\tgene\t6948\t7352\t200\t+\t.\tID=Phabio.11;Name=Phabio.11\n",
        "\n",
        "31 Phabio_draft\tGroup\tmRNA\t6948\t7352\t200\t+\t.\tID=Phabio.11.mRNA;Parent=Phabio.11\n",
        "\n",
        "32 Phabio_draft\tGroup\texon\t6948\t7352\t200\t+\t.\tID=Phabio.11.exon;Parent=Phabio.11.mRNA\n",
        "\n",
        "33 Phabio_draft\tGroup\tgene\t7462\t8271\t200\t+\t.\tID=Phabio.12;Name=Phabio.12\n",
        "\n",
        "34 Phabio_draft\tGroup\tmRNA\t7462\t8271\t200\t+\t.\tID=Phabio.12.mRNA;Parent=Phabio.12\n",
        "\n",
        "35 Phabio_draft\tGroup\texon\t7462\t8271\t200\t+\t.\tID=Phabio.12.exon;Parent=Phabio.12.mRNA\n",
        "\n",
        "36 Phabio_draft\tGroup\tgene\t8385\t9321\t200\t+\t.\tID=Phabio.13;Name=Phabio.13\n",
        "\n",
        "37 Phabio_draft\tGroup\tmRNA\t8385\t9321\t200\t+\t.\tID=Phabio.13.mRNA;Parent=Phabio.13\n",
        "\n",
        "38 Phabio_draft\tGroup\texon\t8385\t9321\t200\t+\t.\tID=Phabio.13.exon;Parent=Phabio.13.mRNA\n",
        "\n",
        "39 Phabio_draft\tGroup\tgene\t9340\t12866\t200\t+\t.\tID=Phabio.14;Name=Phabio.14\n",
        "\n",
        "40 Phabio_draft\tGroup\tmRNA\t9340\t12866\t200\t+\t.\tID=Phabio.14.mRNA;Parent=Phabio.14\n",
        "\n",
        "41 Phabio_draft\tGroup\texon\t9340\t12866\t200\t+\t.\tID=Phabio.14.exon;Parent=Phabio.14.mRNA\n",
        "\n",
        "42 Phabio_draft\tGroup\tgene\t12868\t14577\t200\t+\t.\tID=Phabio.15;Name=Phabio.15\n",
        "\n",
        "43 Phabio_draft\tGroup\tmRNA\t12868\t14577\t200\t+\t.\tID=Phabio.15.mRNA;Parent=Phabio.15\n",
        "\n",
        "44 Phabio_draft\tGroup\texon\t12868\t14577\t200\t+\t.\tID=Phabio.15.exon;Parent=Phabio.15.mRNA\n",
        "\n",
        "45 Phabio_draft\tGroup\tgene\t14663\t16372\t200\t+\t.\tID=Phabio.16;Name=Phabio.16\n",
        "\n",
        "46 Phabio_draft\tGroup\tmRNA\t14663\t16372\t200\t+\t.\tID=Phabio.16.mRNA;Parent=Phabio.16\n",
        "\n",
        "47 Phabio_draft\tGroup\texon\t14663\t16372\t200\t+\t.\tID=Phabio.16.exon;Parent=Phabio.16.mRNA\n",
        "\n",
        "48 Phabio_draft\tGroup\tgene\t16409\t17257\t200\t+\t.\tID=Phabio.17;Name=Phabio.17\n",
        "\n",
        "49 Phabio_draft\tGroup\tmRNA\t16409\t17257\t200\t+\t.\tID=Phabio.17.mRNA;Parent=Phabio.17\n",
        "\n",
        "50 Phabio_draft\tGroup\texon\t16409\t17257\t200\t+\t.\tID=Phabio.17.exon;Parent=Phabio.17.mRNA\n",
        "\n",
        "51 Phabio_draft\tGroup\tgene\t17253\t19791\t200\t+\t.\tID=Phabio.18;Name=Phabio.18\n",
        "\n",
        "52 Phabio_draft\tGroup\tmRNA\t17253\t19791\t200\t+\t.\tID=Phabio.18.mRNA;Parent=Phabio.18\n",
        "\n",
        "53 Phabio_draft\tGroup\texon\t17253\t19791\t200\t+\t.\tID=Phabio.18.exon;Parent=Phabio.18.mRNA\n",
        "\n",
        "54 Phabio_draft\tGroup\tgene\t19788\t21674\t200\t+\t.\tID=Phabio.19;Name=Phabio.19\n",
        "\n",
        "55 Phabio_draft\tGroup\tmRNA\t19788\t21674\t200\t+\t.\tID=Phabio.19.mRNA;Parent=Phabio.19\n",
        "\n",
        "56 Phabio_draft\tGroup\texon\t19788\t21674\t200\t+\t.\tID=Phabio.19.exon;Parent=Phabio.19.mRNA\n",
        "\n",
        "57 Phabio_draft\tGroup\tgene\t21677\t22837\t200\t+\t.\tID=Phabio.20;Name=Phabio.20\n",
        "\n",
        "58 Phabio_draft\tGroup\tmRNA\t21677\t22837\t200\t+\t.\tID=Phabio.20.mRNA;Parent=Phabio.20\n",
        "\n",
        "59 Phabio_draft\tGroup\texon\t21677\t22837\t200\t+\t.\tID=Phabio.20.exon;Parent=Phabio.20.mRNA\n",
        "\n",
        "60 Phabio_draft\tGroup\tgene\t22855\t23196\t200\t+\t.\tID=Phabio.21;Name=Phabio.21\n",
        "\n",
        "61 Phabio_draft\tGroup\tmRNA\t22855\t23196\t200\t+\t.\tID=Phabio.21.mRNA;Parent=Phabio.21\n",
        "\n",
        "62 Phabio_draft\tGroup\texon\t22855\t23196\t200\t+\t.\tID=Phabio.21.exon;Parent=Phabio.21.mRNA\n",
        "\n",
        "63 Phabio_draft\tGroup\tgene\t23219\t23473\t200\t+\t.\tID=Phabio.22;Name=Phabio.22\n",
        "\n",
        "64 Phabio_draft\tGroup\tmRNA\t23219\t23473\t200\t+\t.\tID=Phabio.22.mRNA;Parent=Phabio.22\n",
        "\n",
        "65 Phabio_draft\tGroup\texon\t23219\t23473\t200\t+\t.\tID=Phabio.22.exon;Parent=Phabio.22.mRNA\n",
        "\n",
        "66 Phabio_draft\tGroup\tgene\t23487\t24125\t200\t+\t.\tID=Phabio.23;Name=Phabio.23\n",
        "\n",
        "67 Phabio_draft\tGroup\tmRNA\t23487\t24125\t200\t+\t.\tID=Phabio.23.mRNA;Parent=Phabio.23\n",
        "\n",
        "68 Phabio_draft\tGroup\texon\t23487\t24125\t200\t+\t.\tID=Phabio.23.exon;Parent=Phabio.23.mRNA\n",
        "\n",
        "69 Phabio_draft\tGroup\tgene\t24122\t24946\t200\t+\t.\tID=Phabio.24;Name=Phabio.24\n",
        "\n",
        "70 Phabio_draft\tGroup\tmRNA\t24122\t24946\t200\t+\t.\tID=Phabio.24.mRNA;Parent=Phabio.24\n",
        "\n",
        "71 Phabio_draft\tGroup\texon\t24122\t24946\t200\t+\t.\tID=Phabio.24.exon;Parent=Phabio.24.mRNA\n",
        "\n",
        "72 Phabio_draft\tGroup\tgene\t24951\t25148\t200\t-\t.\tID=Phabio.25;Name=Phabio.25\n",
        "\n",
        "73 Phabio_draft\tGroup\tmRNA\t24951\t25148\t200\t-\t.\tID=Phabio.25.mRNA;Parent=Phabio.25\n",
        "\n",
        "74 Phabio_draft\tGroup\texon\t24951\t25148\t200\t-\t.\tID=Phabio.25.exon;Parent=Phabio.25.mRNA\n",
        "\n",
        "75 Phabio_draft\tGroup\tgene\t25296\t25580\t200\t+\t.\tID=Phabio.26;Name=Phabio.26\n",
        "\n",
        "76 Phabio_draft\tGroup\tmRNA\t25296\t25580\t200\t+\t.\tID=Phabio.26.mRNA;Parent=Phabio.26\n",
        "\n",
        "77 Phabio_draft\tGroup\texon\t25296\t25580\t200\t+\t.\tID=Phabio.26.exon;Parent=Phabio.26.mRNA\n",
        "\n",
        "78 Phabio_draft\tGroup\tgene\t25730\t25897\t200\t+\t.\tID=Phabio.27;Name=Phabio.27\n",
        "\n",
        "79 Phabio_draft\tGroup\tmRNA\t25730\t25897\t200\t+\t.\tID=Phabio.27.mRNA;Parent=Phabio.27\n",
        "\n",
        "80 Phabio_draft\tGroup\texon\t25730\t25897\t200\t+\t.\tID=Phabio.27.exon;Parent=Phabio.27.mRNA\n",
        "\n",
        "81 Phabio_draft\tGroup\tgene\t26201\t27394\t200\t+\t.\tID=Phabio.28;Name=Phabio.28\n",
        "\n",
        "82 Phabio_draft\tGroup\tmRNA\t26201\t27394\t200\t+\t.\tID=Phabio.28.mRNA;Parent=Phabio.28\n",
        "\n",
        "83 Phabio_draft\tGroup\texon\t26201\t27394\t200\t+\t.\tID=Phabio.28.exon;Parent=Phabio.28.mRNA\n",
        "\n",
        "84 Phabio_draft\tGroup\tgene\t27394\t28395\t200\t+\t.\tID=Phabio.29;Name=Phabio.29\n",
        "\n",
        "85 Phabio_draft\tGroup\tmRNA\t27394\t28395\t200\t+\t.\tID=Phabio.29.mRNA;Parent=Phabio.29\n",
        "\n",
        "86 Phabio_draft\tGroup\texon\t27394\t28395\t200\t+\t.\tID=Phabio.29.exon;Parent=Phabio.29.mRNA\n",
        "\n",
        "87 Phabio_draft\tGroup\tgene\t28409\t28642\t200\t+\t.\tID=Phabio.30;Name=Phabio.30\n",
        "\n",
        "88 Phabio_draft\tGroup\tmRNA\t28409\t28642\t200\t+\t.\tID=Phabio.30.mRNA;Parent=Phabio.30\n",
        "\n",
        "89 Phabio_draft\tGroup\texon\t28409\t28642\t200\t+\t.\tID=Phabio.30.exon;Parent=Phabio.30.mRNA\n",
        "\n",
        "90 Phabio_draft\tGroup\tgene\t28639\t29013\t200\t+\t.\tID=Phabio.31;Name=Phabio.31\n",
        "\n",
        "91 Phabio_draft\tGroup\tmRNA\t28639\t29013\t200\t+\t.\tID=Phabio.31.mRNA;Parent=Phabio.31\n",
        "\n",
        "92 Phabio_draft\tGroup\texon\t28639\t29013\t200\t+\t.\tID=Phabio.31.exon;Parent=Phabio.31.mRNA\n",
        "\n",
        "93 Phabio_draft\tGroup\tgene\t29038\t29271\t200\t+\t.\tID=Phabio.32;Name=Phabio.32\n",
        "\n",
        "94 Phabio_draft\tGroup\tmRNA\t29038\t29271\t200\t+\t.\tID=Phabio.32.mRNA;Parent=Phabio.32\n",
        "\n",
        "95 Phabio_draft\tGroup\texon\t29038\t29271\t200\t+\t.\tID=Phabio.32.exon;Parent=Phabio.32.mRNA\n",
        "\n",
        "96 Phabio_draft\tGroup\tgene\t29258\t30079\t200\t+\t.\tID=Phabio.33;Name=Phabio.33\n",
        "\n",
        "97 Phabio_draft\tGroup\tmRNA\t29258\t30079\t200\t+\t.\tID=Phabio.33.mRNA;Parent=Phabio.33\n",
        "\n",
        "98 Phabio_draft\tGroup\texon\t29258\t30079\t200\t+\t.\tID=Phabio.33.exon;Parent=Phabio.33.mRNA\n",
        "\n",
        "99 Phabio_draft\tGroup\tgene\t30097\t30387\t200\t+\t.\tID=Phabio.34;Name=Phabio.34\n",
        "\n",
        "100 Phabio_draft\tGroup\tmRNA\t30097\t30387\t200\t+\t.\tID=Phabio.34.mRNA;Parent=Phabio.34\n",
        "\n",
        "101 Phabio_draft\tGroup\texon\t30097\t30387\t200\t+\t.\tID=Phabio.34.exon;Parent=Phabio.34.mRNA\n",
        "\n",
        "102 Phabio_draft\tGroup\tgene\t30380\t30472\t200\t+\t.\tID=Phabio.35;Name=Phabio.35\n",
        "\n",
        "103 Phabio_draft\tGroup\tmRNA\t30380\t30472\t200\t+\t.\tID=Phabio.35.mRNA;Parent=Phabio.35\n",
        "\n",
        "104 Phabio_draft\tGroup\texon\t30380\t30472\t200\t+\t.\tID=Phabio.35.exon;Parent=Phabio.35.mRNA\n",
        "\n",
        "105 Phabio_draft\tGroup\tgene\t31317\t32618\t200\t+\t.\tID=Phabio.36;Name=Phabio.36\n",
        "\n",
        "106 Phabio_draft\tGroup\tmRNA\t31317\t32618\t200\t+\t.\tID=Phabio.36.mRNA;Parent=Phabio.36\n",
        "\n",
        "107 Phabio_draft\tGroup\texon\t31317\t32618\t200\t+\t.\tID=Phabio.36.exon;Parent=Phabio.36.mRNA\n",
        "\n",
        "108 Phabio_draft\tGroup\tgene\t32700\t33152\t200\t-\t.\tID=Phabio.37;Name=Phabio.37\n",
        "\n",
        "109 Phabio_draft\tGroup\tmRNA\t32700\t33152\t200\t-\t.\tID=Phabio.37.mRNA;Parent=Phabio.37\n",
        "\n",
        "110 Phabio_draft\tGroup\texon\t32700\t33152\t200\t-\t.\tID=Phabio.37.exon;Parent=Phabio.37.mRNA\n",
        "\n",
        "111 Phabio_draft\tGroup\tgene\t33267\t33413\t200\t-\t.\tID=Phabio.38;Name=Phabio.38\n",
        "\n",
        "112 Phabio_draft\tGroup\tmRNA\t33267\t33413\t200\t-\t.\tID=Phabio.38.mRNA;Parent=Phabio.38\n",
        "\n",
        "113 Phabio_draft\tGroup\texon\t33267\t33413\t200\t-\t.\tID=Phabio.38.exon;Parent=Phabio.38.mRNA\n",
        "\n",
        "114 Phabio_draft\tGroup\tgene\t33416\t33949\t200\t-\t.\tID=Phabio.39;Name=Phabio.39\n",
        "\n",
        "115 Phabio_draft\tGroup\tmRNA\t33416\t33949\t200\t-\t.\tID=Phabio.39.mRNA;Parent=Phabio.39\n",
        "\n",
        "116 Phabio_draft\tGroup\texon\t33416\t33949\t200\t-\t.\tID=Phabio.39.exon;Parent=Phabio.39.mRNA\n",
        "\n",
        "117 Phabio_draft\tGroup\tgene\t34117\t34440\t200\t+\t.\tID=Phabio.40;Name=Phabio.40\n",
        "\n",
        "118 Phabio_draft\tGroup\tmRNA\t34117\t34440\t200\t+\t.\tID=Phabio.40.mRNA;Parent=Phabio.40\n",
        "\n",
        "119 Phabio_draft\tGroup\texon\t34117\t34440\t200\t+\t.\tID=Phabio.40.exon;Parent=Phabio.40.mRNA\n",
        "\n",
        "120 Phabio_draft\tGroup\tgene\t34501\t35367\t200\t+\t.\tID=Phabio.41;Name=Phabio.41\n",
        "\n",
        "121 Phabio_draft\tGroup\tmRNA\t34501\t35367\t200\t+\t.\tID=Phabio.41.mRNA;Parent=Phabio.41\n",
        "\n",
        "122 Phabio_draft\tGroup\texon\t34501\t35367\t200\t+\t.\tID=Phabio.41.exon;Parent=Phabio.41.mRNA\n",
        "\n",
        "123 Phabio_draft\tGroup\tgene\t35364\t35663\t200\t+\t.\tID=Phabio.42;Name=Phabio.42\n",
        "\n",
        "124 Phabio_draft\tGroup\tmRNA\t35364\t35663\t200\t+\t.\tID=Phabio.42.mRNA;Parent=Phabio.42\n",
        "\n",
        "125 Phabio_draft\tGroup\texon\t35364\t35663\t200\t+\t.\tID=Phabio.42.exon;Parent=Phabio.42.mRNA\n",
        "\n",
        "126 Phabio_draft\tGroup\tgene\t35660\t35875\t200\t+\t.\tID=Phabio.43;Name=Phabio.43\n",
        "\n",
        "127 Phabio_draft\tGroup\tmRNA\t35660\t35875\t200\t+\t.\tID=Phabio.43.mRNA;Parent=Phabio.43\n",
        "\n",
        "128 Phabio_draft\tGroup\texon\t35660\t35875\t200\t+\t.\tID=Phabio.43.exon;Parent=Phabio.43.mRNA\n",
        "\n",
        "129 Phabio_draft\tGroup\tgene\t35962\t36159\t200\t+\t.\tID=Phabio.44;Name=Phabio.44\n",
        "\n",
        "130 Phabio_draft\tGroup\tmRNA\t35962\t36159\t200\t+\t.\tID=Phabio.44.mRNA;Parent=Phabio.44\n",
        "\n",
        "131 Phabio_draft\tGroup\texon\t35962\t36159\t200\t+\t.\tID=Phabio.44.exon;Parent=Phabio.44.mRNA\n",
        "\n",
        "132 Phabio_draft\tGroup\tgene\t36156\t36347\t200\t+\t.\tID=Phabio.45;Name=Phabio.45\n",
        "\n",
        "133 Phabio_draft\tGroup\tmRNA\t36156\t36347\t200\t+\t.\tID=Phabio.45.mRNA;Parent=Phabio.45\n",
        "\n",
        "134 Phabio_draft\tGroup\texon\t36156\t36347\t200\t+\t.\tID=Phabio.45.exon;Parent=Phabio.45.mRNA\n",
        "\n",
        "135 Phabio_draft\tGroup\tgene\t36344\t36991\t200\t+\t.\tID=Phabio.46;Name=Phabio.46\n",
        "\n",
        "136 Phabio_draft\tGroup\tmRNA\t36344\t36991\t200\t+\t.\tID=Phabio.46.mRNA;Parent=Phabio.46\n",
        "\n",
        "137 Phabio_draft\tGroup\texon\t36344\t36991\t200\t+\t.\tID=Phabio.46.exon;Parent=Phabio.46.mRNA\n",
        "\n",
        "138 Phabio_draft\tGroup\tgene\t37009\t37290\t200\t+\t.\tID=Phabio.47;Name=Phabio.47\n",
        "\n",
        "139 Phabio_draft\tGroup\tmRNA\t37009\t37290\t200\t+\t.\tID=Phabio.47.mRNA;Parent=Phabio.47\n",
        "\n",
        "140 Phabio_draft\tGroup\texon\t37009\t37290\t200\t+\t.\tID=Phabio.47.exon;Parent=Phabio.47.mRNA\n",
        "\n",
        "141 Phabio_draft\tGroup\tgene\t37365\t37643\t200\t+\t.\tID=Phabio.48;Name=Phabio.48\n",
        "\n",
        "142 Phabio_draft\tGroup\tmRNA\t37365\t37643\t200\t+\t.\tID=Phabio.48.mRNA;Parent=Phabio.48\n",
        "\n",
        "143 Phabio_draft\tGroup\texon\t37365\t37643\t200\t+\t.\tID=Phabio.48.exon;Parent=Phabio.48.mRNA\n",
        "\n",
        "144 Phabio_draft\tGroup\tgene\t37640\t37987\t200\t+\t.\tID=Phabio.49;Name=Phabio.49\n",
        "\n",
        "145 Phabio_draft\tGroup\tmRNA\t37640\t37987\t200\t+\t.\tID=Phabio.49.mRNA;Parent=Phabio.49\n",
        "\n",
        "146 Phabio_draft\tGroup\texon\t37640\t37987\t200\t+\t.\tID=Phabio.49.exon;Parent=Phabio.49.mRNA\n",
        "\n",
        "147 Phabio_draft\tGroup\tgene\t38391\t38891\t200\t+\t.\tID=Phabio.50;Name=Phabio.50\n",
        "\n",
        "148 Phabio_draft\tGroup\tmRNA\t38391\t38891\t200\t+\t.\tID=Phabio.50.mRNA;Parent=Phabio.50\n",
        "\n",
        "149 Phabio_draft\tGroup\texon\t38391\t38891\t200\t+\t.\tID=Phabio.50.exon;Parent=Phabio.50.mRNA\n",
        "\n",
        "150 Phabio_draft\tGroup\tgene\t38888\t39253\t200\t+\t.\tID=Phabio.51;Name=Phabio.51\n",
        "\n",
        "151 Phabio_draft\tGroup\tmRNA\t38888\t39253\t200\t+\t.\tID=Phabio.51.mRNA;Parent=Phabio.51\n",
        "\n",
        "152 Phabio_draft\tGroup\texon\t38888\t39253\t200\t+\t.\tID=Phabio.51.exon;Parent=Phabio.51.mRNA\n",
        "\n",
        "153 Phabio_draft\tGroup\tgene\t39405\t39635\t200\t+\t.\tID=Phabio.52;Name=Phabio.52\n",
        "\n",
        "154 Phabio_draft\tGroup\tmRNA\t39405\t39635\t200\t+\t.\tID=Phabio.52.mRNA;Parent=Phabio.52\n",
        "\n",
        "155 Phabio_draft\tGroup\texon\t39405\t39635\t200\t+\t.\tID=Phabio.52.exon;Parent=Phabio.52.mRNA\n",
        "\n",
        "156 Phabio_draft\tGroup\tgene\t39623\t39877\t200\t+\t.\tID=Phabio.53;Name=Phabio.53\n",
        "\n",
        "157 Phabio_draft\tGroup\tmRNA\t39623\t39877\t200\t+\t.\tID=Phabio.53.mRNA;Parent=Phabio.53\n",
        "\n",
        "158 Phabio_draft\tGroup\texon\t39623\t39877\t200\t+\t.\tID=Phabio.53.exon;Parent=Phabio.53.mRNA\n",
        "\n",
        "159 Phabio_draft\tGroup\tgene\t39874\t40851\t200\t+\t.\tID=Phabio.54;Name=Phabio.54\n",
        "\n",
        "160 Phabio_draft\tGroup\tmRNA\t39874\t40851\t200\t+\t.\tID=Phabio.54.mRNA;Parent=Phabio.54\n",
        "\n",
        "161 Phabio_draft\tGroup\texon\t39874\t40851\t200\t+\t.\tID=Phabio.54.exon;Parent=Phabio.54.mRNA\n",
        "\n",
        "162 Phabio_draft\tGroup\tgene\t40848\t41546\t200\t+\t.\tID=Phabio.55;Name=Phabio.55\n",
        "\n",
        "163 Phabio_draft\tGroup\tmRNA\t40848\t41546\t200\t+\t.\tID=Phabio.55.mRNA;Parent=Phabio.55\n",
        "\n",
        "164 Phabio_draft\tGroup\texon\t40848\t41546\t200\t+\t.\tID=Phabio.55.exon;Parent=Phabio.55.mRNA\n",
        "\n",
        "165 Phabio_draft\tGroup\tgene\t42125\t42682\t200\t+\t.\tID=Phabio.56;Name=Phabio.56\n",
        "\n",
        "166 Phabio_draft\tGroup\tmRNA\t42125\t42682\t200\t+\t.\tID=Phabio.56.mRNA;Parent=Phabio.56\n",
        "\n",
        "167 Phabio_draft\tGroup\texon\t42125\t42682\t200\t+\t.\tID=Phabio.56.exon;Parent=Phabio.56.mRNA\n",
        "\n",
        "168 Phabio_draft\tGroup\tgene\t42682\t43641\t200\t+\t.\tID=Phabio.57;Name=Phabio.57\n",
        "\n",
        "169 Phabio_draft\tGroup\tmRNA\t42682\t43641\t200\t+\t.\tID=Phabio.57.mRNA;Parent=Phabio.57\n",
        "\n",
        "170 Phabio_draft\tGroup\texon\t42682\t43641\t200\t+\t.\tID=Phabio.57.exon;Parent=Phabio.57.mRNA\n",
        "\n",
        "171 Phabio_draft\tGroup\tgene\t43641\t43904\t200\t+\t.\tID=Phabio.58;Name=Phabio.58\n",
        "\n",
        "172 Phabio_draft\tGroup\tmRNA\t43641\t43904\t200\t+\t.\tID=Phabio.58.mRNA;Parent=Phabio.58\n",
        "\n",
        "173 Phabio_draft\tGroup\texon\t43641\t43904\t200\t+\t.\tID=Phabio.58.exon;Parent=Phabio.58.mRNA\n",
        "\n",
        "174 Phabio_draft\tGroup\tgene\t43901\t44281\t200\t+\t.\tID=Phabio.59;Name=Phabio.59\n",
        "\n",
        "175 Phabio_draft\tGroup\tmRNA\t43901\t44281\t200\t+\t.\tID=Phabio.59.mRNA;Parent=Phabio.59\n",
        "\n",
        "176 Phabio_draft\tGroup\texon\t43901\t44281\t200\t+\t.\tID=Phabio.59.exon;Parent=Phabio.59.mRNA\n",
        "\n",
        "177 Phabio_draft\tGroup\tgene\t44472\t44603\t200\t+\t.\tID=Phabio.60;Name=Phabio.60\n",
        "\n",
        "178 Phabio_draft\tGroup\tmRNA\t44472\t44603\t200\t+\t.\tID=Phabio.60.mRNA;Parent=Phabio.60\n",
        "\n",
        "179 Phabio_draft\tGroup\texon\t44472\t44603\t200\t+\t.\tID=Phabio.60.exon;Parent=Phabio.60.mRNA\n",
        "\n",
        "180 Phabio_draft\tGroup\tgene\t45389\t45550\t200\t+\t.\tID=Phabio.61;Name=Phabio.61\n",
        "\n",
        "181 Phabio_draft\tGroup\tmRNA\t45389\t45550\t200\t+\t.\tID=Phabio.61.mRNA;Parent=Phabio.61\n",
        "\n",
        "182 Phabio_draft\tGroup\texon\t45389\t45550\t200\t+\t.\tID=Phabio.61.exon;Parent=Phabio.61.mRNA\n",
        "\n",
        "183 Phabio_draft\tGroup\tgene\t45574\t45762\t200\t+\t.\tID=Phabio.62;Name=Phabio.62\n",
        "\n",
        "184 Phabio_draft\tGroup\tmRNA\t45574\t45762\t200\t+\t.\tID=Phabio.62.mRNA;Parent=Phabio.62\n",
        "\n",
        "185 Phabio_draft\tGroup\texon\t45574\t45762\t200\t+\t.\tID=Phabio.62.exon;Parent=Phabio.62.mRNA\n",
        "\n",
        "186 Phabio_draft\tGroup\tgene\t46080\t46541\t200\t+\t.\tID=Phabio.63;Name=Phabio.63\n",
        "\n",
        "187 Phabio_draft\tGroup\tmRNA\t46080\t46541\t200\t+\t.\tID=Phabio.63.mRNA;Parent=Phabio.63\n",
        "\n",
        "188 Phabio_draft\tGroup\texon\t46080\t46541\t200\t+\t.\tID=Phabio.63.exon;Parent=Phabio.63.mRNA\n",
        "\n",
        "189 Phabio_draft\tGroup\tgene\t46755\t47108\t200\t+\t.\tID=Phabio.64;Name=Phabio.64\n",
        "\n",
        "190 Phabio_draft\tGroup\tmRNA\t46755\t47108\t200\t+\t.\tID=Phabio.64.mRNA;Parent=Phabio.64\n",
        "\n",
        "191 Phabio_draft\tGroup\texon\t46755\t47108\t200\t+\t.\tID=Phabio.64.exon;Parent=Phabio.64.mRNA\n",
        "\n",
        "192 Phabio_draft\tGroup\tgene\t47325\t47537\t200\t+\t.\tID=Phabio.65;Name=Phabio.65\n",
        "\n",
        "193 Phabio_draft\tGroup\tmRNA\t47325\t47537\t200\t+\t.\tID=Phabio.65.mRNA;Parent=Phabio.65\n",
        "\n",
        "194 Phabio_draft\tGroup\texon\t47325\t47537\t200\t+\t.\tID=Phabio.65.exon;Parent=Phabio.65.mRNA\n",
        "\n",
        "195 Phabio_draft\tGroup\tgene\t47701\t47991\t200\t+\t.\tID=Phabio.66;Name=Phabio.66\n",
        "\n",
        "196 Phabio_draft\tGroup\tmRNA\t47701\t47991\t200\t+\t.\tID=Phabio.66.mRNA;Parent=Phabio.66\n",
        "\n",
        "197 Phabio_draft\tGroup\texon\t47701\t47991\t200\t+\t.\tID=Phabio.66.exon;Parent=Phabio.66.mRNA\n",
        "\n",
        "198 Phabio_draft\tGroup\tgene\t47988\t48167\t200\t+\t.\tID=Phabio.67;Name=Phabio.67\n",
        "\n",
        "199 Phabio_draft\tGroup\tmRNA\t47988\t48167\t200\t+\t.\tID=Phabio.67.mRNA;Parent=Phabio.67\n",
        "\n",
        "200 Phabio_draft\tGroup\texon\t47988\t48167\t200\t+\t.\tID=Phabio.67.exon;Parent=Phabio.67.mRNA\n",
        "\n",
        "201 Phabio_draft\tGroup\tgene\t48172\t48309\t200\t+\t.\tID=Phabio.68;Name=Phabio.68\n",
        "\n",
        "202 Phabio_draft\tGroup\tmRNA\t48172\t48309\t200\t+\t.\tID=Phabio.68.mRNA;Parent=Phabio.68\n",
        "\n",
        "203 Phabio_draft\tGroup\texon\t48172\t48309\t200\t+\t.\tID=Phabio.68.exon;Parent=Phabio.68.mRNA\n",
        "\n",
        "204 Phabio_draft\tGroup\tgene\t48422\t48658\t200\t+\t.\tID=Phabio.69;Name=Phabio.69\n",
        "\n",
        "205 Phabio_draft\tGroup\tmRNA\t48422\t48658\t200\t+\t.\tID=Phabio.69.mRNA;Parent=Phabio.69\n",
        "\n",
        "206 Phabio_draft\tGroup\texon\t48422\t48658\t200\t+\t.\tID=Phabio.69.exon;Parent=Phabio.69.mRNA\n",
        "\n",
        "207 Phabio_draft\tGroup\tgene\t48651\t48902\t200\t+\t.\tID=Phabio.70;Name=Phabio.70\n",
        "\n",
        "208 Phabio_draft\tGroup\tmRNA\t48651\t48902\t200\t+\t.\tID=Phabio.70.mRNA;Parent=Phabio.70\n",
        "\n",
        "209 Phabio_draft\tGroup\texon\t48651\t48902\t200\t+\t.\tID=Phabio.70.exon;Parent=Phabio.70.mRNA\n",
        "\n",
        "210 Phabio_draft\tGroup\tgene\t48899\t49543\t200\t+\t.\tID=Phabio.71;Name=Phabio.71\n",
        "\n",
        "211 Phabio_draft\tGroup\tmRNA\t48899\t49543\t200\t+\t.\tID=Phabio.71.mRNA;Parent=Phabio.71\n",
        "\n",
        "212 Phabio_draft\tGroup\texon\t48899\t49543\t200\t+\t.\tID=Phabio.71.exon;Parent=Phabio.71.mRNA\n",
        "\n",
        "213 Phabio_draft\tGroup\tgene\t49653\t49877\t200\t+\t.\tID=Phabio.72;Name=Phabio.72\n",
        "\n",
        "214 Phabio_draft\tGroup\tmRNA\t49653\t49877\t200\t+\t.\tID=Phabio.72.mRNA;Parent=Phabio.72\n",
        "\n",
        "215 Phabio_draft\tGroup\texon\t49653\t49877\t200\t+\t.\tID=Phabio.72.exon;Parent=Phabio.72.mRNA\n",
        "\n",
        "216 Phabio_draft\tGroup\tgene\t49874\t50038\t200\t+\t.\tID=Phabio.73;Name=Phabio.73\n",
        "\n",
        "217 Phabio_draft\tGroup\tmRNA\t49874\t50038\t200\t+\t.\tID=Phabio.73.mRNA;Parent=Phabio.73\n",
        "\n",
        "218 Phabio_draft\tGroup\texon\t49874\t50038\t200\t+\t.\tID=Phabio.73.exon;Parent=Phabio.73.mRNA\n",
        "\n",
        "219 Phabio_draft\tGroup\tgene\t50372\t50788\t200\t+\t.\tID=Phabio.74;Name=Phabio.74\n",
        "\n",
        "220 Phabio_draft\tGroup\tmRNA\t50372\t50788\t200\t+\t.\tID=Phabio.74.mRNA;Parent=Phabio.74\n",
        "\n",
        "221 Phabio_draft\tGroup\texon\t50372\t50788\t200\t+\t.\tID=Phabio.74.exon;Parent=Phabio.74.mRNA\n",
        "\n",
        "222 Phabio_draft\tGroup\tgene\t50785\t51027\t200\t+\t.\tID=Phabio.75;Name=Phabio.75\n",
        "\n",
        "223 Phabio_draft\tGroup\tmRNA\t50785\t51027\t200\t+\t.\tID=Phabio.75.mRNA;Parent=Phabio.75\n",
        "\n",
        "224 Phabio_draft\tGroup\texon\t50785\t51027\t200\t+\t.\tID=Phabio.75.exon;Parent=Phabio.75.mRNA\n",
        "\n",
        "225 Phabio_draft\tGroup\tgene\t52643\t51207\t200\t+\t.\tID=Phabio.76;Name=Phabio.76\n",
        "\n",
        "226 Phabio_draft\tGroup\tmRNA\t52643\t51207\t200\t+\t.\tID=Phabio.76.mRNA;Parent=Phabio.76\n",
        "\n",
        "227 Phabio_draft\tGroup\texon\t52643\t51207\t200\t+\t.\tID=Phabio.76.exon;Parent=Phabio.76.mRNA\n",
        "\n",
        "228 Phabio_draft\tGroup\tgene\t52648\t53118\t200\t+\t.\tID=Phabio.77;Name=Phabio.77\n",
        "\n",
        "229 Phabio_draft\tGroup\tmRNA\t52648\t53118\t200\t+\t.\tID=Phabio.77.mRNA;Parent=Phabio.77\n",
        "\n",
        "230 Phabio_draft\tGroup\texon\t52648\t53118\t200\t+\t.\tID=Phabio.77.exon;Parent=Phabio.77.mRNA\n",
        "\n",
        "231 Phabio_draft\tGroup\tgene\t53324\t53935\t200\t+\t.\tID=Phabio.78;Name=Phabio.78\n",
        "\n",
        "232 Phabio_draft\tGroup\tmRNA\t53324\t53935\t200\t+\t.\tID=Phabio.78.mRNA;Parent=Phabio.78\n",
        "\n",
        "233 Phabio_draft\tGroup\texon\t53324\t53935\t200\t+\t.\tID=Phabio.78.exon;Parent=Phabio.78.mRNA\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import re,sys\n",
      "\n",
      "def validHeader (line):\n",
      "    \"\"\"\n",
      "    Return True if line is a valid gff3 header.\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line' - line to check\n",
      "    \"\"\"\n",
      "    if line == \"##gff-version 3\\n\":\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "    \n",
      "def validTabStructure(line):\n",
      "    \"\"\"\n",
      "    Checks a line for valid <tab> structure, a valid GFF3 file should have 9 entries and 8 <Tab> charachters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'line': line to check\n",
      "    \"\"\"\n",
      "    theLine = line.strip().split(\"\\t\")  #strip off extra charachters and split the line on tab\n",
      "    if len(theLine) != 9:               # 9 columns in a gff file, so return False if not exactly 9\n",
      "        return False\n",
      "    else:\n",
      "        return True\n",
      "\n",
      "def validSeqname(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 1 of gff file, for this column it must match\n",
      "    the sequence name on the Gbrowse database and have only valid characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of entry in column 1 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)  # need name check, for now just check for valid charachters\n",
      "\n",
      "def validSource(text):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 2 of gff file, for this column should be source\n",
      "    which only need validity of the characters\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    \"\"\"\n",
      "    return not charCheck(text)\n",
      "\n",
      "def validType(text, validTypes=None):\n",
      "    \"\"\"\n",
      "    Checks a string to make sure it is a valid entry for column 3 of gff file, for this column should be feature type.\n",
      "    For phage this should be one of types typically ['gene','mRNA','exon']\n",
      "    \n",
      "    Parameters:\n",
      "    - 'text': string value of the entry in column 2 to check\n",
      "    - 'validTypes': List of valid types for checking\n",
      "    \"\"\"\n",
      "    if validTypes is None:\n",
      "        validTypes = {'gene','mRNA','exon'}\n",
      "    \n",
      "    return text in validTypes\n",
      "\n",
      "def validCoordinate(coord):\n",
      "    \"\"\"\n",
      "    Checks a coordinate (entries in column 4 and 5), should be a positive integer\n",
      "    \n",
      "    Parameters:\n",
      "    - 'coord': coordinate from column 4 or 5\n",
      "    \"\"\"\n",
      "    try:\n",
      "        x = int(coord)\n",
      "    except:\n",
      "        isCoordinate = False\n",
      "        pass\n",
      "    else:\n",
      "        if x > 0:\n",
      "            print \"tested positive\"\n",
      "            isCoordinate = True\n",
      "        else:\n",
      "            print \"tested negative\"\n",
      "            isCoordinate = False\n",
      "    \n",
      "    return isCoordinate\n",
      "\n",
      "def validScore(score):\n",
      "    \"\"\"\n",
      "    Checks a score (entries in column 6), should be a number or \".\"\n",
      "    \n",
      "    Parameters:\n",
      "    - 'score': score from column 6\n",
      "    \"\"\"\n",
      "    if score == \".\":\n",
      "        return True\n",
      "    \n",
      "    try:\n",
      "        x = float(score)\n",
      "    except:\n",
      "        isScore = False\n",
      "        pass\n",
      "    else:\n",
      "        isScore = True        \n",
      "\n",
      "    return isScore\n",
      "\n",
      "def validStrand(strand):\n",
      "    \"\"\"\n",
      "    Checks a strand (entries in column 7), should be one of (\"=\", \"-\", \".\", \"?\")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'strand': strand entry from column 7\n",
      "    \"\"\"\n",
      "    validValues = {\"+\", \"-\", \".\", \"?\"}\n",
      "    \n",
      "    return strand in validValues\n",
      "\n",
      "def validPhase(phase):\n",
      "    \"\"\"\n",
      "    Checks a phase (entries in column 8), should be one of (\".\", 0, 1, 2, \")\n",
      "    \n",
      "    Parameters:\n",
      "    - 'phase': strand entry from column 8\n",
      "    \"\"\"\n",
      "    validValues = {\".\", 0, \"0\" , 1, \"1\", 2, \"2\"}\n",
      "    \n",
      "    return phase in validValues\n",
      "\n",
      "def validAttributes(attributes):\n",
      "    \"\"\"\n",
      "    Checks a column 9 entry, should be series of key=value entries separated by ; \n",
      "    It is OK to have spaces in the values entry but not the Key value\n",
      "    \n",
      "    \n",
      "    Parameters:\n",
      "    - 'attributes': entire entry from column 9\n",
      "    \"\"\"\n",
      "    \n",
      "    #ok to have a null string\n",
      "    if attributes == '':\n",
      "        return True\n",
      "    \n",
      "     #checking for invalid characters, spaces OK in column 9 values so remove any before checking with charCheck\n",
      "\n",
      "        \n",
      "    #ok go ahead and split into the underlyine key=value items    \n",
      "    attrPairs = attributes.split(\";\")   #list of each attribute pair\n",
      "    \n",
      "    validity = True\n",
      "    \n",
      "    for attrPair in attrPairs:\n",
      "        print attrPair\n",
      "        if attrPair == '':                  #this will happen if file had two ;; in a row\n",
      "            continue\n",
      "            \n",
      "        if attrPair.count('=') != 1:        #for each Key=value there must be only one \"=\"\n",
      "            validity = False\n",
      "        \n",
      "        attrKey, attrValue = attrPair.split('=')\n",
      "        \n",
      "        if len(attrKey) < 1:                #must have an entry for a key\n",
      "            validity = False\n",
      "        \n",
      "        if charCheck(attrKey):              #check key for valid charachters\n",
      "            validity = False\n",
      "        \n",
      "        # check for invalid charachters in value, but spaces are OK, so remove before checking\n",
      "        if charCheck(attrValue.replace(' ','')):\n",
      "            validity = False\n",
      "            \n",
      "    return validity\n",
      "\n",
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters not a-zA-Z0-9.=; and returns True if such a character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    -'search': default search all characters in valid set\n",
      "    \"\"\"\n",
      "    return bool(search(str))\n",
      "\n",
      "def validGene(seq, line):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is define a proper gene.\n",
      "    \n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has one stop codon and it is the last codon.\n",
      "    - begins with a valid start codon [ATG, GTG, CTG].\n",
      "    \n",
      "    Parameters:\n",
      "    -'Seq': string nucleotide sequence of entire phage.\n",
      "    -'line': single line in gff3 format for checking.\n",
      "    \"\"\"\n",
      "    \n",
      "    leftCoordinate= int(line.split(\"\\t\")[3])\n",
      "    rightCoordinate= int(line.split(\"\\t\")[4])\n",
      "    strand = line.split(\"\\t\")[6]\n",
      "    geneSeq = seq[leftCoordinate-1:rightCoordinate]   #not really the gene for negative strand genes but good enough\n",
      "    \n",
      "    if  strand == \"+\":\n",
      "        firstCodon = geneSeq[:3].upper()\n",
      "        lastCodon = geneSeq[-3:].upper()\n",
      "        startCodons = ['ATG', 'TTG', 'GTG']\n",
      "        stopCodons = ['TAA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        firstCodon = geneSeq[-3:].upper()\n",
      "        lastCodon = geneSeq[:3].upper()\n",
      "        startCodons = ['CAT', 'CAA', 'CAC']   #reverse complement of the start codons\n",
      "        stopCodons = ['TTA', 'CTA', 'TCA']   #reverse complement of the stop codons\n",
      "\n",
      "    #first make sure the values of leftCoordinate and rightCoordinate are OK\n",
      "    \n",
      "    if (leftCoordinate > len(seq)) or (rightCoordinate >= len(seq)):  #zero based counting \n",
      "        return False\n",
      "    \n",
      "    if (rightCoordinate - leftCoordinate) < 30:                       #minimum gene size for this check is 10 a.a.\n",
      "        return False\n",
      "    \n",
      "    if len(geneSeq)%3 != 0:\n",
      "        return False\n",
      "    \n",
      "    if firstCodon not in startCodons:\n",
      "        return False\n",
      "    \n",
      "    if lastCodon not in stopCodons:\n",
      "        return False\n",
      "    \n",
      "    stopCodonCount = countStopCodons(geneSeq,strand)\n",
      "    if  stopCodonCount > 1:\n",
      "        return False\n",
      "    \n",
      "    return True\n",
      "\n",
      "def fastaRead(fasta_File):\n",
      "    \"\"\"\n",
      "    Reads in a nucleotide sequence from a text file in .fasta format.\n",
      "\n",
      "    Parameters:\n",
      "    -'fasta_File': uploaded text file in .fasta format\n",
      "\n",
      "    Output:\n",
      "    -'seq': string of a nucleotide sequence.\n",
      "    \"\"\"\n",
      "\n",
      "    seq = []\n",
      "    try:\n",
      "        f1 = open(fasta_File, \"r\")\n",
      "    except:\n",
      "        return \"Error: unable to read fasta file.\"\n",
      "    \n",
      "    for line in f1:\n",
      "        line = line.rstrip()\n",
      "        if line.startswith(\">\"):\n",
      "            continue\n",
      "        else:\n",
      "            seq.append(line)\n",
      "\n",
      "    f1.close()\n",
      "    return ''.join(seq)\n",
      "\n",
      "def countStopCodons(seq, strand):\n",
      "    \"\"\"\n",
      "    Used to count the number of stop codons in a given sequence\n",
      "\n",
      "    Parameters:\n",
      "    -'seq': nucleotide sequence of plus strand.\n",
      "    -'strand': indication if gene is on the plus or minus strand\n",
      "\n",
      "    Output:\n",
      "    -'count': total number of stop codons found in the sequence.\n",
      "    \"\"\"\n",
      "    \n",
      "    count = 0\n",
      "    \n",
      "    if strand == \"+\":\n",
      "        stopCodons = ['TTA', 'TAG', 'TGA']\n",
      "    else:\n",
      "        stopCodons = ['TAA', 'CTA', 'TCA']   ##reverse complement of stop codons\n",
      "        \n",
      "    #now create a list of all codons using list comprehension, this only works well if len%3 == 0\n",
      "    if strand == \"+\":\n",
      "        codons=[seq[x:x+3] for x in range(0,len(seq),3)]\n",
      "    else:\n",
      "        codons=[seq[x:x+3] for x in range(-3,-len(seq),-3)] #works for all but first codon becuase string[-3:0]=''\n",
      "        codons[0] = seq[-3:]                                #so fudge in the \"first\" codon on the negative side\n",
      "    \n",
      "    for stop in stopCodons:\n",
      "        count += codons.count(stop)\n",
      "        \n",
      "    return count"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}