{
 "metadata": {
  "name": "",
  "signature": "sha256:c9afa002f8a94943345ec33b83b7bda68ce07b970de8d9f6071ab0c8a5dca9f5"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pwd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys, re, Bio\n",
      "from Bio.Seq import Seq\n",
      "from Bio.Alphabet import IUPAC\n",
      "from Bio import SeqIO"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "set up functions"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def sortGff3"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def sortGff3(gff3_File, types = ['gene','mRNA','exon']): ## don't forget to add name checking\n",
      "    \"\"\"   \n",
      "    This checks for proper \"structure\" for correct format:\n",
      "    - file starts with proper header\n",
      "    - each line is tab delimited and has 9 components.\n",
      "    - each line has a proper type at the 3rd component.\n",
      "    - example line:\n",
      "    1stComp\\t2ndComp\\t3rdComp\\t4thComp\\t...\n",
      "    \n",
      "    It also sorts each line in the gff file according to the location of the entry and the\n",
      "    type hierarchy. Valid types are 'gene', 'mRNA', 'exon' and must match exactly. \n",
      "    Gene != gene, mrna != mRNA, etc.\n",
      "    \n",
      "    Parameters:\n",
      "    -'gff3_File': the uploaded gff file.\n",
      "    -'types': a list indicating the types of each gff line and the order the types should\n",
      "              be sorted in. The first type in the list will be ordered before the second, \n",
      "              the second type before the third, etc.\n",
      "\n",
      "    Output:\n",
      "    -'keyList': a list of keys sorted in the order that the lines will be sorted.\n",
      "    -'holder': a dictionary of lines paired with keys in 'keyList'.\n",
      "    \"\"\"\n",
      "    \n",
      "    holder = dict()                    #keys are <type>_leftcoord; value is the entire line\n",
      "    counter = dict()                   #keys are leftcoord; values are sum of indexs of types found with types list\n",
      "    structure = []                     #this will hold the errors from this function, which are considered \"structural\"\n",
      "    contigCount = False\n",
      "\n",
      "    formatCount = 0\n",
      "    \n",
      "    with open(gff3_File, \"r\") as file:\n",
      "        fi = file.readlines()              #OK for phages, may fail for large gff files\n",
      "\n",
      "    try:   #check if first line is proper header\n",
      "        if fi[0] != \"##gff-version 3\\n\":\n",
      "            raise LineError(\"0026\")\n",
      "    except LineError as er:\n",
      "        structure.append(\"[ 0 ] \" + er.returnError())\n",
      "\n",
      "        \n",
      "    for lineCount,line in enumerate(fi[1:]):\n",
      "\n",
      "        try:\n",
      "            theLine = line.strip().split(\"\\t\")  #split on tab\n",
      "            if len(theLine) > 9: # 9 columns in a proper gff file\n",
      "                raise FormatError(\"0200\")\n",
      "            elif len(theLine) < 9:\n",
      "                raise FormatError(\"0100\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"[line# \" + str(lineCount+2) + \"] \" + er.returnError())\n",
      "            continue\n",
      "        except:\n",
      "            structure.append(\"Validation Error: unkown error. Check line# \" + str(lineCount+2) + \":\\n\" + line)\n",
      "\n",
      "        if theLine[2] == \"contig\":\n",
      "            holder[theLine[2] + \"_\" + theLine[3]] = line\n",
      "            continue\n",
      "\n",
      "        try:\n",
      "            if theLine[2] not in types:\n",
      "                raise FormatError(\"0700\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"[\" + str(lineCount+2) + \"] \" + er.returnError())\n",
      "            continue\n",
      "\n",
      "        try:\n",
      "            leftPosition = int(theLine[3])\n",
      "            rightPosition = int(theLine[4])\n",
      "        except ValueError as er:\n",
      "            structure.append(\"[\" + str(lineCount+2) + \"] \" + \"coordinates not readable as numbers in column 4 and/or 5\")\n",
      "            continue\n",
      "\n",
      "        #to get here line will is partially validated, it will:\n",
      "        # 1. not be type contig\n",
      "        # 2. have 9 components \n",
      "        # 3. have a valid type in 3rd column\n",
      "        # 4. have integers in 4th and 5th column\n",
      "\n",
      "        holder[theLine[2] + \"_\" + theLine[3]] = line #FIXME: problem if 2 genes with same start number (chaperones)\n",
      "\n",
      "        if theLine[3] in counter:\n",
      "            counter[theLine[3]] = counter[theLine[3]] + types.index(theLine[2])\n",
      "        else:\n",
      "            counter[theLine[3]] = types.index(theLine[2])\n",
      "\n",
      "\n",
      "    #this part checks that you have the correct number of entries for each start coordinate:\n",
      "    counterKeys = counter.keys()\n",
      "    \n",
      "    expectedNum = ((len(types))*(len(types)-1))/2\n",
      "    print \"coutnerKeys are: \" + str(counterKeys)\n",
      "\n",
      "    for key in counterKeys:  \n",
      "        thisNum = counter[key]\n",
      "        try:\n",
      "            if thisNum == expectedNum:\n",
      "                continue\n",
      "            else:\n",
      "                raise FormatError(\"0500\")\n",
      "        except FormatError as er:\n",
      "            structure.append(\"Gene at Coordinate \" + key + \" \" + er.returnError())\n",
      "\n",
      "    Errors[\"structure\"] = structure                    #add the list of structural errors to the Errors dictionary\n",
      "    \n",
      "    #need to sort all the keys \n",
      "    keyList = holder.keys()\n",
      "    \n",
      "\n",
      "    #first sort on the secondary charachteriestic which is type:\n",
      "    typeHierarchy = ['contig','gene','mRNA','exon']\n",
      "    typeSortList = sorted(keyList, key=lambda item: typeHierarchy.index(item.split(\"_\")[0] ))\n",
      "                      \n",
      "    #now sort on the primary charachteristic which is the left coordinate \n",
      "    sortedKeyList = sorted(typeSortList, key=lambda item: int(item.split(\"_\")[1]))\n",
      "            \n",
      "    tempS = str(gff3_File)\n",
      "    newName = tempS[0:-5] + \"_sorted.gff3\"\n",
      "    with open(newName,\"w\") as outfile:\n",
      "        outfile.write(\"##gff-version 3\\n\")\n",
      "        for key in sortedKeyList:\n",
      "            outfile.write(holder[key])\n",
      "            \n",
      "    return [sortedKeyList, holder]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def fileCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fileCheck(keyList, holder, Seq, types = ['gene','mRNA','exon']):\n",
      "    \"\"\"\n",
      "    Checks each component of each line of the gff file for proper format. These are considered \"entry\" errors.\n",
      "    All entry errors will be added to Errors dictionary.\n",
      "\n",
      "    Parameters:\n",
      "    -'keyList': list of sorted keys.\n",
      "    -'holder': dictionary of lines paired with keys in 'keyList'.\n",
      "    -'Seq': string of the nucleotide sequence extracted from uploaded fasta file.\n",
      "    -'types': a list indicating the types of each gff line and the order the types should\n",
      "              be sorted in. The first type in the list will be ordered before the second, \n",
      "              the second type before the third, etc.\n",
      "\n",
      "\"\"\"\n",
      "        \n",
      "    entry = []                  \n",
      "    idList = set()                 #list of ID's seen to check for uniquenss\n",
      "    \n",
      "    \n",
      "    \n",
      "    for line,key in enumerate(keyList[1:]):   #skipping first line as it is the contig\n",
      "        attributes = dict()                   #will be used to hold each Key/value pair of column 9 for each line\n",
      "        \n",
      "        lineCount = line + 1    #adding 1 for unacccounted for header line\n",
      "        \n",
      "        try:\n",
      "            theLine = holder[key].strip().split(\"\\t\") ## try here to catch if  not tab deliminated or adding extra lines\n",
      "        except:\n",
      "            entry.append(\"Format Error: each line needs to be tab deliminated.\")\n",
      "            continue\n",
      "        \n",
      "        \n",
      "        if len(theLine) != 9:   #this is just to skip improper lines, error already caught by sortGFF3\n",
      "            continue\n",
      "        \n",
      "            \n",
      "        ### 1ST and 2ND ITEM ### \n",
      "        try:\n",
      "            if charCheck(theLine[0]):\n",
      "                raise LineError(\"0002\")\n",
      "            elif charCheck(theLine[1]):\n",
      "                raise LineError(\"0003\")\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "        \n",
      "        ### 3RD ITEM ###\n",
      "        try:\n",
      "            if theLine[2] not in types:\n",
      "                raise LineError(\"0004\")  ## types can be changed if more types of line needed\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "            continue                     ##no other checks done if not a proper type\n",
      "        \n",
      "        ### 4TH ITEM ###\n",
      "        try:\n",
      "            if not int(theLine[3]) > 0:\n",
      "                raise LineError(\"0005\") ## needs to be greater than 0\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "        except ValueError:\n",
      "            entry.append(\"[\" + str(lineCount) + \"] coordinates in column 4 not readable as number\") \n",
      "        \n",
      "        \n",
      "        ### 5TH ITEM ###\n",
      "        try:\n",
      "            if not int(theLine[4]) > int(theLine[3]):\n",
      "                raise LineError(\"0006\") ## needs to be greater than 1st coordinate\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "        except ValueError:\n",
      "            entry.append(\"[\" + str(lineCount) + \"] coordinates in column 5 not readable as number\") \n",
      "        \n",
      "        ### 6TH ITEM ###\n",
      "        try:\n",
      "            if theLine[5] != \".\": ## can also be a number\n",
      "                try:\n",
      "                    int(theLine[5])\n",
      "                except ValueError:\n",
      "                    raise LineError(\"0007\")\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "            \n",
      "        ### 7TH ITEM ###\n",
      "        try:\n",
      "            if theLine[6] != \"+\":\n",
      "                if theLine[6] != \"-\":\n",
      "                    raise LineError(\"0008\")  ## + or - strand\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "                \n",
      "         ### 8TH ITEM ### \n",
      "        try:\n",
      "            if theLine[7] != \".\":\n",
      "                raise LineError(\"0009\")\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "                \n",
      "        \n",
      "        ### 9TH ITEM ###      \n",
      "        ### 9th item is an attributes list \";\" separated list of key/value pairs in the form key=value\n",
      "       \n",
      "        #start by checking for invalid characters\n",
      "        try:\n",
      "            if charCheck(theLine[8]):\n",
      "                raise LineError(\"0011\")\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "            continue              #don't even try to parse if invalid charachters, just move to next line \n",
      "\n",
      "\n",
      "        #checking for proper separation of key/value pairs with the ';' \n",
      "        try:\n",
      "            if theLine[8].count(\"=\") > (theLine[8].count(\";\") + 1):\n",
      "                raise LineError(\"0028\")\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "            continue               #again don't even try with this much failure\n",
      "                \n",
      "        #ok go ahead and split into the underlyine key=value items    \n",
      "        attrItems = theLine[8].split(\";\")   #list of each attribute pair\n",
      "\n",
      "   \n",
      "        #run through the list of attrItems; make you they are in key=value and put everything in attributes dict\n",
      "        for attr in attrItems:\n",
      "            if len(attr) == 0:           #catch attr from extra ; but just skip it\n",
      "                continue\n",
      "            try:        ## start with \"=\" search\n",
      "                if attr.find(\"=\") < 0:\n",
      "                    raise LineError(\"0027\")\n",
      "            except LineError as er:\n",
      "                entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "                continue                 #if it doesn't have an \"=\" don't bother with the checks below\n",
      "                    \n",
      "            #to get to here attr has to have an \"=\"; so split to check each of key and value\n",
      "            attributeName, attributeValue = attr.split(\"=\")\n",
      "            try:                        #make sure of both a attributeName (key) and attributeValue (value)\n",
      "                if len(attributeName) < 1 or len(attributeValue) < 1:\n",
      "                    raise LineError(\"0027\")\n",
      "                else:                   #ok to add to attributes dict\n",
      "                    attributes[attributeName] = attributeValue\n",
      "            except LineError as er:\n",
      "                entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "                    \n",
      "        #start parsing the attributes dictionary\n",
      "        try:            #do checks for ID presence and uniqueness\n",
      "            if \"ID\" in attributes:\n",
      "                try:\n",
      "                    if attributes[\"ID\"] in idList:\n",
      "                        raise LineError(\"0012\") ## each id can only be used once \n",
      "                    else:\n",
      "                        idList.add(attributes[\"ID\"])\n",
      "                except LineError as er:\n",
      "                        entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "                                \n",
      "            else:\n",
      "                raise LineError(\"0013\")  ##ID is missing\n",
      "        except LineError as er:\n",
      "            entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "\n",
      "                     \n",
      "\n",
      "        ######################## checks that are specific to certain types #############\n",
      "        if theLine[2] == types[0]:       #is a gene line; check the biology and check for a Name\n",
      "            #geneCheck(int(theLine[3]),int(theLine[4]),Seq,lineCount)\n",
      "            try:  #do checks for Name presence\n",
      "                if \"Name\" in attributes:\n",
      "                    _Name = True\n",
      "                else:\n",
      "                    raise LineError(\"0013\")\n",
      "            except LineError as er:\n",
      "                entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "\n",
      "        elif types.index(theLine[2]) > 0:    #is an mRNA or exon, check for parent\n",
      "            try:  #do checks for parent presence\n",
      "                if \"Parent\" in attributes:\n",
      "                    _Parent = True\n",
      "                else:\n",
      "                    raise LineError(\"0025\")\n",
      "            except LineError as er:\n",
      "                entry.append(\"[\" + str(lineCount) +\"] \" + er.returnError())\n",
      "        \n",
      "        else:\n",
      "            continue                   #this should not happen, but just in case, skip to the next line\n",
      "        \n",
      "                \n",
      "    Errors[\"entry\"] = entry                    #add the list of entry errors to the Errors dictionary\n",
      "\n",
      "    return \"clean\"\n",
      "    \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "global Errors\n",
      "Errors =dict()\n",
      "Keylist,holder = sortGff3(\"docs/Phabio_correct.gff3\")\n",
      "print \"returned from sortGff\"\n",
      "seq = fastaRead(\"docs/Phabio.fasta\")\n",
      "\n",
      "r = fileCheck(Keylist, holder, seq, types = ['gene','mRNA','exon'])\n",
      "print Errors"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "coutnerKeys are: ['24951', '671', '445', '41']\n",
        "returned from sortGff\n",
        "{'entry': [], 'structure': []}\n"
       ]
      }
     ],
     "prompt_number": 27
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def charCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def charCheck(str, search=re.compile(r'[^a-zA-Z0-9.=;_]').search):\n",
      "    \"\"\"\n",
      "    Checks a string for characters not a-zA-Z0-9.=; and returns true if such a character is\n",
      "    found.\n",
      "\n",
      "    Parameters:\n",
      "    -'str': string being checked.\n",
      "    \"\"\"\n",
      "    return bool(search(str))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def geneCheck"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def geneCheck(coord1, coord2, seq, count):       \n",
      "    \"\"\"\n",
      "    Checks if coordinates of sequence given is a proper gene. Writes errors to global list\n",
      "    'Errors'.\n",
      "    \n",
      "    Proper gene:\n",
      "    - can be translated to a protein.\n",
      "    - has a stop codon.\n",
      "    - no internal stop codon.\n",
      "    - begins with a start codon.\n",
      "    \n",
      "    Parameters:\n",
      "    -'coord1': start coordinate of gene.\n",
      "    -'coord2': second coordinate of gene.\n",
      "    -'Seq': string nucleotide sequence.\n",
      "    -'count': line in the file the gene is from.\n",
      "    \"\"\"\n",
      "    #FIXME: need to add check that all coord are within the range of the contig\n",
      "    biology = dict()                     #dictionary of errors based on biology like start codon, stop codon\n",
      "    gene = seq[coord1-1:coord2]\n",
      "    try:\n",
      "        if len(gene)%3 != 0:\n",
      "            raise BiologyError(\"0050\") ## divisible by three\n",
      "    except BiologyError as er:\n",
      "        biology.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "    \n",
      "    try:\n",
      "        protein = translate(gene) ## errors in translating the gene\n",
      "    except:\n",
      "        biology.append(\"[\" + str(count) + \"]  Biology Error: Ensure sequence provided is a nucleotide sequence.\")\n",
      "        return\n",
      "\n",
      "    Start = True\n",
      "    startCodon = seq[coord1-1:coord1+2]\n",
      "    if startCodon == \"ATG\" or startCodon == \"TTG\" or startCodon == \"GTG\":\n",
      "        Start = False\n",
      "    \n",
      "    try:\n",
      "        if protein.count(\"*\") == 0:\n",
      "            raise BiologyError(\"0030\") ## no stop codon\n",
      "        if protein.count(\"*\") > 1:\n",
      "            raise BiologyError(\"0040\") ## internal stop codons\n",
      "        if Start:\n",
      "            raise BiologyError(\"0020\") ## has to start with M\n",
      "    except BiologyError as er:\n",
      "        biology.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "    \n",
      "    Errors[\"biology\"] = biology                    #add the list of biology errors to the Errors dictionary\n",
      "    return\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def fastaRead"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fastaRead(fasta_File):\n",
      "    \"\"\"\n",
      "    Reads in a nucleotide sequence from a text file in .fasta format.\n",
      "\n",
      "    Parameters:\n",
      "    -'fasta_File': uploaded text file in .fasta format\n",
      "\n",
      "    Output:\n",
      "    -'seq': string of a nucleotide sequence.\n",
      "    \"\"\"\n",
      "\n",
      "    seq = []\n",
      "    try:\n",
      "        f1 = open(fasta_File, \"r\")\n",
      "    except:\n",
      "        return \"Error: unable to read fasta file.\"\n",
      "    \n",
      "    for line in f1:\n",
      "        line = line.rstrip()\n",
      "        if line.startswith(\">\"):\n",
      "            continue\n",
      "        else:\n",
      "            seq.append(line)\n",
      "\n",
      "    f1.close()\n",
      "    return ''.join(seq)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def translate"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def translate(nucSeq):\n",
      "    \"\"\"\n",
      "    Translates a nucleotide sequence into a protein sequence. Uses a dictionary of nucleotides\n",
      "    paired with the protein they translate for.\n",
      "\n",
      "    Parameters:\n",
      "    -'nucSeq': string of nucleotides.\n",
      "\n",
      "    Output:\n",
      "    -'protSeq': string of proteins.\n",
      "    \"\"\"\n",
      "    \n",
      "    codonLib = {'TTT':'F','TTC':'F','TTA':'L','TTG':'L','CTT':'L','CTC':'L','CTA':'L',\n",
      "                'CTG':'L','ATT':'I','ATC':'I','ATA':'I','ATG':'M','GTT':'V','GTC':'V',\n",
      "                'GTA':'V','GTG':'V','TCT':'S','TCC':'S','TCA':'S','TCG':'S','CCT':'P',\n",
      "                'CCC':'P','CCA':'P','CCG':'P','ACT':'T','ACC':'T','ACA':'T','ACG':'T',\n",
      "                'GCT':'A','GCC':'A','GCA':'A','GCG':'A','TAT':'Y','TAC':'Y','TAA':'*',\n",
      "                'TAG':'*','CAT':'H','CAC':'H','CAA':'Q','CAG':'Q','AAT':'N','AAC':'N',\n",
      "                'AAA':'K','AAG':'K','GAT':'D','GAC':'D','GAA':'E','GAG':'E','TGT':'C',\n",
      "                'TGC':'C','TGA':'*','TGG':'W','CGT':'R','CGC':'R','CGA':'R','CGG':'R',\n",
      "                'AGT':'S','AGC':'S','AGA':'R','AGG':'R','GGT':'G','GGC':'G','GGA':'G','GGG':'G'}\n",
      "    \n",
      "    cntLoc = 0\n",
      "    cntCodon = 0\n",
      "    codon = ''\n",
      "    protSeq = ''\n",
      "    while cntLoc < len(nucSeq):\n",
      "        codon = codon + nucSeq[cntLoc]\n",
      "        cntCodon += 1\n",
      "        if cntCodon == 3:\n",
      "            temp = codonLib[codon]\n",
      "            protSeq = protSeq + temp\n",
      "            codon = ''\n",
      "            cntCodon = 0\n",
      "        cntLoc += 1\n",
      "        \n",
      "    return protSeq"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "def outFile"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def outFile(keyS, holderS, nameS, errors, nameE, incLine):\n",
      "    \"\"\"\n",
      "    Writes a sorted gff file and an errors text file.\n",
      "    \n",
      "    Parameters:\n",
      "    -'keyS': list of sorted keys.\n",
      "    -'holderS': dictionary of lines from the gff file paired with keys in 'keyS'.\n",
      "    -'nameS': name of new sorted gff file. Also location where the file is written.\n",
      "    -'errors': list of all errors found in uploaded files to be written to text file.\n",
      "    -'nameE': name of new errors text file. Also location where the file is written.\n",
      "    -'incLine': boolean indicating whether lines from the sorted file should be included in \n",
      "                errors file.\n",
      "    \"\"\"\n",
      "    outE = nameE\n",
      "    outS = nameS\n",
      "   \n",
      "    if incLine:\n",
      "       for item in errors:\n",
      "          if \"Coordinate\" in item:\n",
      "             tempE = item.split(\" \")[1]\n",
      "             for key in keyS:\n",
      "                if tempE == key.split(\"_\")[1]:\n",
      "                   outE.write(holderS[key])\n",
      "                   break\n",
      "          else:\n",
      "             tempE = int(item[1])\n",
      "             outE.write(holderS[keyS[tempE]])\n",
      "         \n",
      "          outE.write(item)\n",
      "          outE.write(\"\\n\")\n",
      "          outE.write(\"\\n\")\n",
      "    else:\n",
      "       for item in errors:\n",
      "          outE.write(item)\n",
      "          outE.write(\"\\n\")\n",
      "   \n",
      "    for key in keyS:\n",
      "       outS.write(holderS[key])\n",
      "       outS.write(\"\\n\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Define Error Classes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class ValidationError(Exception):\n",
      "    \"\"\"\n",
      "    Exception classes:\n",
      "    - 'Validation Error': basic exception class. Used to catch unknown errors. \n",
      "    - 'Format Error': catches errors dealing with line format.\n",
      "    - 'Line Error': catches errors dealing with individuel components of a line.\n",
      "    - 'Biology Error': catches errors dealing with the described genes.\n",
      "    \"\"\"\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        _dict = dict()\n",
      "        _dict['0000'] = \"Usage Error: Unknown error code given used. Use printDict() to see dictionary of errors\"\n",
      "        _dict['1000'] = \"Validation Error: unknown\"\n",
      "        \n",
      "        self.code = str(code)\n",
      "        self._dict = _dict\n",
      "        self.message = message\n",
      "        \n",
      "        super(ValidationError, self).__init__(code, message)\n",
      "        \n",
      "    def printDict(self):\n",
      "        for k, v in self._dict.iteritems():\n",
      "            print k + \":  \" + v\n",
      "            \n",
      "    def returnError(self):\n",
      "        try:\n",
      "            return self._dict[self.code]\n",
      "        except:\n",
      "            return self._dict['0000']\n",
      "\n",
      "class FormatError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(FormatError, self).__init__(code, message)\n",
      "        \n",
      "        self._dict['0100'] = \"Format Error: less than 9 columns, no other checks done.\"\n",
      "        self._dict['0200'] = \"Format Error: more than 9 columns, no other checks done.\"\n",
      "        self._dict['0300'] = \"Format Error: tab deliminated.\"\n",
      "        self._dict['0400'] = \"Format Error: multiple contigs. There should only be one contig line per file.\"\n",
      "        self._dict['0500'] = \"Format Error: Errors in, or improper number of types for one location. Default types are [gene, mRNA, exon]\"\n",
      "        self._dict['0600'] = \"Format Error: ensure only one sequence in fasta file.\"\n",
      "        self._dict['0700'] = \"Format Error: invalid type in column 3; must be EXACTLY 'gene' or 'mRNA' or 'exon'.\"\n",
      "        \n",
      "        \n",
      "class LineError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(LineError, self).__init__(code, message)\n",
      "\n",
      "        self._dict['0001'] = \"Line Error: unknown\"\n",
      "        self._dict['0002'] = \"Line Error: 1st component = restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0003'] = \"Line Error: 2nd component = restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0004'] = \"Line Error: 3rd component = type given not found, must exactly match one of [gene, mRNA, exon].\"\n",
      "        self._dict['0005'] = \"Line Error: 4th component = 1st coordinate must be a positive integer.\"\n",
      "        self._dict['0006'] = \"Line Error: 5th component = 2nd coordinate must both an integer and be greater than the 1st coordinate.\"\n",
      "        self._dict['0007'] = \"Line Error: 6th component = must be a number or '.'\"\n",
      "        self._dict['0008'] = \"Line Error: 7th component = must indicate strand as either '+' or '-'; no other characters allowed\"\n",
      "        self._dict['0009'] = \"Line Error: 8th component = must be '.'\"\n",
      "        self._dict['0011'] = \"Line Error: 9th component = restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0012'] = \"Line Error: 9th component = ID already used. Each gene must have a unique ID.\"\n",
      "        self._dict['0013'] = \"Line Error: 9th component = Item must have an entry for ID\"\n",
      "        self._dict['0021'] = \"Line Error: 9th component = bad sort or a gene does not have a line for each type. Default types are [gene, mRNA, exon].\"\n",
      "        self._dict['0022'] = \"Line Error: 9th component = restrict characters used to a-Z/0-9/./=/;/_\"\n",
      "        self._dict['0023'] = \"Line Error: 9th component = ID already used. Each ID must be unique.\"\n",
      "        self._dict['0024'] = \"Line Error: 9th component = last type, must at least have a Parent.\"\n",
      "        self._dict['0025'] = \"Line Error: 9th component = Lines of this type must have a Parent.\"\n",
      "        self._dict['0026'] = \"Line Error: First line must be EXACTLY: '##gff-version 3' (HINT: check for unseen spaces or tabs)\"\n",
      "        self._dict['0027'] = \"Line Error: 9th component = items in this column must be in the form <name>=<value>, separated by ';'\"\n",
      "        self._dict['0028'] = \"Line Error: 9th component = items must be separated by ';', too many '=' without enough ';' \"\n",
      "\n",
      "        \n",
      "class BiologyError(ValidationError):\n",
      "    def __init__(self, code, message = \"\"):\n",
      "        \n",
      "        super(BiologyError, self).__init__(code, message)\n",
      "        \n",
      "        self._dict['0010'] = \"Biology Error: unknown\"\n",
      "        self._dict['0020'] = \"Biology Error: Incorrect start codon. Must start with M\"\n",
      "        self._dict['0030'] = \"Biology Error: No stop codon detected.\"\n",
      "        self._dict['0040'] = \"Biology Error: Internal stop codons\"\n",
      "        self._dict['0050'] = \"Biology Error: total nucleotide count not divisible by three.\"\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Main"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def main(gff3_File, seq_File):\n",
      "    global Errors\n",
      "    Errors = dict()\n",
      "    seq = fastaRead(seq_File)\n",
      "    \n",
      "    sorted_File = sortGff3(gff3_File)       \n",
      "    \n",
      "    report = fileCheck(sorted_File, seq)\n",
      "    \n",
      "    newName = gff3_File[0:-5] + \"_sorted_Errors.txt\"\n",
      "    out = open(newName,\"w\")\n",
      "    for item in Errors:\n",
      "        out.write(item)\n",
      "        out.write(\"\\n\")\n",
      "    out.close()\n",
      "    return [sorted_File, newName]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "test code"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "read fasta"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "phabioSeq = fastaRead(\"docs/Phabio.fasta\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "phabioSeq[40:371]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 21,
       "text": [
        "'GTGCCGTTCGCCAACACGACTGTTCGGAATCGTCGCCGCGCTCAGGTTCGTCAGCGTGACGGTGATGCTCCATGCGCGTTGCGGATCACGGCGGATTGCCAGGCACTGGGCGGGGTTATCGACTATGACGCTCGCCCACCTCATCCGCGGTCGTTTGAGGTTGACCACATCGTGAGTTCTGTTGAGGCCGCGCGACTGGGCTGGTCTCAGGAAGATTCGGATGGGCTGGATAACTGCCAGGCAGTGTGTCGGCAGTGTAATCGCGCGAAGTCGTCTGGGGATCGTGCTGTCCCGGAAGTGAGGGAGTCGTACGTGAATCCGCGCTTTGCCT'"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "test sortGFF3"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "global Errors\n",
      "Errors = []\n",
      "\n",
      "#f1 = open(\"Jerm2-final-genes.txt\", \"r\")\n",
      "#for line in f1:\n",
      "#    theLine = line.strip().split(\"\\t\")\n",
      "#    for comp in theLine:\n",
      "#        print comp\n",
      "#    print \"\\n\\n\"\n",
      "\n",
      "derp = sortGff3(\"Jerm2-final-genes.txt\")\n",
      "print derp[1]['gene_599']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fileCheck(\"Jerm2-final-gene_sorted.gff3\", \"Jerm2.fasta\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Errors=[]\n",
      "derp = fastaRead(\"Jerm2.fasta\")\n",
      "geneCheck(31199,31429,derp,21)\n",
      "print Errors"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene = \"N/A\"\n",
      "smlCount = 0\n",
      "\n",
      "for theSeq in SeqIO.parse(\"Jerm2.fasta\", \"fasta\"):\n",
      "    gene = theSeq\n",
      "gene.seq.translate() \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Errors = []\n",
      "prot = geneCheck(1, 60, \"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAAATGATGATGATGATGATGATG\", 35)\n",
      "print Errors"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(\"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAA\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def geneCheckBio(coord1, coord2, seq, count):\n",
      "    gene = \"N/A\"\n",
      "    smlCount = 0\n",
      "    \n",
      "    try:\n",
      "        for theSeq in SeqIO.parse(seq, \"fasta\"):\n",
      "            gene = theSeq[coord1:coord2]\n",
      "            smlCount += 1\n",
      "        if smlCount != 1:\n",
      "            raise FormatError(\"0600\")\n",
      "    except FormatError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "        return\n",
      "    except:\n",
      "        Errors.append(\"Error in parsing the fasta file. Ensure correct fasta format\")\n",
      "        return\n",
      "    \n",
      "    try:\n",
      "        if len(gene)%3 != 0:\n",
      "            raise BiologyError(\"0050\") ## divisible by three\n",
      "    except BiologyError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "        \n",
      "    try:\n",
      "        protein = gene.seq.translate() ## errors in translating the gene\n",
      "    except:\n",
      "        Errors.append(\"[\" + str(count) + \"]  Biology Error: unable to translate sequence. Ensure sequence provided is a nucleotide sequence.\")\n",
      "        return\n",
      "    \n",
      "    try:\n",
      "        if protein.count(\"*\") == 0:\n",
      "            raise BiologyError(\"0030\") ## no stop codon\n",
      "        if protein.count(\"*\") > 1:\n",
      "            raise BiologyError(\"0040\") ## internal stop codons\n",
      "        if protein[0] != \"M\":\n",
      "            raise BiologyError(\"0020\") ## has to start with M\n",
      "    except BiologyError as er:\n",
      "        Errors.append(\"[\" + str(count) +\"] \" + er.returnError())\n",
      "    \n",
      "    return "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "81%3"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(\"TTGTCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATCATAAATGATGATGATGATGATGATG\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def errorsToFile(errorsList, gff3_File_name):\n",
      "    \n",
      "    \n",
      "    \n",
      "    newName = gff3_File_name[0:-5] + \"_sorted_Errors.txt\"\n",
      "    out = open(newName,\"w\")\n",
      "    for item in errorsList:\n",
      "        out.write(item)\n",
      "    out.close()\n",
      "    return newName\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def main2():\n",
      "    global bigG\n",
      "    bigG = ['derp','merp']\n",
      "    sec()\n",
      "    printGl()\n",
      "    \n",
      "def sec():\n",
      "    bigG.append('poke')\n",
      "    \n",
      "def printGl():\n",
      "    print bigG"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "once = ['one','two','three']\n",
      "once.append('fourth')\n",
      "print once[3]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cay = False\n",
      "\n",
      "try:\n",
      "    if cay:\n",
      "        print \"nom\"\n",
      "    else:\n",
      "        raise ValidationError(\"1000\")\n",
      "except ValidationError as e:\n",
      "    print \"HELLO \" + e.returnError()\n",
      "except:\n",
      "    print \"Missed\"\n",
      "print \"END\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "kay = False\n",
      "\n",
      "try:\n",
      "    if kay:\n",
      "        print \"om\"\n",
      "    else:\n",
      "        raise FormatError(\"100\")\n",
      "except FormatError as e:\n",
      "    print e.printDict()\n",
      "except:\n",
      "    print \"missed me\"\n",
      "print \"end\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "try:\n",
      "    raise FormatError(\"010\")\n",
      "except FormatError as e:\n",
      "    print \"meep\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "prac = dict()\n",
      "prac[\"0300\"] = \"5\"\n",
      "prac[\"0001\"] = \"6\"\n",
      "prac[\"0002\"] = \"7\"\n",
      "prac[\"0003\"] = \"8\"\n",
      "prac[\"0000\"] = \"1\"\n",
      "prac[\"1000\"] = \"2\"\n",
      "prac[\"0100\"] = \"3\"\n",
      "prac[\"0200\"] = \"4\"\n",
      "prac[\"0010\"] = \"11\"\n",
      "prac[\"0020\"] = \"12\"\n",
      "prac[\"0030\"] = \"13\"\n",
      "prac[\"0040\"] = \"14\"\n",
      "\n",
      "prac.keys()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "listKeys = prac.keys()\n",
      "codeNum = len(listKeys)\n",
      "codeLen = len(listKeys[0])\n",
      "prevKey = \"\"\n",
      "index = 0\n",
      "while index < codeNum:\n",
      "    currKey = listKeys[index]\n",
      "    if index == 0:\n",
      "        prevKey = currKey\n",
      "        index += 1\n",
      "        continue\n",
      "    else:\n",
      "        for y in range(0, codeLen):\n",
      "            if currKey[y] == prevKey[y]:\n",
      "                continue\n",
      "            elif currKey[y] > prevKey[y]:\n",
      "                listKeys[index-1] = currKey\n",
      "                listKeys[index] = prevKey\n",
      "                index -= 2\n",
      "                break\n",
      "            else:\n",
      "                continue\n",
      "        index += 1\n",
      "        prevKey = currKey\n",
      "for key in listKeys:\n",
      "    print key\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dir()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}